// @teambit/react.mounter@1.1.5/dist/mounter.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import React from 'react';
import { ErrorBoundary } from 'react-error-boundary';
import { HighlighterProvider } from '@teambit/react.ui.highlighter-provider';
import { needLiveControls, LivePreview } from '@bitdev/react.live-controls';
import { getDefaultControlsHook } from '@teambit/react.ui.mounter.use-default-controls';
import { getLiveControls } from '@teambit/react.modules.react-live-controls-from-schema';
// Prepare <AutoLiveControlsPreview /> to wrap <LivePreview />
// with auto controls feature.
const useDefaultControls = getDefaultControlsHook(getLiveControls);
function AutoLiveControlsPreview(props) {
    const resultProps = { ...props };
    const enabled = !resultProps.controls && !resultProps.noControls;
    const defaultControls = useDefaultControls(enabled);
    if (enabled && defaultControls.length) {
        resultProps.controls = defaultControls;
    }
    return (_jsx(LivePreview, { ...resultProps }));
}
// TODO: remove when stable.
function activateLiveControls() {
    if (window.parent) {
        window.parent.postMessage({
            type: 'composition-live-controls:activate',
        }, '*');
    }
}
activateLiveControls();
// Both imports are dynamic so the bundler never statically validates named exports
// against whichever React version is installed.
// React 18/19: react-dom/client resolves → use createRoot.
// React 17:    react-dom/client rejects  → fall back to legacy ReactDOM.render.
let createRootImpl = null;
import('react-dom/client')
    .then((m) => {
    createRootImpl = (container) => m.createRoot(container);
})
    .catch(() => {
    import('react-dom').then((m) => {
        createRootImpl = (container) => ({
            render: (element) => m.render(element, container),
            unmount: () => m.unmountComponentAtNode(container),
        });
    });
});
function createCompatRoot(container) {
    if (createRootImpl)
        return createRootImpl(container);
    throw new Error('[react-compat] createCompatRoot called before module init settled');
}
/**
 * a reference to a root instance.
 */
let root = null;
/**
 * a reference to the root document.
 */
let rootDoc = null;
/**
 * create a mounter.
 */
export function createMounter(Provider = React.Fragment) {
    return (Composition) => {
        const isLiveComposition = typeof Composition === 'object';
        if (isLiveComposition && !needLiveControls(window.location)) {
            // eslint-disable-next-line no-param-reassign
            Composition.noControls = true;
        }
        if (root && rootDoc) {
            // Unmount the existing root before mounting a new one.
            root.unmount();
            root = null;
            rootDoc = null;
        }
        rootDoc = document.getElementById('root');
        root = createCompatRoot(rootDoc);
        root.render(
        // @ts-ignore TODO: remove this after release.
        _jsx(ErrorBoundary, { fallback: _jsx(_Fragment, {}), children: _jsx(HighlighterProvider, { children: _jsx(Provider, { children: isLiveComposition ? (_jsx(AutoLiveControlsPreview, { ...Composition })) : (_jsx(Composition, {})) }) }) }));
    };
}
export default createMounter;
