// @teambit/react.ui.mounter.use-default-controls@0.0.2/dist/use-default-controls.js
/* eslint-disable consistent-return */
import React from 'react';
const DEFAULT_CONTROLS_REQUEST = 'composition-live-controls:request-default-controls';
const DEFAULT_CONTROLS_RESPONSE = 'composition-live-controls:default-controls';
function getCompositionHref() {
    const { hash } = window.location;
    const componentId = hash.split('?')[0].slice(1);
    const params = hash
        .split('?')[1]
        .split('&')
        .map((x) => x.split('='))
        .reduce((acc, [key, value]) => {
        acc[key] = value;
        return acc;
    }, {});
    if (params.preview === 'compositions' && params.livecontrols) {
        return JSON.stringify({
            id: componentId,
            name: params.name,
        });
    }
    return null;
}
export function getDefaultControlsHook(defaultControlsParser) {
    return function useDefaultControls(enabled) {
        const [defaultControls, setDefaultControls] = React.useState([]);
        const currentHref = getCompositionHref();
        React.useEffect(() => {
            if (!currentHref)
                return;
            if (!enabled)
                return;
            const { id, name } = JSON.parse(currentHref);
            function requestDefaultControls() {
                if (!window.parent || window.parent === window)
                    return;
                window.parent.postMessage({
                    type: DEFAULT_CONTROLS_REQUEST,
                    payload: { id, name },
                }, '*');
            }
            function onMessage(event) {
                if (event.source !== window.parent)
                    return;
                const { data } = event;
                if (!data?.type || data.type !== DEFAULT_CONTROLS_RESPONSE)
                    return;
                const payload = data.payload;
                if (!payload)
                    return;
                if (payload.id !== id || payload.name !== name)
                    return;
                if (Array.isArray(payload.controls)) {
                    setDefaultControls(payload.controls);
                    return;
                }
                if (payload.schema) {
                    const schema = defaultControlsParser(payload.schema);
                    if (schema[name]) {
                        setDefaultControls(schema[name]);
                    }
                }
            }
            window.addEventListener('message', onMessage);
            function reset() {
                requestDefaultControls();
            }
            // webpack hmr manual integration
            // @ts-ignore
            if (typeof module !== 'undefined') {
                // @ts-ignore
                const webpackHot = module?.hot;
                if (webpackHot?.addStatusHandler) {
                    webpackHot.addStatusHandler((status) => {
                        // React to the current status...
                        if (status === 'ready') {
                            reset();
                        }
                    });
                }
            }
            // Vite hmr manual integration
            if (import.meta.hot) {
                // Also listen to Vite-specific afterUpdate event (if available)
                import.meta.hot.on?.('vite:afterUpdate', reset);
            }
            reset();
            return () => {
                window.removeEventListener('message', onMessage);
            };
        }, [enabled, currentHref]);
        return defaultControls;
    };
}
