// html-webpack-inject-plugin@5.3.1/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _htmlWebpackPlugin = _interopRequireDefault(require("html-webpack-plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class HtmlWebpackInjectPlugin {
  constructor(config) {
    _defineProperty(this, "assets", void 0);

    _defineProperty(this, "prepend", void 0);

    _defineProperty(this, "apply", compiler => {
      compiler.hooks.compilation.tap('HtmlWebpackInjectPlugin', compilation => {
        const hooks = _htmlWebpackPlugin.default.getHooks(compilation).alterAssetTags;

        hooks.tapAsync('HtmlWebpackInjectPlugin', (htmlPluginData, cb) => {
          this.assets.filter(({
            include
          }) => {
            if (include) {
              if (Array.isArray(include)) {
                return include.includes(htmlPluginData.outputName);
              } else {
                return include(htmlPluginData.outputName);
              }
            } else {
              return true;
            }
          }).forEach(asset => {
            const tagName = asset.tagName !== 'meta' ? `${asset.tagName}s` : asset.tagName;
            const tags = htmlPluginData.assetTags[tagName];
            htmlPluginData.assetTags[tagName] = this.prepend ? [asset].concat(tags) : tags.concat(asset);
          });
          return cb(null, htmlPluginData);
        });
      });
    });

    const {
      externals = [],
      prepend = false
    } = config;
    this.assets = externals.map(({
      tag = 'meta',
      tagName,
      attributes,
      attrs = {},
      innerHTML,
      voidTag = false,
      include
    }) => {
      return {
        tagName: tagName || tag,
        attributes: attributes || attrs,
        innerHTML,
        voidTag,
        meta: {
          plugin: 'HtmlWebpackInjectPlugin'
        },
        include
      };
    });
    this.prepend = prepend;
  }

}

exports.default = HtmlWebpackInjectPlugin;
