// @teambit/react.modules.react-live-controls-from-schema@0.0.2/dist/react-live-controls-from-schema.js
import { dirname, join } from 'path';
export class LiveCompositionSchema {
    constructor(schema) {
        this.schema = schema;
        this.exports = {};
        this.initAllExports();
        // console.log('Exports:', this.exports);
    }
    analyze() {
        // 1. get compositions
        // 2. filter out non-react
        // 3. get props type
        // 4. get controls type
        const output = {};
        const compositions = this.getCompositions();
        const liveCompositions = compositions.filter(this.isLiveComposition);
        liveCompositions.forEach((liveComposition) => {
            const name = this.getCompositionName(liveComposition);
            const componentNode = this.getLiveComponent(liveComposition);
            if (!this.isReactComponent(componentNode))
                return;
            const propsTypeNode = this.getPropsType(componentNode);
            if (!propsTypeNode)
                return;
            const resolvedPropsTypeNode = this.resolvePropsType(propsTypeNode);
            if (!resolvedPropsTypeNode)
                return;
            const controlsType = this.getControlsType(resolvedPropsTypeNode);
            if (!controlsType)
                return;
            output[name] = controlsType;
        });
        return output;
    }
    initAllExports() {
        this.schema.module.exports.forEach((node) => {
            const { name, exportNode } = node;
            const { filePath } = exportNode.location;
            this.setExport(filePath, name, exportNode);
        });
        this.schema.internals.forEach((moduleNode) => {
            const { filePath } = moduleNode.location;
            moduleNode.exports.forEach((node) => {
                const { name } = node;
                this.setExport(filePath, name, node);
            });
            moduleNode.internals.forEach((node) => {
                if (node.exports) {
                    node.exports.forEach((subNode) => {
                        const { name } = subNode;
                        this.setExport(filePath, name, subNode);
                    });
                }
                else {
                    const { name } = node;
                    this.setExport(filePath, name, node);
                }
            });
        });
    }
    setExport(filePath, name, node) {
        // remove .js/.ts/.tsx/.jsx extension for easier lookup
        const resolvedFilePath = filePath.replace(/\.(js|ts|tsx|jsx)$/, '');
        const resolvedName = name.includes(' (default)') ? 'default' : name;
        if (!this.exports[resolvedFilePath]) {
            this.exports[resolvedFilePath] = {};
        }
        this.exports[resolvedFilePath][resolvedName] = node;
    }
    getExport(filePath, name) {
        const resolvedFilePath = filePath.replace(/\.(js|ts|tsx|jsx)$/, '');
        if (this.exports[resolvedFilePath] && this.exports[resolvedFilePath][name]) {
            return this.exports[resolvedFilePath][name];
        }
        return null;
    }
    getCompositions() {
        const compositions = this.schema.internals.find((node) => node.location.filePath.match(LiveCompositionSchema.compositionFilePattern));
        const exports = compositions ? compositions.exports : [];
        return exports.map((node) => {
            if (node.__schema === 'ExportSchema') {
                return node.exportNode;
            }
            return node;
        });
    }
    isLiveComposition(compositionNode) {
        // console.log('isLiveComposition', compositionNode, compositionNode.__schema);
        // 1. it's an object
        // 2. it has a 'component' member
        if (compositionNode.__schema === 'VariableLikeSchema' &&
            compositionNode.type?.__schema === 'ObjectLiteralExpressionSchema') {
            const componentMember = compositionNode.type.members.find((m) => m.name === 'component');
            return !!componentMember;
        }
        return false;
    }
    getCompositionName(compositionNode) {
        // console.log('getCompositionName', compositionNode.name);
        return compositionNode.name;
    }
    getLiveComponent(compositionNode) {
        // console.log('getLiveComponent', compositionNode.name, compositionNode.type);
        const componentMember = compositionNode.type.members.find((m) => m.name === 'component');
        if (componentMember.value?.__schema === 'TypeRefSchema') {
            const { name, sourceFilePath } = componentMember.value;
            const result = this.getExport(LiveCompositionSchema.resolveSourceFilePath(sourceFilePath, compositionNode.location.filePath), name);
            return result;
        }
        return null;
    }
    isReactComponent(node) {
        const returnType = node?.returnType;
        // console.log('isReactComponent', node, returnType);
        if (returnType?.packageName === 'react') {
            return LiveCompositionSchema.REACT_NODE_NAMES.includes(returnType.name);
        }
        return false;
    }
    getPropsType(reactComponentNode) {
        // console.log('getPropsType', reactComponentNode);
        const paramType = reactComponentNode?.props?.type || reactComponentNode?.params?.[0]?.type;
        return paramType;
    }
    resolvePropsType(propsType) {
        // console.log('resolvePropsType', propsType);
        if (propsType.__schema === 'TypeLiteralSchema') {
            const newTypeNode = {
                members: [],
                __schema: 'TypeLiteralSchema',
            };
            if (propsType.members) {
                newTypeNode.members.push(...propsType.members);
            }
            return newTypeNode;
        }
        if (propsType.__schema === 'TypeIntersectionSchema') {
            const newTypeNode = {
                members: [],
                __schema: 'TypeLiteralSchema',
            };
            propsType.types.forEach((typeNode) => {
                const resolvedTypeNode = this.resolvePropsType(typeNode);
                if (resolvedTypeNode && resolvedTypeNode.members) {
                    newTypeNode.members.push(...resolvedTypeNode.members);
                }
            });
            return newTypeNode;
        }
        if (propsType.__schema === 'TypeRefSchema') {
            const result = this.getExport(propsType.location.filePath, propsType.name);
            return result ? this.resolvePropsType(result.type) : null;
        }
        return null;
    }
    getControlsType(propsTypeNode) {
        // console.log('getControlsType', propsTypeNode);
        const controls = propsTypeNode.members.map((member) => {
            // console.log('member', member.name, this.getMemberType(member.type));
            const memberType = this.getMemberType(member.type);
            if (memberType) {
                if (!memberType.package && memberType.name) {
                    // internal type
                    if (memberType.name === 'select' && memberType.options) {
                        return { id: member.name, type: 'select', options: memberType.options };
                    }
                    return { id: member.name, type: memberType.name };
                }
                if (member.name === 'children' &&
                    memberType.package === 'react' &&
                    memberType.name === 'ReactNode') {
                    // external type (children: React.ReactNode)
                    return { id: member.name, type: 'string' };
                }
            }
            return null;
        }).filter(Boolean);
        return controls;
    }
    getMemberType(node) {
        // console.log('getMemberType', node);
        // type refs
        if (node.__schema === 'TypeRefSchema') {
            if (node.packageName) {
                return { name: node.name, package: node.packageName };
            }
            if (node.internalFilePath) {
                const filePath = node.internalFilePath;
                const internalType = this.getExport(filePath, node.name);
                if (!internalType)
                    return null;
                // console.log('  internalType', to, from, node.name, internalType);
                if (internalType.__schema === 'KeywordTypeSchema') {
                    return { name: internalType.name };
                }
                return null;
            }
        }
        // literal union
        // e.g. 'small' | 'medium' | "large"
        if (node.__schema === 'TypeUnionSchema') {
            const isStringUnion = node.types.every((t) => t.__schema === 'LiteralTypeSchema' &&
                LiveCompositionSchema.isLiteralString(t.name));
            if (isStringUnion) {
                return {
                    name: 'select',
                    options: node.types.map((t) => LiveCompositionSchema.getLiteralStringValue(t.name)).filter((x) => x !== null),
                };
            }
        }
        // keyword type
        // e.g. string, boolean, number, any
        if (node.__schema === 'KeywordTypeSchema') {
            if (['string', 'boolean', 'number'].includes(node.name)) {
                return { name: node.name };
            }
            return null;
        }
        // array type
        if (node.__schema === 'TypeArraySchema') {
            return null;
        }
        // others
        return null;
    }
    static isLiteralString(name) {
        // console.log('isLiteralString', name, /^'[^']*'$/.test(name));
        return /^'[^']*'$/.test(name) || /^"[^"]*"$/.test(name);
    }
    static getLiteralStringValue(name) {
        try {
            // eslint-disable-next-line no-eval
            return eval(name);
        }
        catch {
            return null;
        }
    }
    static resolveSourceFilePath(filePath, base) {
        const baseDirname = dirname(base);
        const resolvedPath = join(baseDirname, filePath);
        return resolvedPath;
    }
}
LiveCompositionSchema.compositionFilePattern = /\.compositions?\.\w+$/;
LiveCompositionSchema.REACT_NODE_NAMES = [
    'ReactNode',
    'JSX.Element',
    'React.ReactNode',
    'React.JSX.Element'
];
export function getLiveControls(schema) {
    const analyzer = new LiveCompositionSchema(schema);
    return analyzer.analyze();
}
