// @teambit/oxc.linter.oxlint-linter@2.0.4/oxlint-config-writer.ts
import fs from 'fs';
import crypto from 'crypto';
import type { EnvHandler } from '@teambit/envs';
import type {
  ConfigWriterEntry,
  ConfigFile,
  GenerateExtendingConfigFilesArgs,
  ExtendingConfigFile,
} from '@teambit/workspace-config-files';

const BIT_GENERATED_OXLINT_CONFIG_COMMENT = '// bit-generated-oxlint-config';

const sha1 = (str: string) => crypto.createHash('sha1').update(str).digest('hex');

export type OxlintConfigWriterOptions = {
  name?: string;
  /**
   * absolute path to the oxlint config file to use as the base config.
   */
  configPath: string;
};

export class OxlintConfigWriter implements ConfigWriterEntry {
  id = 'oxlint';

  patterns = ['**/.oxlintrc.json'];

  constructor(public name: string, private configPath: string) {}

  calcConfigFiles(): ConfigFile[] | undefined {
    const raw = fs.readFileSync(this.configPath, 'utf-8');
    // Strip $schema — the relative path in the source config is valid there but
    // would point to the wrong location once the file is copied to node_modules/.cache.
    const parsed = JSON.parse(raw);
    delete parsed.$schema;
    const content = JSON.stringify(parsed, null, 2);
    const hash = sha1(raw);
    return [{ name: `oxlintrc.bit.${hash}.json`, content, hash }];
  }

  generateExtendingFile({ writtenConfigFiles }: GenerateExtendingConfigFilesArgs): ExtendingConfigFile | undefined {
    const configFile = writtenConfigFiles.find((f) => f.name.includes('oxlintrc.bit.'));
    if (!configFile) return undefined;
    const config = { extends: [`{${configFile.name}}`] };
    const content = `${BIT_GENERATED_OXLINT_CONFIG_COMMENT}\n${JSON.stringify(config, null, 2)}`;
    return { content, name: '.oxlintrc.json', extendingTarget: configFile, useAbsPaths: false };
  }

  isBitGenerated(filePath: string): boolean {
    const content = fs.readFileSync(filePath, 'utf-8');
    return content.includes(BIT_GENERATED_OXLINT_CONFIG_COMMENT);
  }

  static from(options: OxlintConfigWriterOptions): { name: string; handler: EnvHandler<OxlintConfigWriter> } {
    const name = options.name || 'OxlintConfigWriter';
    return {
      name,
      handler: () => new OxlintConfigWriter(name, options.configPath),
    };
  }
}
