// @teambit/vite.vitest-tester@0.4.13/dist/coverage-utils.js
import { join } from 'path';
import minimatch from 'minimatch';
import { getCoverage, coverageByComponent } from '@teambit/defender.coverage-report';
const METRIC_KEY_SET = new Set(['lines', 'functions', 'branches', 'statements']);
const GENERAL_THRESHOLD_KEYS = new Set([
    'lines',
    'functions',
    'branches',
    'statements',
    '100',
    'perFile',
    'autoUpdate',
]);
function getRootDirMap(context) {
    const rootDirMap = {};
    const capsules = context?.capsuleNetwork?.seedersCapsules || [];
    capsules.forEach(capsule => {
        rootDirMap[capsule.component.id.toString()] = capsule.path;
    });
    return rootDirMap;
}
function matchPattern(filePath, rootDir, pattern) {
    const absoluteFilePath = join(rootDir, filePath);
    return minimatch(absoluteFilePath, pattern);
}
function collectMetricThresholds(source) {
    const thresholds = {};
    if (!source)
        return thresholds;
    const is100 = source[100];
    for (const metric of METRIC_KEY_SET) {
        const value = source[metric];
        if (typeof value === 'number' && Number.isFinite(value)) {
            thresholds[metric] = value;
        }
        else if (is100) {
            thresholds[metric] = 100;
        }
    }
    return thresholds;
}
function checkMetricThresholds(data, thresholds, createMessage) {
    const errors = [];
    if (!data)
        return errors;
    for (const metric of METRIC_KEY_SET) {
        const thresholdValue = thresholds[metric];
        if (typeof thresholdValue !== 'number')
            continue;
        const metricSummary = data?.[metric];
        const actualPct = typeof metricSummary?.pct === 'number' ? metricSummary.pct : undefined;
        if (actualPct === undefined)
            continue;
        if (actualPct === 0 && metricSummary.total === 0)
            continue;
        if (actualPct + Number.EPSILON < thresholdValue) {
            errors.push(new Error(createMessage(metric, actualPct, thresholdValue)));
        }
    }
    return errors;
}
export function applyThresholdsToResults(thresholds, rootDirMap, covByComponent, result) {
    const generalThresholds = collectMetricThresholds(thresholds);
    for (const [componentId, componentCoverage] of Object.entries(covByComponent)) {
        if (!componentCoverage)
            continue;
        const componentRootDir = rootDirMap[componentId];
        const coverageErrors = [];
        // Check general thresholds
        const generalCoverage = componentCoverage.total;
        if (generalCoverage && Object.keys(generalThresholds).length) {
            coverageErrors.push(...checkMetricThresholds(generalCoverage, generalThresholds, (metric, actualPct, thresholdValue) => `Coverage threshold not met for ${componentId} (${metric}: ${actualPct}% < ${thresholdValue}%)`));
        }
        // Check per-pattern thresholds
        const patternThresholds = Object.entries(thresholds).filter(([key]) => !GENERAL_THRESHOLD_KEYS.has(key)).map(([pattern, value]) => {
            return [pattern, collectMetricThresholds(value)];
        });
        if (patternThresholds.length && componentRootDir) {
            for (const [pattern, patternThreshold] of patternThresholds) {
                if (!patternThreshold)
                    continue;
                componentCoverage.files.forEach(file => {
                    if (!matchPattern(file.path, componentRootDir, pattern))
                        return;
                    coverageErrors.push(...checkMetricThresholds(file.data, patternThreshold, (metric, actualPct, thresholdValue) => `Coverage threshold not met for ${componentId} (${metric}: ${actualPct}% < ${thresholdValue}% in ${file.path} matching "${pattern}")`));
                });
            }
        }
        const componentResult = result.components.find(c => c.componentId.toString() === componentId);
        if (componentResult?.results) {
            Object.assign(componentResult.results, {
                coverage: componentCoverage,
            });
        }
        if (coverageErrors.length) {
            componentResult?.errors?.push(...coverageErrors);
        }
    }
    return result;
}
export async function checkCoverageThresholds(coverageOptions, context, result) {
    const { thresholds } = coverageOptions;
    if (!thresholds) {
        return result;
    }
    const rootDirMap = getRootDirMap(context);
    const covByComponent = coverageByComponent(context, await getCoverage(context));
    return applyThresholdsToResults(thresholds, rootDirMap, covByComponent, result);
}
