// @teambit/vite.vitest-tester@0.4.13/dist/report-utils.js
import { relative } from 'path';
import minimatch from 'minimatch';
import lodash from 'lodash';
import { Tests, } from '@teambit/tester';
import { TestsFiles, TestResult, TestsResult } from '@teambit/tests-results';
import { ComponentMap } from '@teambit/component';
import { VitestError } from './vitest-error.js';
import { getPatternsMap, resolveComponentPattern } from './config-utils.js';
const { compact } = lodash;
// output utils
const getTaskState = (status) => {
    switch (status) {
        case 'pass':
            return 'passed';
        case 'fail':
            return 'failed';
        case 'skip':
            return 'skipped';
        case 'todo':
            return 'todo';
        default:
            return 'pending';
    }
};
const traverseTask = (task, ancestor = []) => {
    const resultList = [];
    const errorList = [];
    const info = {
        resultList,
        pass: 0,
        failed: 0,
        pending: 0,
        duration: 0,
        slow: false, // TODO: calculate slow
        errorList,
    };
    if (task.type === 'test') {
        const error = task.result?.errors?.[0];
        const errorStr = error?.message || error?.stack;
        const status = getTaskState(task.result?.state);
        switch (status) {
            case 'passed':
                info.pass += 1;
                break;
            case 'failed':
                info.failed += 1;
                break;
            case 'pending':
                info.pending += 1;
                break;
            default:
                break;
        }
        const isFailure = status === 'failed';
        info.duration += task.result?.duration || 0;
        resultList.push(new TestResult(ancestor, task.name, status, task.result?.duration, isFailure ? undefined : errorStr, isFailure ? errorStr : undefined));
    }
    if (task.type === 'suite') {
        errorList.push(...(task.result?.errors || []));
        task.tasks.forEach(subTask => {
            const subInfo = traverseTask(subTask, [...ancestor]);
            info.pass += subInfo.pass;
            info.failed += subInfo.failed;
            info.pending += subInfo.pending;
            info.duration += subInfo.duration;
            info.slow = info.slow || subInfo.slow;
            resultList.push(...subInfo.resultList);
            errorList.push(...subInfo.errorList);
        });
    }
    return info;
};
const genErrorsFromTests = (files, unhandledErrors) => {
    const errors = [];
    files.forEach(file => {
        file.tests.forEach(test => {
            if (test.error) {
                errors.push(new VitestError(test.error));
            }
            if (test.failure) {
                errors.push(new VitestError(test.failure));
            }
        });
    });
    unhandledErrors.forEach((error) => {
        errors.push(error);
    });
    return errors;
};
export const getResultFromFiles = (files, errors, opts, context) => {
    const patternsMap = getPatternsMap(opts, context);
    const componentMap = ComponentMap.as(patternsMap.components, (component) => {
        const componentPatternValue = patternsMap.get(component);
        if (!componentPatternValue)
            return [];
        const [currComponent, patternEntry] = componentPatternValue;
        const resolvedPatterns = resolveComponentPattern(currComponent, patternEntry, opts, context);
        const matchedFiles = files.filter(file => {
            // Split patterns into include and exclude
            const includePatterns = resolvedPatterns.filter(p => !p.startsWith('!'));
            const excludePatterns = resolvedPatterns.filter(p => p.startsWith('!')).map(p => p.slice(1));
            // File must match at least one include pattern
            const matchesInclude = includePatterns.some(pattern => minimatch(file.filepath, pattern));
            // File must not match any exclude patterns
            const matchesExclude = excludePatterns.some(pattern => minimatch(file.filepath, pattern));
            const matches = matchesInclude && !matchesExclude;
            return matches;
        });
        return matchedFiles;
    });
    const componentResults = componentMap.toArray().map(([component, componentFiles]) => {
        const componentPatternValue = patternsMap.get(component);
        if (!componentPatternValue)
            return undefined;
        const [, patternEntry] = componentPatternValue;
        const tests = [];
        const fileErrors = [];
        componentFiles.forEach(file => {
            const { pass, failed, pending, duration, slow, resultList, errorList, } = traverseTask(file);
            let fileError;
            if (errorList.length > 0) {
                const errorMessage = errorList.map(e => e.message || e.stack || '').join('\n');
                const errorStack = errorList.map(e => e.stack || '').join('\n');
                // Convert to plain object for JSON serialization
                fileError = {
                    message: errorMessage,
                    stack: errorStack,
                    name: 'VitestError'
                };
                fileErrors.push(new VitestError(errorMessage, errorStack));
            }
            tests.push(new TestsFiles(relative(patternEntry.componentDir, file.filepath), resultList, pass, failed, pending, duration, slow, fileError));
        });
        const totalErrors = genErrorsFromTests(tests, [...errors, ...fileErrors]);
        return {
            componentId: component.id,
            results: new TestsResult(tests, totalErrors.length === 0, 0),
            errors: totalErrors,
        };
    });
    const result = compact(componentResults);
    return new Tests(result);
};
