// @teambit/vite.vitest-tester@0.4.13/dist/vitest-tester.js
/* eslint-disable max-classes-per-file */
import { createRequire } from 'module';
import fsExtra from 'fs-extra';
import { proxy } from 'comlink';
import { WorkspaceAspect } from '@teambit/workspace';
import { DependencyResolverAspect } from '@teambit/dependency-resolver';
import { DevFilesAspect } from '@teambit/dev-files';
import { startVitest, VitestPackageInstaller } from 'vitest/node';
import { patternsToArray, getExclude, getConfig, } from './config-utils.js';
import { getResultFromFiles, } from './report-utils.js';
import { checkCoverageThresholds, } from './coverage-utils.js';
const { readFileSync, readJsonSync } = fsExtra;
const require = createRequire(import.meta.url);
export class VitestTester {
    constructor(id, configPath, vitestWorker, logger, workspace, depResolver, devFiles, opts = {}) {
        this.id = id;
        this.configPath = configPath;
        this.vitestWorker = vitestWorker;
        this.logger = logger;
        this.workspace = workspace;
        this.depResolver = depResolver;
        this.devFiles = devFiles;
        this.opts = opts;
        this.displayName = 'Vitest';
    }
    static from(options) {
        const id = options.name || 'vitest-tester';
        return (context) => {
            const logger = context.createLogger(id);
            const worker = context.createWorker(id, require.resolve('./vitest.worker'));
            let workspace;
            try {
                workspace = context.getAspect(WorkspaceAspect.id);
            }
            catch (err) {
                logger.info('unable to load workspace aspect');
            }
            const depResolver = context.getAspect(DependencyResolverAspect.id);
            const devFiles = context.getAspect(DevFilesAspect.id);
            return new VitestTester(id, options.config, worker, logger, workspace, depResolver, devFiles, options);
        };
    }
    version() {
        const filePath = require.resolve('vitest/package.json');
        const vitestPackageInfo = readJsonSync(filePath);
        return vitestPackageInfo.version;
    }
    displayConfig() {
        return readFileSync(this.configPath, 'utf8');
    }
    async onTestRunComplete(callback) {
        this.onTestRunCompleteCb = callback;
    }
    async test(context) {
        if (this.opts.coverage !== false && context.release) {
            context.coverage = true;
        }
        const viteConfig = await getConfig(this.configPath, context, this.workspace, this.depResolver, this.opts, false);
        if (context.coverage) {
            for (const component of context.components) {
                const patterns = this.devFiles.getDevPatterns(component);
                // @ts-ignore
                viteConfig.test.coverage.exclude.push(...patterns.map((pattern) => `${pattern}`));
            }
            // @ts-ignore
            viteConfig.test.coverage.exclude.push('**/*.bit-app.*');
        }
        const include = patternsToArray(this.opts, context);
        const exclude = getExclude(this.opts);
        // @ts-ignore
        viteConfig.test.include = include;
        if (exclude) {
            // @ts-ignore
            viteConfig.test.exclude.push(...exclude);
        }
        if (context.updateSnapshot) {
            // @ts-ignore
            viteConfig.test.update = true;
        }
        if (context.watch) {
            // @ts-ignore
            viteConfig.test.watch = true;
        }
        // skip installation checks
        const installErrors = [];
        class CustomPackageInstaller extends VitestPackageInstaller {
            async ensureInstalled(dependency) {
                try {
                    require.resolve(dependency, { paths: [viteConfig.root] });
                }
                catch (err) {
                    installErrors.push(err);
                    return false;
                }
                return true;
            }
        }
        const vitest = await startVitest('test', undefined, undefined, viteConfig, {
            packageInstaller: new CustomPackageInstaller()
        });
        if (context.watch) {
            // hold the process
            await new Promise(() => { });
        }
        await vitest.close();
        const files = vitest.state.getFiles();
        const errors = [
            ...installErrors,
            ...Array.from(vitest.state.errorsSet),
        ];
        const result = getResultFromFiles(files, errors, this.opts, context);
        if (viteConfig.test.coverage) {
            await checkCoverageThresholds(viteConfig.test.coverage, context, result);
        }
        return result;
    }
    async watch(context) {
        const viteConfig = await getConfig(this.configPath, context, this.workspace, this.depResolver, this.opts, true);
        // @ts-ignore
        viteConfig.test.include = patternsToArray(this.opts, context);
        // @ts-ignore
        viteConfig.test.exclude = getExclude(this.opts);
        const workerApi = this.vitestWorker.initiate(context.ui
            ? { stdout: true, stderr: true, stdin: true }
            : { stdout: false, stderr: false, stdin: false });
        const onTestCompleteCb = proxy((files, errors) => {
            if (!this.onTestRunCompleteCb || !files)
                return;
            const tests = getResultFromFiles(files || [], errors || [], this.opts, context);
            this.onTestRunCompleteCb(tests);
        });
        await workerApi.onTestComplete(onTestCompleteCb);
        await workerApi.watch(viteConfig);
        return {};
    }
}
