// @bitdev/react.live-controls@0.0.4/live-controls.tsx
import React from 'react';
import {
  type Controls,
  broadcastReady,
  broadcastDestroy,
  getUpdateListener,
  LiveComposition,
  resolveAll,
  getLiveControlsChannel,
} from '@teambit/compositions.ui.composition-live-controls';

export function useLiveControls(
  values: Record<string, any>,
  controls: Controls,
  noControls: boolean | undefined,
): Record<string, any> {
  const [vs, setValues] = React.useState<Record<string, any>>(() => {
    const [ivs] = resolveAll(values, controls);
    return ivs;
  });

  const cs = React.useMemo(() => {
    const [, ics] = resolveAll(values, controls);
    return ics;
  }, [values, controls]);

  const timestampRef = React.useRef<number>(Date.now());
  const channel = getLiveControlsChannel(window.location);

  React.useEffect(() => {
    if (noControls) return () => {};

    const timestamp = timestampRef.current;

    broadcastReady(window.parent, timestamp, cs, vs, channel);

    return () => {
      broadcastDestroy(window.parent, timestamp, channel);
    };
  }, [noControls, channel]);

  React.useEffect(() => {
    if (noControls) return () => {};

    function onLiveControlsUpdate(event: MessageEvent) {
      getUpdateListener(
        event,
        ({ key, value, channel: payloadChannel = 'default' }) => {
          if (payloadChannel !== channel) return;

          const def = cs.find((c) => c.id === key);
          let newValue: any = value;
          if (def?.input === 'date') {
            newValue = new Date(value);
          }

          setValues((prevValues) => ({
            ...prevValues,
            [key]: newValue,
          }));
        }
      );
    }

    window.addEventListener('message', onLiveControlsUpdate);

    return () => {
      window.removeEventListener('message', onLiveControlsUpdate);
    };
  }, [noControls, channel, cs]);

  return vs;
}

export type LivePreviewProps = LiveComposition<
  React.ComponentType,
  React.ReactNode | any
>;

export function LivePreview({
  component: Comp,
  props,
  controls = [],
  render,
  noControls = false,
}: LivePreviewProps) {
  const values = useLiveControls(props, controls, noControls);
  return render ? render(values, Comp) : <Comp {...values} />;
}

export function SimplePreview({
  component: Comp,
  props,
  controls = [],
  render,
}: LivePreviewProps) {
  const [vs] = resolveAll(props, controls);
  return render ? render(vs, Comp) : <Comp {...vs} />;
}

export type ReactLiveComposition = LivePreviewProps;
