// @bitdev/react.app-types.vite-react@3.0.11/dist/vite-react.js
import { CAPSULE_ARTIFACTS_DIR } from '@teambit/builder';
import { createRequire } from "node:module";
import { join, resolve } from 'node:path';
import { existsSync } from 'node:fs';
import { compact } from 'lodash-es';
import { Port } from '@teambit/toolbox.network.get-port';
import { WorkspaceAspect } from '@teambit/workspace';
import { createSsrServer } from './server.js';
import { loadVite } from './load-vite.js';
const require = createRequire(import.meta.url);
// add `source` into the list particularly for bit components
const DEFAULT_MAIN_FIELDS = ['module', 'source', 'main', 'jsnext:main', 'jsnext'];
const DEFAULT_PEERS = ['react', 'react-dom', 'graphql', 'react-router-dom', '@apollo/client'];
const DEFAULT_CONFIG_FILENAME = 'vite.config';
const DEFAULT_SERVER_CONFIG_FILENAME = 'vite-server.config';
export class ViteReact {
    /**
     * WARNING:
     * PLEASE DO NOT CALL THIS CONSTRUCTOR DIRECTLY.
     * USE `ViteReact.from()` INSTEAD.
     *
     * This constructor may have breaking changes in the future.
     * For better long-term maintenance, it's recommended to call
     * `ViteReact.from(options)` instead.
     */
    constructor(
    /**
     * name of the application.
     */
    name = 'react-vite', 
    /**
     * default port for the dev server and preview.
     */
    portRange = [3000, 3200], 
    /**
     * peer dependencies to alias from the app component dependencies.
     */
    peers = DEFAULT_PEERS, 
    /**
     * vit config. overrides default react config.
     */
    viteConfigPath = '', 
    /**
     * use true to ignite vit in SSR mode.
     */
    ssr = false, 
    /**
     * entries to include.
     */
    serverRoot = 'server.app-root', 
    /**
     * filename of the server.
     */
    serverFilename = 'server', 
    /**
     * build config for the vite build of the server side code.
     * used for both the express server, and the ssr.
     * defaults to cjs.
     */
    viteServerBuildConfigPath = '', 
    /**
     * name of the artifact.
     */
    artifactName = 'app-bundle', 
    /**
     * deploy function.
     */
    deploy, 
    /**
     * instance of vite.
     * uses the any type annotation to avoid type issues between versions. can be either a promise or a value.
     * @type typeof Vite
     */
    vite = loadVite(), 
    /**
     * determine whether to expose the network.
     */
    exposeNetwork = true) {
        this.name = name;
        this.portRange = portRange;
        this.peers = peers;
        this.viteConfigPath = viteConfigPath;
        this.ssr = ssr;
        this.serverRoot = serverRoot;
        this.serverFilename = serverFilename;
        this.viteServerBuildConfigPath = viteServerBuildConfigPath;
        this.artifactName = artifactName;
        this.deploy = deploy;
        this.vite = vite;
        this.exposeNetwork = exposeNetwork;
        this.defaultPeers = DEFAULT_PEERS;
    }
    /**
     * run the dev server for a given context.
     * can run in both pwa or ssr modes.
     */
    async run(context) {
        const vite = await this.vite;
        const workspace = context.getAspect(WorkspaceAspect.id);
        const [fromPort, toPort] = this.portRange;
        const port = Number.isNaN(context.port)
            ? await Port.getPort(fromPort, toPort)
            : context.port;
        const rootComponentDir = context.hostRootDir;
        const appId = context.appComponent.id;
        const isInWorkspace = workspace ? workspace.hasId(appId) : false;
        const componentDir = isInWorkspace
            ? workspace?.componentDir(context.appComponent.id)
            : rootComponentDir;
        const publicDir = componentDir;
        const mode = 'development';
        const userConfig = await this.loadViteConfig({
            mode,
            command: 'serve'
        }, componentDir, this.viteConfigPath);
        const env = await this.getEnvFile(mode, componentDir, context.envVariables);
        const viteConfig = await this.computeConfig(componentDir, userConfig, workspace, env);
        const devServer = await vite.createServer(viteConfig);
        if (this.ssr) {
            const server = await createSsrServer({
                publicDir,
                indexHtmlPath: join(componentDir, 'index.html'),
                serverPath: join(componentDir, this.serverRoot),
                serverEntryFile: join(componentDir, this.serverRoot),
                dev: true,
                viteServer: devServer,
            });
            server.listen(port);
            // eslint-disable-next-line no-console
            console.log(`🚀 ${this.name} app is listening on port ${port}`);
            return {
                appName: this.name,
                port
            };
        }
        await devServer.listen(port);
        devServer.printUrls();
        return {
            appName: this.name,
            port
        };
    }
    async getEnvFile(mode, rootDir, overrides) {
        const vite = await this.vite;
        const dotenv = vite.loadEnv(mode, rootDir);
        return {
            ...overrides,
            ...dotenv
        };
    }
    findViteConfig(componentDir, configPath, defaultConfigPath) {
        if (configPath) {
            return resolve(componentDir, configPath);
        }
        const defaultPath = defaultConfigPath || DEFAULT_CONFIG_FILENAME;
        const defaultPaths = [
            resolve(componentDir, `${defaultPath}.ts`),
            resolve(componentDir, `${defaultPath}.mjs`),
            resolve(componentDir, `${defaultPath}.js`),
        ];
        return defaultPaths.find((candidatePath) => {
            return existsSync(candidatePath);
        });
    }
    async loadViteConfig(config, componentDir, path, defaultConfigFilename = DEFAULT_CONFIG_FILENAME) {
        const vite = await this.vite;
        const configPath = this.findViteConfig(componentDir, path, defaultConfigFilename);
        if (!configPath)
            return {};
        try {
            const viteConfig = await vite.loadConfigFromFile(config, configPath);
            return viteConfig?.config;
        }
        catch (e) {
            return {};
        }
    }
    async computeConfig(root, config = {}, workspace, envVars = {}) {
        const vite = await this.vite;
        const components = workspace ? await workspace.list() : [];
        const packageList = compact(components.map((c) => workspace?.componentPackageName(c)));
        const ignored = packageList.map((p) => `!**/node_modules/${p}/**`);
        const userHosts = config.server?.allowedHosts;
        const allHosts = userHosts ? config.server?.allowedHosts : true;
        const extraConfig = {
            configFile: false,
            root,
            define: {
                'process.env': JSON.stringify({ ...envVars }),
            },
            cacheDir: join(root, 'node_modules', '.vite'),
            optimizeDeps: { exclude: packageList },
            server: {
                allowedHosts: allHosts,
                host: this.exposeNetwork,
                middlewareMode: this.ssr,
                watch: { ignored },
            },
            resolve: {
                mainFields: DEFAULT_MAIN_FIELDS,
                dedupe: this.peers,
            },
            appType: this.ssr ? 'custom' : undefined,
        };
        return vite.mergeConfig(config, extraConfig);
    }
    async computeBuildClientConfig(root, config = {}) {
        const vite = await this.vite;
        const extraConfig = {
            configFile: false,
            root,
            resolve: {
                mainFields: DEFAULT_MAIN_FIELDS,
                dedupe: this.peers,
            },
            build: {
                emptyOutDir: false,
                outDir: join(root, this.outputDir),
                chunkSizeWarningLimit: 1000
            },
        };
        return vite.mergeConfig(config, extraConfig);
    }
    async computeBuildConfig(root, input, outputFilename, config = {}) {
        const vite = await this.vite;
        const outDir = join(root, this.outputDir);
        const extraConfig = {
            configFile: false,
            root,
            resolve: {
                mainFields: DEFAULT_MAIN_FIELDS,
                dedupe: this.peers,
            },
            build: {
                emptyOutDir: false,
                commonjsOptions: {
                    ignoreDynamicRequires: true,
                },
                rollupOptions: {
                    input,
                    output: {
                        entryFileNames: `${outputFilename}.mjs`,
                        chunkFileNames: `${outputFilename}-chunk.mjs`,
                        format: 'esm',
                    },
                },
                ssr: this.ssr,
                outDir,
            }
        };
        return vite.mergeConfig(config, extraConfig);
    }
    get outputDir() {
        return join(CAPSULE_ARTIFACTS_DIR, this.artifactName);
    }
    /**
     * app build.
     * building three entries, one for the server
     */
    async build(context) {
        const vite = await this.vite;
        const root = context.capsule.path;
        const config = await this.loadViteConfig({
            mode: 'production',
            command: 'build'
        }, root, this.viteConfigPath);
        const serverConfig = await this.loadViteConfig({
            mode: 'production',
            command: 'build'
        }, root, this.viteServerBuildConfigPath, DEFAULT_SERVER_CONFIG_FILENAME);
        // client build
        const viteConfig = await this.computeBuildClientConfig(root, config);
        // (ssr-only) server build
        const viteSsrConfig = this.ssr ? await this.computeBuildConfig(root, join(root, this.serverRoot), 'server-ssr', config) : undefined;
        // (ssr-only) runner build
        const ssrRunner = this.ssr ? await this.computeBuildConfig(root, require.resolve('./server-runner.js'), this.serverFilename, serverConfig) : undefined;
        // avoid to be affected by custom `NODE_ENV` by like Vitest
        const currentNodeEnv = process.env.NODE_ENV;
        process.env.NODE_ENV = 'production';
        await Promise.all([viteConfig, viteSsrConfig, ssrRunner]
            .filter(Boolean)
            .map(async (vConfig) => vite.build(vConfig)));
        process.env.NODE_ENV = currentNodeEnv;
        return {
            artifacts: [
                {
                    name: this.artifactName,
                    globPatterns: [this.outputDir],
                },
            ],
            metadata: {
                publicDir: this.outputDir
            }
        };
    }
    static from(options) {
        return new ViteReact(options.name, options.defaultPort, options.peers, options.viteConfigPath, options.ssr, options.serverRoot, options.serverFilename, options.viteServerBuildConfigPath, options.artifactName, options.deploy, options.vite, options.exposeNetwork);
    }
}
