// @teambit/oxc.linter.oxlint-linter@2.0.4/dist/oxlint-linter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OxlintLinter = void 0;
const path_1 = __importDefault(require("path"));
const lodash_1 = require("lodash");
const oxc_linter_oxlint_node_1 = require("@teambit/oxc.linter.oxlint-node");
const compute_options_1 = require("./compute-options");
class OxlintLinter {
    constructor(id = 'oxlint-linter', logger, options) {
        this.id = id;
        this.logger = logger;
        this.options = options;
        // eslint-disable-next-line react/static-property-placement
        this.displayName = 'OXlint';
        this.oxlintNode = oxc_linter_oxlint_node_1.OxlintNode.create(options.oxlintNodeOptions);
    }
    displayConfig() {
        return JSON.stringify(this.options, null, 2);
    }
    version() {
        return this.oxlintNode.version();
    }
    async lint(context, buildContext) {
        const longProcessLogger = this.logger.createLongProcessLogger('linting components', context.components.length);
        const computedOptions = (0, compute_options_1.computeOxlintNodeOptions)(this.options.oxlintNodeOptions, context);
        this.oxlintNode = oxc_linter_oxlint_node_1.OxlintNode.create(computedOptions);
        // Build a map from the paths oxlint will report (relative from cwd) to their component.
        // oxlint normalizes absolute paths to cwd-relative paths in its JSON output.
        const fileToComponent = new Map();
        const allFiles = [];
        context.componentsDirMap.components.forEach((component) => {
            longProcessLogger.logProgress(`component: ${component.id.toString()}, # of files: ${component.filesystem.files.length}`);
            const files = this.getFilesPaths(component, buildContext);
            files.forEach((file) => {
                allFiles.push(file);
                const key = path_1.default.isAbsolute(file) ? path_1.default.relative(process.cwd(), file) : file;
                fileToComponent.set(key, component);
                fileToComponent.set(file, component); // also store absolute as fallback
            });
        });
        // Single oxlint invocation across all component files
        let lintResults;
        if (allFiles.length === 0) {
            // No files to lint — skip the oxlint invocation. Without paths, oxlint
            // would fall back to scanning via tsconfig and hang on large workspaces.
            lintResults = { json: { diagnostics: [] }, default: '' };
        }
        else {
            try {
                lintResults = await this.oxlintNode.run(allFiles);
            }
            catch (err) {
                this.logger.error(err.message ?? String(err));
                throw err;
            }
        }
        const hasIssues = lintResults.json?.diagnostics?.length > 0;
        const fixHint = hasIssues ? '\nSome issues may be auto-fixable. Run bit lint with the --fix flag to apply automatic fixes.' : '';
        const componentOutput = (lintResults.default ?? '') + fixHint;
        // Distribute diagnostics back to their originating component
        const diagsByComponent = new Map();
        lintResults.json.diagnostics.forEach((diag) => {
            const component = fileToComponent.get(diag.filename);
            if (!component)
                return;
            const compId = component.id.toString();
            const existing = diagsByComponent.get(compId) ?? [];
            existing.push(diag);
            diagsByComponent.set(compId, existing);
        });
        const results = context.componentsDirMap.components.map((component) => {
            const diagnostics = diagsByComponent.get(component.id.toString()) ?? [];
            const jsonResult = { ...lintResults.json, diagnostics };
            return this.computeComponentResultsWithTotals(component, jsonResult, componentOutput);
        });
        const { totalErrorCount, totalFatalErrorCount, totalFixableErrorCount, totalFixableWarningCount, totalWarningCount, totalComponentsWithErrorCount, totalComponentsWithFatalErrorCount, totalComponentsWithFixableErrorCount, totalComponentsWithFixableWarningCount, totalComponentsWithWarningCount, isClean } = this.computeManyComponentsTotals(results);
        return {
            totalErrorCount,
            totalFatalErrorCount,
            totalFixableErrorCount,
            totalFixableWarningCount,
            totalWarningCount,
            totalComponentsWithErrorCount,
            totalComponentsWithFatalErrorCount,
            totalComponentsWithFixableErrorCount,
            totalComponentsWithFixableWarningCount,
            totalComponentsWithWarningCount,
            results,
            isClean,
            errors: [],
        };
    }
    getFilesPaths(component, buildContext) {
        let capsule;
        if (buildContext) {
            capsule = buildContext.capsuleNetwork.graphCapsules.getCapsule(component.id);
        }
        const files = component.filesystem.files.map((file) => {
            if (this.options.extensions && !this.options.extensions?.includes(file.extname))
                return undefined;
            if (!capsule)
                return file.path;
            return path_1.default.join(capsule.path, file.relative);
        });
        return (0, lodash_1.compact)(files);
    }
    computeComponentResultsWithTotals(component, jsonResult, output) {
        const files = {};
        let totalErrorCount = 0;
        const totalFatalErrorCount = 0;
        const totalFixableErrorCount = 0;
        const totalFixableWarningCount = 0;
        let totalWarningCount = 0;
        jsonResult.diagnostics.reduce((acc, result) => {
            totalErrorCount += result.severity === 'error' ? 1 : 0;
            totalWarningCount += result.severity === 'warning' ? 1 : 0;
            if (!files[result.filename]) {
                files[result.filename] = {
                    errorCount: 0,
                    warningCount: 0,
                    messages: [],
                    raw: []
                };
            }
            const file = files[result.filename];
            file.errorCount += result.severity === 'error' ? 1 : 0;
            file.warningCount += result.severity === 'warning' ? 1 : 0;
            file.messages.push({
                severity: result.severity,
                message: result.message,
                suggestions: [result.help],
                // TODO: calculate the line and column based on offset and length
            });
            file.raw.push(JSON.parse(JSON.stringify(result)));
            return acc;
        }, files);
        const filesResults = Object.entries(files).map(([filePath, fileResults]) => {
            return {
                filePath,
                ...fileResults
            };
        });
        const isClean = totalErrorCount === 0 && totalWarningCount === 0 && totalFatalErrorCount === 0;
        return {
            component,
            output,
            totalErrorCount,
            totalFatalErrorCount,
            totalFixableErrorCount,
            totalFixableWarningCount,
            totalWarningCount,
            results: filesResults,
            isClean
        };
    }
    computeManyComponentsTotals(componentsResults) {
        let totalErrorCount = 0;
        let totalFatalErrorCount = 0;
        let totalFixableErrorCount = 0;
        let totalFixableWarningCount = 0;
        let totalWarningCount = 0;
        let totalComponentsWithErrorCount = 0;
        let totalComponentsWithFatalErrorCount = 0;
        let totalComponentsWithFixableErrorCount = 0;
        let totalComponentsWithFixableWarningCount = 0;
        let totalComponentsWithWarningCount = 0;
        let isClean = true;
        componentsResults.forEach((result) => {
            if (result.totalErrorCount) {
                totalErrorCount += result.totalErrorCount;
                totalComponentsWithErrorCount += 1;
                isClean = false;
            }
            // @ts-ignore - missing from the @types/eslint lib
            if (result.totalFatalErrorCount) {
                totalFatalErrorCount += result.totalFatalErrorCount;
                totalComponentsWithFatalErrorCount += 1;
                isClean = false;
            }
            if (result.totalFixableErrorCount) {
                totalFixableErrorCount += result.totalFixableErrorCount;
                totalComponentsWithFixableErrorCount += 1;
            }
            if (result.totalFixableWarningCount) {
                totalFixableWarningCount += result.totalFixableWarningCount;
                totalComponentsWithFixableWarningCount += 1;
            }
            if (result.totalWarningCount) {
                totalWarningCount += result.totalWarningCount;
                totalComponentsWithWarningCount += 1;
                isClean = false;
            }
        });
        return {
            totalErrorCount,
            totalFatalErrorCount,
            totalFixableErrorCount,
            totalFixableWarningCount,
            totalWarningCount,
            componentsResults,
            totalComponentsWithErrorCount,
            totalComponentsWithFatalErrorCount,
            totalComponentsWithFixableErrorCount,
            totalComponentsWithFixableWarningCount,
            totalComponentsWithWarningCount,
            isClean
        };
    }
    static resolveBinPath(binName, executable, specificPath, fallbackDirs) {
        if (specificPath)
            return specificPath;
        let resolvedPath;
        fallbackDirs.find((dir) => {
            try {
                resolvedPath = (0, oxc_linter_oxlint_node_1.resolveBin)(binName, { executable, paths: [dir] });
                return true;
            }
            catch {
                // not found in this dir, try next
                return false;
            }
        });
        return resolvedPath;
    }
    static create(options, { logger }) {
        const name = options.name || 'oxlint-linter';
        const fallbackDirs = options.binDir ? [options.binDir, __dirname] : [__dirname];
        const binPath = OxlintLinter.resolveBinPath('oxlint', 'oxlint', options.binPath || options.oxlintNodeOptions.binPath, fallbackDirs);
        const tsgolintBinPath = OxlintLinter.resolveBinPath('oxlint-tsgolint', 'tsgolint', options.tsgolintBinPath || options.oxlintNodeOptions.tsgolintBinPath, fallbackDirs);
        const oxlintNodeOptions = { ...options.oxlintNodeOptions, binPath, tsgolintBinPath };
        return new OxlintLinter(name, logger, { ...options, oxlintNodeOptions });
    }
    static from(options) {
        // @ts-ignore - Linter.version allows Promise<string> but ServiceHandler.version requires string.
        // Fixed upstream: https://github.com/teambit/bit/pull/10226
        return (context) => {
            const logger = context.createLogger(options.name || 'oxlint-linter');
            return OxlintLinter.create(options, { logger });
        };
    }
}
exports.OxlintLinter = OxlintLinter;
