// @teambit/oxc.linter.oxlint-linter@2.0.4/dist/oxlint-config-writer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OxlintConfigWriter = void 0;
const fs_1 = __importDefault(require("fs"));
const crypto_1 = __importDefault(require("crypto"));
const BIT_GENERATED_OXLINT_CONFIG_COMMENT = '// bit-generated-oxlint-config';
const sha1 = (str) => crypto_1.default.createHash('sha1').update(str).digest('hex');
class OxlintConfigWriter {
    constructor(name, configPath) {
        this.name = name;
        this.configPath = configPath;
        this.id = 'oxlint';
        this.patterns = ['**/.oxlintrc.json'];
    }
    calcConfigFiles() {
        const raw = fs_1.default.readFileSync(this.configPath, 'utf-8');
        // Strip $schema — the relative path in the source config is valid there but
        // would point to the wrong location once the file is copied to node_modules/.cache.
        const parsed = JSON.parse(raw);
        delete parsed.$schema;
        const content = JSON.stringify(parsed, null, 2);
        const hash = sha1(raw);
        return [{ name: `oxlintrc.bit.${hash}.json`, content, hash }];
    }
    generateExtendingFile({ writtenConfigFiles }) {
        const configFile = writtenConfigFiles.find((f) => f.name.includes('oxlintrc.bit.'));
        if (!configFile)
            return undefined;
        const config = { extends: [`{${configFile.name}}`] };
        const content = `${BIT_GENERATED_OXLINT_CONFIG_COMMENT}\n${JSON.stringify(config, null, 2)}`;
        return { content, name: '.oxlintrc.json', extendingTarget: configFile, useAbsPaths: false };
    }
    isBitGenerated(filePath) {
        const content = fs_1.default.readFileSync(filePath, 'utf-8');
        return content.includes(BIT_GENERATED_OXLINT_CONFIG_COMMENT);
    }
    static from(options) {
        const name = options.name || 'OxlintConfigWriter';
        return {
            name,
            handler: () => new OxlintConfigWriter(name, options.configPath),
        };
    }
}
exports.OxlintConfigWriter = OxlintConfigWriter;
