// @teambit/defender.coverage-report@0.1.0/coverage-report.ts
import type { TesterContext } from '@teambit/tester';
import path, { join, relative } from 'node:path';
import fsExtra from 'fs-extra';
import { promises } from 'node:fs';

const { readFile } = promises;

// Define the new types according to the GraphQL schema
export interface CoverageDetails {
  total: number;
  covered: number;
  skipped: number;
  pct: number;
}

export interface CoverageData {
  lines: CoverageDetails;
  functions: CoverageDetails;
  statements: CoverageDetails;
  branches: CoverageDetails;
}

export interface FileCoverage {
  path: string;
  data: CoverageData;
}

export interface CoverageResult {
  files: FileCoverage[];
  total: CoverageData;
}

export async function getCoverage(
  context: TesterContext
): Promise<CoverageResult> {
  const coveragePath = join(
    context.rootPath,
    'coverage',
    'coverage-summary.json'
  );

  if (!fsExtra.existsSync(coveragePath)) {
    return {
      files: [],
      total: {
        lines: { total: 0, covered: 0, skipped: 0, pct: 0 },
        functions: { total: 0, covered: 0, skipped: 0, pct: 0 },
        statements: { total: 0, covered: 0, skipped: 0, pct: 0 },
        branches: { total: 0, covered: 0, skipped: 0, pct: 0 },
      },
    };
  }

  try {
    const coverage = await readFile(coveragePath, 'utf8');
    const parsedCoverage = JSON.parse(coverage) as CoverageData;

    // Transform the parsed coverage to match the new schema
    const files: FileCoverage[] = Object.entries(parsedCoverage).map(
      ([filePath, metrics]) => ({
        path: filePath,
        data: {
          lines: metrics.lines,
          functions: metrics.functions,
          statements: metrics.statements,
          branches: metrics.branches,
        },
      })
    );

    const total = computeCoverageTotal(files.map((file) => file.data));

    return {
      files,
      total,
    };
  } catch (err) {
    console.log('no coverage found');
    return {
      files: [],
      total: {
        lines: { total: 0, covered: 0, skipped: 0, pct: 0 },
        functions: { total: 0, covered: 0, skipped: 0, pct: 0 },
        statements: { total: 0, covered: 0, skipped: 0, pct: 0 },
        branches: { total: 0, covered: 0, skipped: 0, pct: 0 },
      },
    };
  } finally {
    if (fsExtra.existsSync(coveragePath)) {
      await fsExtra.remove(join(context.rootPath, 'coverage'));
    }
  }
}

/**
 * Sums up coverage details for a specified key (lines, functions, etc.)
 * and computes the coverage percentage.
 */
function computeCoverageTotal(files: CoverageData[]): CoverageData {
  const totalLines = computeCoverageDetails(files, 'lines');
  const totalFunctions = computeCoverageDetails(files, 'functions');
  const totalStatements = computeCoverageDetails(files, 'statements');
  const totalBranches = computeCoverageDetails(files, 'branches');

  return {
    lines: totalLines,
    functions: totalFunctions,
    statements: totalStatements,
    branches: totalBranches,
  };
}

function computeCoverageDetails(
  files: CoverageData[],
  key: 'lines' | 'functions' | 'statements' | 'branches'
): CoverageDetails {
  const total = files.reduce((sum, file) => sum + file[key].total, 0);
  const covered = files.reduce((sum, file) => sum + file[key].covered, 0);
  const skipped = files.reduce((sum, file) => sum + file[key].skipped, 0);

  // Avoid division by zero
  const pct = total > 0 ? (covered / total) * 100 : 0;

  return {
    total,
    covered,
    skipped,
    pct: Number.parseInt(pct.toFixed(0)),
  };
}

export function coverageByComponent(
  context: TesterContext,
  coverage: CoverageResult
): { [componentId: string]: CoverageResult } {
  const coverageByComponent: { [componentId: string]: CoverageResult } = {};

  // Group coverage data by component
  for (const component of context.components) {
    const compIdStr = component.id.toString();

    // Determine the correct root directory for each component
    const componentWorkspacePath = context.release
      ? join(
          context.rootPath,
          `${component.id.toObject().scope}_` +
            `${component.id.toObject().name.replace(/\//g, '_')}` +
            `@${component.id.toObject().version}`
        )
      : join(
          context.rootPath,
          component.state._consumer.componentMap.rootDir
        );

    const { mainFile } = component;
    const componentFiles: FileCoverage[] = [];

    // Collect files for this component
    for (const file of coverage.files) {
      if (path.basename(file.path) === mainFile.basename) {
        // If the main file has only 0% coverage in every check, skip it
        if (
          file.data.lines.pct === 0 &&
          file.data.functions.pct === 0 &&
          file.data.statements.pct === 0 &&
          file.data.branches.pct === 0
        ) {
          continue;
        }
      }

      // Only include files that match this component's path
      if (file.path.startsWith(componentWorkspacePath)) {
        const relativePath = relative(componentWorkspacePath, file.path);
        componentFiles.push({
          path: relativePath,
          data: file.data,
        });
      }
    }

    // Compute totals for this component
    const totalCoverage = computeCoverageTotal(
      componentFiles.map((file) => file.data)
    );

    // Store the result for this component
    coverageByComponent[compIdStr] = {
      files: componentFiles,
      total: totalCoverage,
    };
  }

  return coverageByComponent;
}
