// @bitdev/react.generators.react-starters@2.0.1/workspace-common/files/workspace-config.ts
import { WorkspaceContext } from '@teambit/generator';
import {
  getWorkspaceConfigTemplateParsed,
  stringifyWorkspaceConfig,
} from '@teambit/config';
import { assign } from 'comment-json';
import { DEFAULT_SCOPE, FORKED_ENV_NAME } from '../constants';

export async function workspaceConfig(
  { name, defaultScope, aspectComponent }: WorkspaceContext,
  extraConfig?: Record<string, any>,
  additionalEnvs: string[] = [],
  env?: string
) {
  const scope = defaultScope || DEFAULT_SCOPE;
  const envId = env || aspectComponent?.id.toStringWithoutVersion();
  if (!envId) {
    throw new Error('env is required when aspectComponent is not provided');
  }
  /**
   * a ws that was generated by the react-env should use the forked env
   */
  const generatorEnv =
    envId === 'teambit.react/react-env' ? `${scope}/${FORKED_ENV_NAME}` : envId;
  const configParsed = await getWorkspaceConfigTemplateParsed();
  configParsed['teambit.workspace/workspace'].name = name;
  configParsed['teambit.workspace/workspace'].defaultScope = scope;
  configParsed["teambit.workspace/workspace"].resolveEnvsFromRoots = true;
  configParsed['teambit.dependencies/dependency-resolver'].packageManager =
    'teambit.dependencies/pnpm';
  configParsed['teambit.dependencies/dependency-resolver'].rootComponents =
    true;
  configParsed['teambit.dependencies/dependency-resolver'].policy = {
    dependencies: {
      // eslint: '^7.28.0',
    },
  };
  configParsed['teambit.generator/generator'] = {
    envs: [generatorEnv, ...additionalEnvs],
  };

  delete configParsed['teambit.workspace/variants'];

  const configMerged = extraConfig
    ? assign(configParsed, extraConfig)
    : configParsed;

  return stringifyWorkspaceConfig(configMerged);
}
