// @teambit/defender.coverage-report@0.1.0/dist/coverage-report.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCoverage = getCoverage;
exports.coverageByComponent = coverageByComponent;
const node_path_1 = __importStar(require("node:path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const node_fs_1 = require("node:fs");
const { readFile } = node_fs_1.promises;
function getCoverage(context) {
    return __awaiter(this, void 0, void 0, function* () {
        const coveragePath = (0, node_path_1.join)(context.rootPath, 'coverage', 'coverage-summary.json');
        if (!fs_extra_1.default.existsSync(coveragePath)) {
            return {
                files: [],
                total: {
                    lines: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    functions: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    statements: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    branches: { total: 0, covered: 0, skipped: 0, pct: 0 },
                },
            };
        }
        try {
            const coverage = yield readFile(coveragePath, 'utf8');
            const parsedCoverage = JSON.parse(coverage);
            // Transform the parsed coverage to match the new schema
            const files = Object.entries(parsedCoverage).map(([filePath, metrics]) => ({
                path: filePath,
                data: {
                    lines: metrics.lines,
                    functions: metrics.functions,
                    statements: metrics.statements,
                    branches: metrics.branches,
                },
            }));
            const total = computeCoverageTotal(files.map((file) => file.data));
            return {
                files,
                total,
            };
        }
        catch (err) {
            console.log('no coverage found');
            return {
                files: [],
                total: {
                    lines: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    functions: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    statements: { total: 0, covered: 0, skipped: 0, pct: 0 },
                    branches: { total: 0, covered: 0, skipped: 0, pct: 0 },
                },
            };
        }
        finally {
            if (fs_extra_1.default.existsSync(coveragePath)) {
                yield fs_extra_1.default.remove((0, node_path_1.join)(context.rootPath, 'coverage'));
            }
        }
    });
}
/**
 * Sums up coverage details for a specified key (lines, functions, etc.)
 * and computes the coverage percentage.
 */
function computeCoverageTotal(files) {
    const totalLines = computeCoverageDetails(files, 'lines');
    const totalFunctions = computeCoverageDetails(files, 'functions');
    const totalStatements = computeCoverageDetails(files, 'statements');
    const totalBranches = computeCoverageDetails(files, 'branches');
    return {
        lines: totalLines,
        functions: totalFunctions,
        statements: totalStatements,
        branches: totalBranches,
    };
}
function computeCoverageDetails(files, key) {
    const total = files.reduce((sum, file) => sum + file[key].total, 0);
    const covered = files.reduce((sum, file) => sum + file[key].covered, 0);
    const skipped = files.reduce((sum, file) => sum + file[key].skipped, 0);
    // Avoid division by zero
    const pct = total > 0 ? (covered / total) * 100 : 0;
    return {
        total,
        covered,
        skipped,
        pct: Number.parseInt(pct.toFixed(0)),
    };
}
function coverageByComponent(context, coverage) {
    const coverageByComponent = {};
    // Group coverage data by component
    for (const component of context.components) {
        const compIdStr = component.id.toString();
        // Determine the correct root directory for each component
        const componentWorkspacePath = context.release
            ? (0, node_path_1.join)(context.rootPath, `${component.id.toObject().scope}_` +
                `${component.id.toObject().name.replace(/\//g, '_')}` +
                `@${component.id.toObject().version}`)
            : (0, node_path_1.join)(context.rootPath, component.state._consumer.componentMap.rootDir);
        const { mainFile } = component;
        const componentFiles = [];
        // Collect files for this component
        for (const file of coverage.files) {
            if (node_path_1.default.basename(file.path) === mainFile.basename) {
                // If the main file has only 0% coverage in every check, skip it
                if (file.data.lines.pct === 0 &&
                    file.data.functions.pct === 0 &&
                    file.data.statements.pct === 0 &&
                    file.data.branches.pct === 0) {
                    continue;
                }
            }
            // Only include files that match this component's path
            if (file.path.startsWith(componentWorkspacePath)) {
                const relativePath = (0, node_path_1.relative)(componentWorkspacePath, file.path);
                componentFiles.push({
                    path: relativePath,
                    data: file.data,
                });
            }
        }
        // Compute totals for this component
        const totalCoverage = computeCoverageTotal(componentFiles.map((file) => file.data));
        // Store the result for this component
        coverageByComponent[compIdStr] = {
            files: componentFiles,
            total: totalCoverage,
        };
    }
    return coverageByComponent;
}
