// @bitdev/react.generators.react-templates@6.0.5/dist/theme-template/files/dark-tokens-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.darkTokensFile = darkTokensFile;
function darkTokensFile(context) {
    return {
        relativePath: `dark-tokens.ts`,
        content: `import { DeepPartial } from '@bitdesign/sparks.sparks-theme';
import { ${context.namePascalCase}Schema } from "./${context.name}-tokens.js";

/**
 * override tokens for the dark theme.
 * overrides the default light theme tokens.
 */
export const darkThemeSchema: DeepPartial<${context.namePascalCase}Schema> = {
  colors: {
    primary: {
      default: '#90caf9',
      hover: '#64b5f6',
      active: '#42a5f5'
    },
    secondary: {
      default: '#ededed',
      hover: '#e0e0e0',
      active: '#d5d5d5'
    },
    surface: {
      background: '#121212',
      primary: '#1e1e1e',
      secondary: '#282828',
    },
    text: {
      primary: '#e0e0e0',
      default: '#e0e0e0',
      secondary: '#9e9e9e',
      inverse: '#212121',
    },
    status: {
      positive: { default: '#28a745', subtle: '#4CAF50' },
      negative: { default: '#dc3545', subtle: '#f44336' },
      warning: { default: '#ffd069', subtle: '#ffc107' },
      info: { default: '#bbdefb', subtle: '#90caf9' },
    },
    overlay: 'rgba(0, 0, 0, 0.8)',
  },
  borders: {
    default: {
      color: '#e0e0e0',
      width: '1px',
      style: 'solid',
    },
    focus: {
      color: '#6750A4',
      width: '2px',
      style: 'solid',
      offset: '2px',
    },
    radius: {
      small: '4px',
      medium: '8px',
      large: '16px',
    },
  }
};
`
    };
}
