// @bitdev/react.generators.react-templates@6.0.5/dist/theme-template/files/theme-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.themeFile = themeFile;
function themeFile(context) {
    return {
        relativePath: `${context.name}.tsx`,
        content: `import { ReactNode, useCallback, useState } from 'react';
import classNames from 'classnames';
import { mergeTokenSchema } from '@bitdesign/sparks.sparks-theme';
import { ${context.namePascalCase}Provider } from './${context.name}-provider.js';
import { ${context.namePascalCase}Schema } from './${context.name}-tokens.js';
import { ThemeContext, ThemeContextValue, ThemeMode } from './theme-controller.js';
import { themeOptions } from './theme-options.js';
import styles from './${context.name}.module.css';

export type ${context.namePascalCase}Props = {
  /**
   * a root ReactNode for the component tree
   * applied with the theme.
   */
  children?: ReactNode;

  /**
   * inject a class name to override to the theme.
   * this allows people to affect your theme. remove to avoid.
   */
  className?: string;

  /**
   * override tokens in the schema
   */
  overrides?: Partial<${context.namePascalCase}Schema>,

  /**
   * preset of the theme.
   */
  initialTheme?: ThemeMode;

  /**
   * style tags to include.
   */
  style?: React.CSSProperties,
};

/**
 * a theme for the ${context.namePascalCase} organization.
 * it provides tokens, fonts and general styling for all components.
 */
export function ${context.namePascalCase}({ children, initialTheme, overrides, className, style, ...rest }: ${context.namePascalCase}Props) {
  const [themeMode, setThemeModeState] = useState<ThemeMode>(initialTheme);
  const themePreset = themeOptions[themeMode];

  const setThemeMode = useCallback((mode: ThemeMode) => {
    setThemeModeState(mode);
  }, []);

  const toggleTheme = useCallback(() => {
    setThemeModeState(prevMode => prevMode === 'light' ? 'dark' : 'light');
  }, []);

  const themeContextValue: ThemeContextValue = {
    themeMode,
    toggleTheme,
    setThemeMode,
  };

  const themeOverrides = mergeTokenSchema(themePreset, overrides);

  return (
    <ThemeContext.Provider value={themeContextValue}>
      <${context.namePascalCase}Provider.ThemeProvider
        className={classNames(styles.${context.nameCamelCase}, className)}
        overrides={themeOverrides}
        {...rest}
      >
        {children}
      </${context.namePascalCase}Provider.ThemeProvider>
    </ThemeContext.Provider>
  );
}
`
    };
}
