// @teambit/react.babel.bit-react-transformer@1.0.48/dist/bit-react-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createBitReactTransformer = createBitReactTransformer;
function _fsExtra() {
  const data = require("fs-extra");
  _fsExtra = function () {
    return data;
  };
  return data;
}
function _lodash() {
  const data = require("lodash");
  _lodash = function () {
    return data;
  };
  return data;
}
function _reactUiHighlighterComponentMetadata() {
  const data = require("@teambit/react.ui.highlighter.component-metadata.bit-component-meta");
  _reactUiHighlighterComponentMetadata = function () {
    return data;
  };
  return data;
}
function _metaFromPkgJson() {
  const data = require("./meta-from-pkg-json");
  _metaFromPkgJson = function () {
    return data;
  };
  return data;
}
function _helpers() {
  const data = require("./helpers");
  _helpers = function () {
    return data;
  };
  return data;
}
// @babel/types, not @types/babel!

const PLUGIN_NAME = 'bit-react-transformer';
function metaToDeclaration(meta, types) {
  const properties = [
  // e.g. "id": "teambit.base-ui/input/button@0.6.10"
  types.objectProperty(types.identifier(_reactUiHighlighterComponentMetadata().componentMetaProperties.componentId), types.stringLiteral(meta.id)),
  // e.g. "homepage": "https://bit.dev/teambit/base-ui/input/button"
  meta.homepage && types.objectProperty(types.identifier(_reactUiHighlighterComponentMetadata().componentMetaProperties.homepageUrl), types.stringLiteral(meta.homepage)),
  // "exported": true / false
  meta.exported && types.objectProperty(types.identifier(_reactUiHighlighterComponentMetadata().componentMetaProperties.isExported), types.booleanLiteral(meta.exported))].filter(x => x);

  // variable deceleration, e.g. `var __bit_component = { ... };`
  const deceleration = types.variableDeclaration('var', [types.variableDeclarator(types.identifier(_reactUiHighlighterComponentMetadata().componentMetaField), types.objectExpression(properties))]);
  return deceleration;
}

/**
 * the bit babel transformer adds a `componentId` property on React components
 * for showcase and debugging purposes.
 */
function createBitReactTransformer(api, opts = {}) {
  let componentMap;
  const types = api.types;
  function setMap(mapPath) {
    try {
      const json = (0, _fsExtra().readFileSync)(mapPath, 'utf-8');
      componentMap = JSON.parse(json);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('bit-react-transformer: error reading map file ', e);
    }
  }
  const extractMeta = (0, _lodash().memoize)(filePath => componentMap?.[filePath] || (0, _metaFromPkgJson().metaFromPackageJson)(filePath));
  function addComponentId(path, filePath, identifier) {
    // add meta property, e.g. `Button.__bit_component = __bit_component;`
    const componentIdStaticProp = types.expressionStatement(types.assignmentExpression('=', types.memberExpression(types.identifier(identifier), types.identifier(_reactUiHighlighterComponentMetadata().componentMetaField)), types.identifier(_reactUiHighlighterComponentMetadata().componentMetaField)));

    // always append *after the nearest statement*, never inside an expression
    const parentStatement = path.getStatementParent() ?? path;
    // Note: Type assertion needed due to @babel/types version incompatibility
    parentStatement.insertAfter(componentIdStaticProp);
  }
  const visitor = {
    // visits the start of the file, right after `"use strict"`
    Program(path, state) {
      // // Do not use .stop() or .skip(), it will stop all other babel plugins as well.
      const filename = state.file.opts.filename;
      if (!filename) return;
      const meta = extractMeta(filename);
      if (!meta) return;
      const deceleration = metaToDeclaration(meta, types);

      // inserts to the top of file
      // Note: Type assertion needed due to @babel/types version incompatibility
      path.unshiftContainer('body', deceleration);
    },
    FunctionDeclaration(path, state) {
      // if (!isFunctionComponent(path.node.body)) return;
      const name = path.node.id?.name;
      const filename = state.file.opts.filename;
      if (!name || !filename || !extractMeta(filename)) return;
      addComponentId(path, filename, name);
    },
    VariableDeclarator(path, state) {
      const filename = state.file.opts.filename;
      if (!filename || !extractMeta(filename)) return;
      const node = path.node;
      if (!node.init) return;
      if (node.id.type !== 'Identifier') return;
      const id = node.id;
      switch (node.init.type) {
        case 'FunctionExpression':
          if ((0, _helpers().isFunctionComponent)(node.init.body)) {
            addComponentId(path.parentPath, filename, id.name);
          }
          break;
        case 'ArrowFunctionExpression':
          addComponentId(path.parentPath, filename, id.name);
          break;

        // handle forwardRef
        case 'CallExpression':
          // direct forwardRef, e.g `const Comp = forwardRef(() => <div>comp</div>);`
          if (node.init.callee.type === 'Identifier' && node.init.callee.name === 'forwardRef') addComponentId(path.parentPath, filename, id.name);

          // react.forwardRef, e.g `const Comp = React.forwardRef(() => <div>comp</div>);`
          if (node.init.callee.type === 'MemberExpression' && node.init.callee.property.type === 'Identifier' && node.init.callee.property.name === 'forwardRef') {
            addComponentId(path.parentPath, filename, id.name);
          }
          break;
        default:
          break;
      }
    },
    ClassDeclaration(path, state) {
      const filename = state.file.opts.filename;
      if (!filename || !extractMeta(filename)) return;
      if (!(0, _helpers().isClassComponent)(path.node)) return;
      const name = path.node.id?.name;
      if (!name) return;
      addComponentId(path, filename, name);
    }
  };
  const Plugin = {
    name: PLUGIN_NAME,
    visitor,
    pre() {
      const filepath = opts.componentFilesPath;
      if (filepath && !componentMap) setMap(filepath);
    },
    post() {
      // reset memoization, in case any file changes between runs
      extractMeta.cache.clear?.();
    }
  };
  return Plugin;
}
