// @teambit/styling.build-tasks.css-modules-transform@0.0.4/dist/css-modules-transform.js
import postcss from 'postcss';
import postcssModules from 'postcss-modules';
import { createHash } from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';
/**
 * Escapes special regex characters in a string so it can be safely used
 * inside a `new RegExp(...)` constructor as a literal match.
 *
 * This is needed because CSS module filenames (e.g. "button.module.css")
 * contain dots which are regex wildcards. Without escaping, "button.module.css"
 * would also match "buttonXmoduleYcss" or similar unintended patterns.
 */
function escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
/**
 * Recursively finds files with a given extension in a directory.
 */
export function findFilesByExtension(dir, ext) {
    const results = [];
    const walk = (currentDir) => {
        const entries = fs.readdirSync(currentDir, { withFileTypes: true });
        entries.forEach((entry) => {
            const fullPath = path.join(currentDir, entry.name);
            if (entry.isDirectory() && entry.name !== 'node_modules') {
                walk(fullPath);
            }
            else if (entry.name.endsWith(ext)) {
                results.push(fullPath);
            }
        });
    };
    walk(dir);
    return results;
}
/**
 * Transforms JS files to replace CSS module imports with plain CSS imports
 * and an inlined class name mapping object.
 */
export function transformJsFiles(distDir, originalCssPath, classNameMap) {
    const jsFiles = findFilesByExtension(distDir, '.js');
    const moduleCssBasename = path.basename(originalCssPath); // e.g. "checkbox.module.css"
    const plainCssBasename = moduleCssBasename.replace('.module.css', '.css');
    const escaped = escapeRegex(moduleCssBasename);
    // ESM default import: import styles from './checkbox.module.css';
    const esmDefaultRe = new RegExp(`import\\s+(\\w+)\\s+from\\s+['"]([./\\\\\\w-]*${escaped})['"]\\s*;?`, 'g');
    // ESM namespace import: import * as styles from './checkbox.module.css';
    const esmStarRe = new RegExp(`import\\s+\\*\\s+as\\s+(\\w+)\\s+from\\s+['"]([./\\\\\\w-]*${escaped})['"]\\s*;?`, 'g');
    // CJS require: const/var/let styles = require('./checkbox.module.css');
    const cjsRe = new RegExp(`(?:const|var|let)\\s+(\\w+)\\s*=\\s*(?:__importDefault\\()?require\\(['"]([./\\\\\\w-]*${escaped})['"]\\)\\)?\\s*;?`, 'g');
    // CJS interop: var checkbox_module_css_1 = require("./checkbox.module.css");
    const cjsInteropRe = new RegExp(`(?:const|var|let)\\s+(\\w+)\\s*=\\s*(?:__importDefault\\(\\s*)?require\\(['"]([./\\\\\\w-]*${escaped})['"]\\)\\s*\\)?\\s*;?`, 'g');
    const jsonMap = JSON.stringify(classNameMap);
    const modifiedFiles = [];
    jsFiles.forEach((jsFile) => {
        let content = fs.readFileSync(jsFile, 'utf8');
        let modified = false;
        content = content.replace(esmDefaultRe, (_, varName, importPath) => {
            modified = true;
            const newPath = importPath.replace(moduleCssBasename, plainCssBasename);
            return `import '${newPath}';\nconst ${varName} = ${jsonMap};`;
        });
        content = content.replace(esmStarRe, (_, varName, importPath) => {
            modified = true;
            const newPath = importPath.replace(moduleCssBasename, plainCssBasename);
            return `import '${newPath}';\nconst ${varName} = ${jsonMap};`;
        });
        // Only apply CJS transforms if not already modified (avoid double-transform)
        if (!modified) {
            content = content.replace(cjsRe, (_, varName, importPath) => {
                modified = true;
                const newPath = importPath.replace(moduleCssBasename, plainCssBasename);
                return `require('${newPath}');\nconst ${varName} = ${jsonMap};`;
            });
            if (!modified) {
                content = content.replace(cjsInteropRe, (_, varName, importPath) => {
                    modified = true;
                    const newPath = importPath.replace(moduleCssBasename, plainCssBasename);
                    return `require('${newPath}');\nconst ${varName} = ${jsonMap};`;
                });
            }
        }
        if (modified) {
            fs.writeFileSync(jsFile, content, 'utf8');
            modifiedFiles.push(jsFile);
        }
    });
    return modifiedFiles;
}
async function processOneCssFile(cssFile, distDir, componentId) {
    const cssContent = fs.readFileSync(cssFile, 'utf8');
    let classNameMap = {};
    const result = await postcss([
        postcssModules({
            generateScopedName: (name, filename) => {
                const fileBaseName = path.basename(filename, '.module.css');
                const input = `${componentId}_${fileBaseName}_${name}_${cssContent}`;
                const hash = createHash('md5')
                    .update(input)
                    .digest('base64url')
                    .substring(0, 5);
                return `${fileBaseName}_${name}_${hash}`;
            },
            getJSON: (_, json) => {
                classNameMap = json;
            },
        }),
    ]).process(cssContent, { from: cssFile });
    // Write plain CSS (rename .module.css -> .css)
    const plainCssPath = cssFile.replace('.module.css', '.css');
    fs.writeFileSync(plainCssPath, result.css, 'utf8');
    // Transform all JS files that import this CSS module
    const modifiedJsFiles = transformJsFiles(distDir, cssFile, classNameMap);
    return [plainCssPath, ...modifiedJsFiles];
}
async function processCapsule(distDir, componentId) {
    const moduleCssFiles = findFilesByExtension(distDir, '.module.css');
    if (moduleCssFiles.length === 0)
        return [];
    // Process CSS files sequentially since they may share the same JS files
    const allFiles = [];
    await moduleCssFiles.reduce((prev, cssFile) => prev.then(async () => {
        const files = await processOneCssFile(cssFile, distDir, componentId);
        allFiles.push(...files);
    }), Promise.resolve());
    return allFiles;
}
export class CssModulesTransformTask {
    constructor() {
        this.aspectId = '';
        this.name = 'CssModulesTransform';
        this.description = 'Transforms CSS Modules into plain CSS with hashed class names so consumers need no special bundler config';
    }
    async execute(context) {
        const capsules = context.capsuleNetwork.originalSeedersCapsules;
        const artifactPaths = new Set();
        const componentsResults = await Promise.all(capsules.map(async (capsule) => {
            const capsuleDir = capsule.path;
            const distDir = path.join(capsuleDir, 'dist');
            if (!fs.existsSync(distDir)) {
                return { component: capsule.component };
            }
            try {
                const files = await processCapsule(distDir, capsule.component.id.toStringWithoutVersion());
                for (const file of files) {
                    artifactPaths.add(path.relative(capsuleDir, file));
                }
                return { component: capsule.component };
            }
            catch (err) {
                return { component: capsule.component, errors: [err] };
            }
        }));
        return {
            componentsResults,
            artifacts: [
                {
                    generatedBy: this.aspectId,
                    name: this.name,
                    globPatterns: [...artifactPaths],
                },
            ],
        };
    }
    static from(options) {
        const name = options?.name || 'CssModulesTransform';
        return {
            name,
            handler: () => new CssModulesTransformTask(),
        };
    }
}
