// @teambit/webpack.modules.style-regexps@1.0.10/dist/style-regexps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sassRegex = exports.sassNoModuleRegex = exports.sassModuleRegex = exports.lessRegex = exports.lessNoModuleRegex = exports.lessModuleRegex = exports.cssRegex = exports.cssNoModulesRegex = exports.cssModuleRegex = void 0;
// style files regexps

// css
const cssRegex = exports.cssRegex = /\.css$/;
// css regex - will catch .css but not .module.css
const cssNoModulesRegex = exports.cssNoModulesRegex = /(?<!\.module)\.css$/;
const cssModuleRegex = exports.cssModuleRegex = /\.module\.css$/;

// sass | scss
const sassRegex = exports.sassRegex = /\.(scss|sass)$/;
// scss|sass regex - will catch .scss|sass but not .module.scss|sass
const sassNoModuleRegex = exports.sassNoModuleRegex = /(?<!\.module)\.(scss|sass)$/;
const sassModuleRegex = exports.sassModuleRegex = /\.module\.(scss|sass)$/;

// less
const lessRegex = exports.lessRegex = /\.less$/;
// less regex - will catch .less but not .module.less
const lessNoModuleRegex = exports.lessNoModuleRegex = /(?<!\.module)\.less$/;
const lessModuleRegex = exports.lessModuleRegex = /\.module\.less$/;
