// @teambit/preview.react-preview@1.1.8/webpack-error-reporting.transformer.ts
import fs from 'fs';
import path from 'path';
import HtmlWebpackInjectPlugin from 'html-webpack-inject-plugin';
import { WebpackConfigMutator, WebpackConfigTransformer } from '@teambit/webpack';


// define it here instead of importing it from @teambit/ui-foundation.ui.rendering.html to avoid extra dep
const ERROR_EVENT = "_ERROR_OCCURRED_";

/**
 * Creates a webpack transformer that uses html-webpack-inject-plugin to inject
 * the error reporting script into HTML files
 */
export function createErrorReportingTransformer(): WebpackConfigTransformer {
  return (configMutator: WebpackConfigMutator) => {
    const config = configMutator.raw;
    let errorReportingScript: string;
    try {
      const scriptPath = path.resolve(__dirname, './bit-error-reporting.js');
      errorReportingScript = fs.readFileSync(scriptPath, 'utf8');
      errorReportingScript = errorReportingScript.replace(/\$\{ERROR_EVENT\}/g, ERROR_EVENT);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.warn('Failed to load error reporting script:', error);
      errorReportingScript = `
        (function() {
          if (window.parent && window !== window.parent) {
            window.addEventListener('error', function(event) {
              window.parent.postMessage({
                event: '${ERROR_EVENT}',
                payload: {
                  message: event.message || 'Unknown error',
                  filename: event.filename,
                  type: 'runtime-error',
                  context: { url: window.location.href }
                }
              }, '*');
            }, true);
          }
        })();
      `;
    }
    config.plugins = config.plugins || [];
    config.plugins.push(
      new HtmlWebpackInjectPlugin({
        externals: [
          {
            tagName: 'script',
            attributes: {},
            innerHTML: errorReportingScript
          }
        ],
        prepend: true
      })
    );
    return configMutator;
  };
}
