// @teambit/preview.react-preview@1.1.8/webpack-caching.transformer.ts
import { WebpackConfigMutator, WebpackConfigTransformer } from '@teambit/webpack';
import type { Configuration as DevServerConfiguration } from 'webpack-dev-server';

export function createCacheControlTransformer(): WebpackConfigTransformer {
  return (configMutator: WebpackConfigMutator) => {
    const config = configMutator.raw;
    config.devtool = 'eval-cheap-module-source-map';

    if (!config.optimization) {
      config.optimization = {};
    }
    if (!config.devServer) {
      config.devServer = {};
    }
    if (!config.devServer.setupMiddlewares) {
      config.devServer.setupMiddlewares = [];
    }
    config.optimization = {
      ...(config.optimization || {}),
      minimize: false,
      concatenateModules: true,
      moduleIds: 'named',
      runtimeChunk: 'single',
      splitChunks: {
        chunks: 'all',
        maxInitialRequests: 2,
        minSize: 2000000,
        maxSize: 5000000,
        cacheGroups: {
          vendors: {
            test: /[\\/]node_modules[\\/]/,
            name: 'vendors',
            priority: 30,
            enforce: true,
            reuseExistingChunk: true,
          },
          main: {
            test: /\.(js|jsx|ts|tsx)$/,
            name: 'main',
            priority: 10,
            enforce: true,
            chunks: 'all',
          }
        },
      },
    };

    const originalSetupMiddlewares = config.devServer.setupMiddlewares;

    config.devServer.setupMiddlewares = (middlewares: any[], devServer: DevServerConfiguration) => {
      middlewares.unshift({
        name: 'performance-middleware',
        middleware: (req: any, res: any, next: any) => {
          if (req.path.endsWith('.js')) {
            res.setHeader('Cache-Control', 'no-cache');
          }
          if (req.path === '/' || req.path.endsWith('.html')) {
            res.setHeader('Cache-Control', 'public, max-age=1');
          }
          next();
        }
      });

      return typeof originalSetupMiddlewares === 'function'
        ? originalSetupMiddlewares(middlewares, devServer)
        : middlewares;
    };

    return configMutator;
  };
}
