// @teambit/oxc.linter.oxlint-node@2.0.2/options.ts
import { FixesFlags } from './fixes';
import { PluginsFlags } from './plugins';
import { RulesFlags } from './rules';

export type OxlintFormat = 'default' | 'json' | 'unix' | 'checkstyle' | 'github' | 'gitlab' | 'junit' | 'stylish';

export type OxlintDisableDirectivesSeverity = 'error' | 'warn' | 'off';

export type OxlintNodeOptions = {
  /**
   * path to oxlint config file to use during linting
   */
  configPath?: string;

  /**
   * path to tsconfig to use during compilation.
   */
  tsconfigPath?: string;

  /**
   * decide the format for the CLI output.
   */
  formats?: OxlintFormat[];

  /**
   * decide the plugins to enable.
   */
  pluginsFlags?: PluginsFlags;

  /**
   * decide the rules / categories to enable.
   */
  rulesFlags?: RulesFlags;

  /**
   * decide the fixes to enable.
   */
  fixesFlags?: FixesFlags;

  /**
   * path to oxlint binary.
   */
  binPath?: any;

  /**
   * path to tsgolint binary. When set, passed to oxlint via the
   * OXLINT_TSGOLINT_PATH environment variable so it can locate
   * the tsgolint executable in isolated environments (e.g. capsules).
   */
  tsgolintBinPath?: string;

  // File handling

  /**
   * use a custom .eslintignore-style file for ignore patterns.
   */
  ignorePath?: string;

  /**
   * add glob/gitignore patterns to ignore (can be specified multiple times).
   */
  ignorePatterns?: string[];

  /**
   * disable all ignore file processing.
   */
  noIgnore?: boolean;

  // Warning & error control

  /**
   * report only errors; suppress warnings.
   */
  quiet?: boolean;

  /**
   * treat all warnings as errors (non-zero exit code).
   */
  denyWarnings?: boolean;

  /**
   * force a non-zero exit code if the number of warnings exceeds this threshold.
   */
  maxWarnings?: number;

  // Output

  /**
   * suppress all diagnostic output.
   */
  silent?: boolean;

  // Advanced

  /**
   * number of CPU threads to use (1 for single-core).
   */
  threads?: number;

  /**
   * print the active configuration without linting.
   */
  printConfig?: boolean;

  /**
   * flag unused oxlint-disable-line comments.
   */
  reportUnusedDisableDirectives?: boolean;

  /**
   * same as reportUnusedDisableDirectives but with a custom severity level.
   */
  reportUnusedDisableDirectivesSeverity?: OxlintDisableDirectivesSeverity;

  /**
   * skip automatic nested configuration loading.
   */
  disableNestedConfig?: boolean;

  /**
   * enable rules that require type information.
   */
  typeAware?: boolean;

  /**
   * enable experimental TypeScript compiler diagnostics.
   */
  typeCheck?: boolean;
};
