// @teambit/oxc.linter.oxlint-node@2.0.2/oxlint-node.ts
import execa from 'execa';
import { execFileSync } from 'child_process';
import { resolveBin } from './resolve-bin';
import { OxlintFormat, OxlintNodeOptions } from './options';
import { Plugins } from './plugins';
import { Rules } from './rules';
import { Fixes } from './fixes';

export type OxlintMultiFormatResult = Partial<Record<OxlintFormat, any>>;

export class OxlintNode {
  constructor(
    public binPath: string,
    public plugins: Plugins,
    public rules: Rules,
    public fixes: Fixes,
    public formats: OxlintFormat[] = ['default'],
    public configPath?: string,
    public tsconfigPath?: string,
    private options: Omit<OxlintNodeOptions, 'pluginsFlags' | 'rulesFlags' | 'fixesFlags' | 'formats' | 'configPath' | 'tsconfigPath' | 'binPath' | 'tsgolintBinPath'> = {},
    private tsgolintBinPath?: string
  ) {}

  version() {
    return execFileSync(this.binPath, ['--version']).toString().trim();
  }

  private getConfigCliArg(): string {
    return this.configPath ? `--config=${this.configPath}` : '';
  }

  private getTsConfigCliArg(): string {
    return this.tsconfigPath ? `--tsconfig=${this.tsconfigPath}` : '';
  }

  private getFormatCliArg(format: OxlintFormat): string {
    return `--format=${format}`;
  }

  private getExtraCliArgs(): string[] {
    const args: string[] = [];
    const o = this.options;
    if (o.ignorePath) args.push(`--ignore-path=${o.ignorePath}`);
    if (o.ignorePatterns) o.ignorePatterns.forEach((p) => args.push(`--ignore-pattern=${p}`));
    if (o.noIgnore) args.push('--no-ignore');
    if (o.quiet) args.push('--quiet');
    if (o.denyWarnings) args.push('--deny-warnings');
    if (o.maxWarnings !== undefined) args.push(`--max-warnings=${o.maxWarnings}`);
    if (o.silent) args.push('--silent');
    if (o.threads !== undefined) args.push(`--threads=${o.threads}`);
    if (o.printConfig) args.push('--print-config');
    if (o.reportUnusedDisableDirectivesSeverity) {
      args.push(`--report-unused-disable-directives-severity=${o.reportUnusedDisableDirectivesSeverity}`);
    } else if (o.reportUnusedDisableDirectives) {
      args.push('--report-unused-disable-directives');
    }
    if (o.disableNestedConfig) args.push('--disable-nested-config');
    if (o.typeAware) args.push('--type-aware');
    if (o.typeCheck) args.push('--type-check');
    return args;
  }

  async run(paths: string[] = []): Promise<OxlintMultiFormatResult> {
    const result: OxlintMultiFormatResult = {};
    await Promise.all(
      this.formats.map(async (format) => {
        const args = this.toCliArgs(format, paths);
        let stdout: string;
        const env: Record<string, string> = {};
        if (format !== 'json') env.FORCE_COLOR = '1';
        if (this.tsgolintBinPath) env.OXLINT_TSGOLINT_PATH = this.tsgolintBinPath;
        const execaOpts = Object.keys(env).length > 0 ? { env } : {};
        try {
          ({ stdout } = await execa(this.binPath, args, execaOpts));
        } catch (err: any) {
          // oxlint exits with code 1 when it finds lint issues — that is expected.
          // Only re-throw if there is no stdout, which indicates a real failure
          // (e.g. binary not found, bad arguments).
          if (!err.stdout) throw err;
          stdout = err.stdout;
        }
        result[format] = stdout;
        if (format === 'json') {
          try {
            result[format] = JSON.parse(stdout);
          } catch {
            throw new Error(`oxlint returned non-JSON output:\n${stdout}`);
          }
        }
      })
    );
    return result;
  }

  toCliArgs(format: OxlintFormat, paths: string[] = []): string[] {
    const args = [
      ...paths,
      ...this.plugins.toCliArgs(),
      ...this.rules.toCliArgs(),
      ...this.fixes.toCliArgs(),
      ...this.getExtraCliArgs(),
      this.getConfigCliArg(),
      this.getTsConfigCliArg(),
      this.getFormatCliArg(format),
    ];
    return args.filter(Boolean);
  }

  static create(options: OxlintNodeOptions): OxlintNode {
    const plugins = Plugins.from(options.pluginsFlags || {});
    const rules = Rules.from(options.rulesFlags || []);
    const fixes = Fixes.from(options.fixesFlags || {});
    const { pluginsFlags, rulesFlags, fixesFlags, formats, configPath, tsconfigPath, binPath: optBinPath, tsgolintBinPath: optTsgolintBinPath, ...rest } = options;
    const binPath = optBinPath || resolveBin('oxlint', { executable: 'oxlint', paths: [__dirname] });

    let tsgolintBinPath = optTsgolintBinPath;
    if (!tsgolintBinPath && (rest.typeAware || rest.typeCheck)) {
      try {
        tsgolintBinPath = resolveBin('oxlint-tsgolint', { executable: 'tsgolint', paths: [__dirname] });
      } catch {
        // tsgolint not found — oxlint will report its own error if needed
      }
    }

    return new OxlintNode(
      binPath,
      plugins,
      rules,
      fixes,
      formats,
      configPath,
      tsconfigPath,
      rest,
      tsgolintBinPath
    );
  }
}
