// @teambit/node.deps-detectors.detective-es6@0.0.6/detective-es6.ts
/**
 * this file had been forked (and changed since then) from https://github.com/dependents/node-detective-es6
 */

import {
  getDependenciesFromCallExpression,
  getDependenciesFromMemberExpression,
  getSpecifierValueForImportDeclaration,
} from '@teambit/node.deps-detectors.parser-helper';

/**
 * Import Specifier data.
 * For example, `import foo from './bar' `, "foo" is the import-specifier and is default.
 * Conversely, `import { foo } from './bar' `, here, "foo" is non-default.
 */
export type Specifier = {
  isDefault: boolean;
  name: string;
  exported?: boolean;
};
type Node = any;
type DepData = {
  importSpecifiers?: Specifier[];
};
type Dependencies = Record<string, DepData>;

// @ts-ignore
const Walker = require('node-source-walk');

/**
 * Extracts the dependencies of the supplied es6 module
 *
 * @param  {String|Object} src - File's content or AST
 * @return {String[]}
 */
export function detective(src: any): Dependencies {
  const walker = new Walker();

  const dependencies: Dependencies = {};
  const addDependency = (dependency: string) => {
    if (!dependencies[dependency]) {
      dependencies[dependency] = {};
    }
  };
  const addImportSpecifier = (dependency: string, importSpecifier: Specifier) => {
    const dep = dependencies[dependency];
    if (dep.importSpecifiers) {
      dep.importSpecifiers.push(importSpecifier);
    } else {
      dep.importSpecifiers = [importSpecifier];
    }
  };
  const addExportedToImportSpecifier = (name: string) => {
    Object.keys(dependencies).forEach((dependency) => {
      const dep = dependencies[dependency];
      if (!dep.importSpecifiers) return;
      const specifier = dep.importSpecifiers.find((i) => i.name === name);
      if (specifier) specifier.exported = true;
    });
  };

  if (typeof src === 'undefined') {
    throw new Error('src not given');
  }

  if (src === '') {
    return dependencies;
  }

  walker.walk(src, (node: Node) => {
    switch (node.type) {
      case 'ImportDeclaration':
        if (node.source && node.source.value) {
          const dependency = node.source.value;
          addDependency(dependency);
          node.specifiers.forEach((specifier: any) => {
            const specifierValue = getSpecifierValueForImportDeclaration(specifier);
            addImportSpecifier(dependency, specifierValue);
          });
        }
        break;
      case 'ExportNamedDeclaration':
      case 'ExportAllDeclaration':
        if (node.source && node.source.value) {
          const dependency = node.source.value;
          addDependency(dependency);
          if (node.specifiers) {
            // in case of "export * from" there are no node.specifiers
            node.specifiers.forEach((specifier: any) => {
              const specifierValue = {
                isDefault: !specifier.local || specifier.local.name === 'default', // e.g. export { default as isArray } from './is-array';
                name: specifier.exported.name,
                exported: true,
              };
              addImportSpecifier(dependency, specifierValue);
            });
          }
        } else if (node.specifiers && node.specifiers.length) {
          node.specifiers.forEach((exportSpecifier: any) => {
            addExportedToImportSpecifier(exportSpecifier.exported.name);
          });
        }
        break;
      case 'ExportDefaultDeclaration':
        addExportedToImportSpecifier(node.declaration.name);
        break;
      case 'ImportExpression': {
        // node represents Dynamic Imports such as import(source)
        if (node.source?.value) addDependency(node.source?.value);
        break;
      }
      case 'CallExpression':
        {
          const value = getDependenciesFromCallExpression(node);
          if (value) addDependency(value);
        }
        break;
      case 'MemberExpression':
        {
          const value = getDependenciesFromMemberExpression(node);
          if (value) addDependency(value);
        }
        break;
      default:
        break;
    }
  });

  return dependencies;
}
