// @teambit/react.modules.react-live-controls-from-schema@0.0.2/react-live-controls-from-schema.ts
import { dirname, join } from 'path';
import { type Control } from '@teambit/compositions.ui.composition-live-controls';

export type Schema = any;

type Node = any;

type MemberType = { name: string; package?: string, options?: string[] };

type Exports = { [filePath: string]: { [name: string]: Node } };

export type MetaInfo = {
  [key: string]: Control[];
};

export class LiveCompositionSchema {
  constructor(private schema: Schema) {
    this.initAllExports();
    // console.log('Exports:', this.exports);
  }

  private exports: Exports = {};

  static compositionFilePattern = /\.compositions?\.\w+$/;

  analyze(): MetaInfo {
    // 1. get compositions
    // 2. filter out non-react
    // 3. get props type
    // 4. get controls type
    const output: MetaInfo = {};
    const compositions = this.getCompositions();
    const liveCompositions = compositions.filter(this.isLiveComposition);
    liveCompositions.forEach((liveComposition: Node) => {
      const name = this.getCompositionName(liveComposition);
      const componentNode = this.getLiveComponent(liveComposition);
      if (!this.isReactComponent(componentNode)) return;
      const propsTypeNode = this.getPropsType(componentNode);
      if (!propsTypeNode) return;
      const resolvedPropsTypeNode = this.resolvePropsType(propsTypeNode);
      if (!resolvedPropsTypeNode) return;
      const controlsType = this.getControlsType(resolvedPropsTypeNode);
      if (!controlsType) return;
      output[name] = controlsType;
    })
    return output;
  }

  initAllExports(): void {
    this.schema.module.exports.forEach((node: Node) => {
      const { name, exportNode } = node;
      const { filePath } = exportNode.location;
      this.setExport(filePath, name, exportNode);
    });
    this.schema.internals.forEach((moduleNode: Node) => {
      const { filePath } = moduleNode.location;
      moduleNode.exports.forEach((node: Node) => {
        const { name } = node;
        this.setExport(filePath, name, node);
      });
      moduleNode.internals.forEach((node: Node) => {
        if (node.exports) {
          node.exports.forEach((subNode: Node) => {
            const { name } = subNode;
            this.setExport(filePath, name, subNode);
          });
        } else {
          const { name } = node;
          this.setExport(filePath, name, node);
        }
      });
    });
  }

  setExport(filePath: string, name: string, node: Node): void {
    // remove .js/.ts/.tsx/.jsx extension for easier lookup
    const resolvedFilePath = filePath.replace(/\.(js|ts|tsx|jsx)$/, '');
    const resolvedName = name.includes(' (default)') ? 'default' : name;
    if (!this.exports[resolvedFilePath]) {
      this.exports[resolvedFilePath] = {};
    }
    this.exports[resolvedFilePath][resolvedName] = node;
  }

  getExport(filePath: string, name: string): Node | null {
    const resolvedFilePath = filePath.replace(/\.(js|ts|tsx|jsx)$/, '');
    if (this.exports[resolvedFilePath] && this.exports[resolvedFilePath][name]) {
      return this.exports[resolvedFilePath][name];
    }
    return null;
  }

  getCompositions(): Node[] {
    const compositions = this.schema.internals.find(
      (node: Node) => node.location.filePath.match(
        LiveCompositionSchema.compositionFilePattern
      )
    );
    const exports = compositions ? compositions.exports : [];
    return exports.map((node: Node) => {
      if (node.__schema === 'ExportSchema') {
        return node.exportNode;
      }
      return node;
    });
  }

  isLiveComposition(compositionNode: Node): boolean {
    // console.log('isLiveComposition', compositionNode, compositionNode.__schema);
    // 1. it's an object
    // 2. it has a 'component' member
    if (
      compositionNode.__schema === 'VariableLikeSchema' &&
      compositionNode.type?.__schema === 'ObjectLiteralExpressionSchema'
    ) {
      const componentMember = compositionNode.type.members.find((m: Node) => m.name === 'component');
      return !!componentMember;
    }
    return false;
  }

  getCompositionName(compositionNode: Node): string {
    // console.log('getCompositionName', compositionNode.name);
    return compositionNode.name;
  }

  getLiveComponent(compositionNode: Node) {
    // console.log('getLiveComponent', compositionNode.name, compositionNode.type);
    const componentMember = compositionNode.type.members.find((m: Node) => m.name === 'component');
    if (componentMember.value?.__schema === 'TypeRefSchema') {
      const { name, sourceFilePath } = componentMember.value;
      const result = this.getExport(LiveCompositionSchema.resolveSourceFilePath(sourceFilePath, compositionNode.location.filePath), name);
      return result;
    }
    return null;
  }

  isReactComponent(node: Node): boolean {
    const returnType = node?.returnType;
    // console.log('isReactComponent', node, returnType);
    if (returnType?.packageName === 'react') {
      return LiveCompositionSchema.REACT_NODE_NAMES.includes(returnType.name);
    }
    return false;
  }

  getPropsType(reactComponentNode: Node): Node | null {
    // console.log('getPropsType', reactComponentNode);
    const paramType = reactComponentNode?.props?.type || reactComponentNode?.params?.[0]?.type;
    return paramType;
  }

  resolvePropsType(propsType: Node): Node | null {
    // console.log('resolvePropsType', propsType);
    if (propsType.__schema === 'TypeLiteralSchema') {
      const newTypeNode = {
        members: [] as Node[],
        __schema: 'TypeLiteralSchema',
      };
      if (propsType.members) {
        newTypeNode.members.push(...propsType.members);
      }
      return newTypeNode;
    }
    if (propsType.__schema === 'TypeIntersectionSchema') {
      const newTypeNode = {
        members: [] as Node[],
        __schema: 'TypeLiteralSchema',
      };
      propsType.types.forEach((typeNode: Node) => {
        const resolvedTypeNode = this.resolvePropsType(typeNode);
        if (resolvedTypeNode && resolvedTypeNode.members) {
          newTypeNode.members.push(...resolvedTypeNode.members);
        }
      });
      return newTypeNode;
    }
    if (propsType.__schema === 'TypeRefSchema') {
      const result = this.getExport(propsType.location.filePath, propsType.name);
      return result ? this.resolvePropsType(result.type) : null;
    }
    return null;
  }

  getControlsType(propsTypeNode: Node): Control[] {
    // console.log('getControlsType', propsTypeNode);
    const controls: Control[] = propsTypeNode.members.map((member: Node) => {
      // console.log('member', member.name, this.getMemberType(member.type));
      const memberType = this.getMemberType(member.type);
      if (memberType) {
        if (!memberType.package && memberType.name) {
          // internal type
          if (memberType.name === 'select' && memberType.options) {
            return { id: member.name, type: 'select', options: memberType.options };
          }
          return { id: member.name, type: memberType.name };
        }
        if (
          member.name === 'children' &&
          memberType.package === 'react' &&
          memberType.name === 'ReactNode'
        ) {
          // external type (children: React.ReactNode)
          return { id: member.name, type: 'string' };
        }
      }
      return null;
    }).filter(Boolean);
    return controls;
  }

  getMemberType(node: Node): MemberType | null {
    // console.log('getMemberType', node);
    // type refs
    if (node.__schema === 'TypeRefSchema') {
      if (node.packageName) {
        return { name: node.name, package: node.packageName };
      }
      if (node.internalFilePath) {
        const filePath = node.internalFilePath;
        const internalType = this.getExport(filePath, node.name);
        if (!internalType) return null;
        // console.log('  internalType', to, from, node.name, internalType);
        if (internalType.__schema === 'KeywordTypeSchema') {
          return { name: internalType.name };
        }
        return null;
      }
    }
    // literal union
    // e.g. 'small' | 'medium' | "large"
    if (node.__schema === 'TypeUnionSchema') {
      const isStringUnion = node.types.every(
        (t: Node) =>
          t.__schema === 'LiteralTypeSchema' &&
          LiveCompositionSchema.isLiteralString(t.name)
      );
      if (isStringUnion) {
        return {
          name: 'select',
          options: node.types.map(
            (t: Node) => LiveCompositionSchema.getLiteralStringValue(t.name)
          ).filter((x: string | null) => x !== null),
        };
      }
    }
    // keyword type
    // e.g. string, boolean, number, any
    if (node.__schema === 'KeywordTypeSchema') {
      if (['string', 'boolean', 'number'].includes(node.name)) {
        return { name: node.name };
      }
      return null;
    }
    // array type
    if (node.__schema === 'TypeArraySchema') {
      return null;
    }
    // others
    return null;
  }

  static REACT_NODE_NAMES = [
    'ReactNode',
    'JSX.Element',
    'React.ReactNode',
    'React.JSX.Element'
  ];

  static isLiteralString(name: string): boolean {
    // console.log('isLiteralString', name, /^'[^']*'$/.test(name));
    return /^'[^']*'$/.test(name) || /^"[^"]*"$/.test(name);
  }

  static getLiteralStringValue(name: string): string | null {
    try {
      // eslint-disable-next-line no-eval
      return eval(name);
    } catch {
      return null;
    }
  }

  static resolveSourceFilePath(filePath: string, base: string): string {
    const baseDirname = dirname(base);
    const resolvedPath = join(baseDirname, filePath);
    return resolvedPath;
  }
}

export function getLiveControls(schema: Schema): Record<string, Control[]> {
  const analyzer = new LiveCompositionSchema(schema);
  return analyzer.analyze();
}
