// @teambit/node.deps-detectors.detective-es6@0.0.6/dist/detective-es6.spec.js
"use strict";
/* eslint-disable jest/expect-expect, jest/valid-expect */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * this file had been forked (and changed since then) from https://github.com/dependents/node-detective-es6
 */
const chai_1 = require("chai");
const detective_es6_1 = require("./detective-es6");
const assert = require('assert');
describe('detective-es6', () => {
    const ast = {
        type: 'Program',
        body: [
            {
                type: 'VariableDeclaration',
                declarations: [
                    {
                        type: 'VariableDeclarator',
                        id: {
                            type: 'Identifier',
                            name: 'x',
                        },
                        init: {
                            type: 'Literal',
                            value: 4,
                            raw: '4',
                        },
                    },
                ],
                kind: 'let',
            },
        ],
    };
    it('accepts an ast', () => {
        const deps = (0, detective_es6_1.detective)(ast);
        const depsKeys = Object.keys(deps);
        assert(!depsKeys.length);
    });
    it('retrieves the dependencies of es6 modules', () => {
        const deps = (0, detective_es6_1.detective)('import {foo, bar} from "mylib";');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'mylib');
    });
    it('retrieves the re-export dependencies of es6 modules', () => {
        const deps = (0, detective_es6_1.detective)('export {foo, bar} from "mylib";');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'mylib');
    });
    it('retrieves the re-export * dependencies of es6 modules', () => {
        const deps = (0, detective_es6_1.detective)('export * from "mylib";');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'mylib');
    });
    it('handles multiple imports', () => {
        const deps = (0, detective_es6_1.detective)('import {foo, bar} from "mylib";\nimport "mylib2"');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 2);
        assert(depsKeys[0] === 'mylib');
        assert(depsKeys[1] === 'mylib2');
    });
    it('handles default imports', () => {
        const deps = (0, detective_es6_1.detective)('import foo from "foo";');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'foo');
    });
    it('handles dynamic imports', () => {
        const deps = (0, detective_es6_1.detective)('import("foo").then(foo => foo());');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'foo');
    });
    it('should support commonJS syntax', () => {
        const deps = (0, detective_es6_1.detective)('var foo = require("foo");');
        const depsKeys = Object.keys(deps);
        assert(depsKeys.length === 1);
        assert(depsKeys[0] === 'foo');
    });
    it('returns an empty list for empty files', () => {
        const deps = (0, detective_es6_1.detective)('');
        const depsKeys = Object.keys(deps);
        assert.equal(depsKeys.length, 0);
    });
    it('throws when content is not provided', () => {
        assert.throws(() => {
            // @ts-ignore
            (0, detective_es6_1.detective)();
        }, Error, 'src not given');
    });
    it('does not throw with jsx in a module', () => {
        assert.doesNotThrow(() => {
            (0, detective_es6_1.detective)("import foo from 'foo'; var templ = <jsx />;");
        });
    });
    it('does not throw on an async ES7 function', () => {
        assert.doesNotThrow(() => {
            (0, detective_es6_1.detective)("import foo from 'foo'; export default async function bar() {}");
        });
    });
    describe('string in apostrophes', () => {
        it('should recognize when using require statement', () => {
            const deps = (0, detective_es6_1.detective)('const foo = require(`foo`);'); // eslint-disable-line
            const depsKeys = Object.keys(deps);
            assert.equal(depsKeys.length, 1);
            assert.equal(depsKeys[0], 'foo');
        });
        it('should throw when using import syntax', () => {
            (0, chai_1.expect)(() => (0, detective_es6_1.detective)('import foo from `foo`;')).throw(); // eslint-disable-line
        });
    });
    describe('import-specifiers detection (for tree shaking)', () => {
        it('should recognize default imports as default', () => {
            const deps = (0, detective_es6_1.detective)('import foo from "foo";');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).property('importSpecifiers');
            // @ts-ignore
            const importSpecifier = deps.foo.importSpecifiers[0];
            (0, chai_1.expect)(importSpecifier.name).equal('foo');
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            (0, chai_1.expect)(importSpecifier.isDefault).be.true;
        });
        it('should recognize non-default imports as non-default', () => {
            const deps = (0, detective_es6_1.detective)('import { foo } from "foo";');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).property('importSpecifiers');
            // @ts-ignore
            const importSpecifier = deps.foo.importSpecifiers[0];
            (0, chai_1.expect)(importSpecifier.name).equal('foo');
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            (0, chai_1.expect)(importSpecifier.isDefault).be.false;
        });
        it('should support export-default-as syntax', () => {
            const deps = (0, detective_es6_1.detective)('export { default as foo } from "foo";');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).property('importSpecifiers');
            // @ts-ignore
            const importSpecifier = deps.foo.importSpecifiers[0];
            (0, chai_1.expect)(importSpecifier.name).equal('foo');
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            (0, chai_1.expect)(importSpecifier.isDefault).be.true;
        });
        it('should not be supported for CommonJS', () => {
            const deps = (0, detective_es6_1.detective)('const foo = require("foo");');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).not.property('importSpecifiers');
        });
        it('should add "exported": true if the same variable has been imported and exported', () => {
            const deps = (0, detective_es6_1.detective)('import { foo } from "foo"; export default foo;');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).property('importSpecifiers');
            // @ts-ignore
            const importSpecifier = deps.foo.importSpecifiers[0];
            (0, chai_1.expect)(importSpecifier.name).equal('foo');
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            (0, chai_1.expect)(importSpecifier.exported).be.true;
        });
        it('should not add "exported" property if the variable has been imported but not exported', () => {
            const deps = (0, detective_es6_1.detective)('import { foo } from "foo";');
            (0, chai_1.expect)(deps).property('foo');
            // @ts-ignore
            (0, chai_1.expect)(deps.foo).property('importSpecifiers');
            // @ts-ignore
            const importSpecifier = deps.foo.importSpecifiers[0];
            (0, chai_1.expect)(importSpecifier.name).equal('foo');
            (0, chai_1.expect)(importSpecifier).not.property('exported');
        });
    });
});
