// @bitdev/react.react-env@6.0.14/dist/react-env.bit-env.js
import { compact } from 'lodash-es';
import findRoot from 'find-root';
import { CssModulesTransformTask } from '@teambit/styling.build-tasks.css-modules-transform';
import { VitestTask, VitestTester } from '@teambit/vite.vitest-tester';
// import { JestTask, JestTester } from '@teambit/defender.jest-tester';
// import { createReactEsbuildType } from '@bitdev/react.app-types.react-esbuild';
// import { ReactAppType } from '@teambit/react.apps.react-app-types';
// import { ViteReactType } from '@bitdev/react.app-types.vite-react';
// import { ReactSsrType } from '@bitdev/react.app-types.react-ssr';
import { StarterList, TemplateList } from '@teambit/generator';
import { ReactComponentTemplate, ReactEnvTemplate, ReactHookTemplate, ReactJSComponentTemplate, ReactViteTemplate, ReactWebpackTemplate, ThemeTemplate, } from '@bitdev/react.generators.react-templates';
import { ReactWorkspaceStarter } from '@bitdev/react.generators.react-starters';
// import { ReactAppType } from '@teambit/react.apps.react-app-types';
import { TypeScriptExtractor } from '@teambit/typescript';
// import { ReactVitePreview } from '@bitdev/react.preview.react-vite-preview';
// import { ReactPreview as RspackReactPreview } from '@teambit/rspack.dev-services.preview.react-preview';
import { ReactPreview } from '@teambit/preview.react-preview';
import { GLOBAL_TYPES_DIR, resolveTypes, TypescriptCompiler, TypescriptConfigWriter, TypescriptTask, } from '@teambit/typescript.typescript-compiler';
import { OxlintLinter, OxlintTask, OxlintConfigWriter } from '@teambit/oxc.linter.oxlint-linter';
// import { AppTypeList } from '@teambit/application';
import { PrettierConfigWriter, PrettierFormatter, } from '@teambit/defender.prettier-formatter';
import typescript from 'typescript';
import { CAPSULE_ARTIFACTS_DIR, Pipeline } from '@teambit/builder';
import { ConfigWriterList } from '@teambit/workspace-config-files';
import { PackageGenerator } from '@teambit/pkg';
import { createRequire } from 'node:module';
import { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
const esmRequire = createRequire(import.meta.url);
const fileFolderPath = dirname(fileURLToPath(import.meta.url));
export class ReactEnv {
    constructor() {
        /**
         * name of the environment. used for friendly mentions across bit.
         */
        this.name = 'react';
        this.type = 'react';
        /**
         * icon for the env. use this to build a more friendly env.
         * uses react by default.
         */
        this.icon = 'https://static.bit.dev/extensions-icons/react.svg';
        this.envCurrentPath = fileFolderPath;
        this.tsconfigPath = esmRequire.resolve('./config/tsconfig.json');
        this.tsTypesPath = './types';
        this.vitestConfigPath = esmRequire.resolve('./config/vitest.config.mjs');
        this.oxlintConfigPath = esmRequire.resolve('./config/oxlintrc.json');
        this.prettierConfigPath = esmRequire.resolve('./config/prettier.config.cjs');
        this.prettierExtensions = [
            '.js',
            '.jsx',
            '.ts',
            '.tsx',
            '.mjs',
            '.cjs',
            '.json',
            '.css',
            '.scss',
            '.md',
            '.mdx',
            '.html',
            '.yml',
            '.yaml',
        ];
        this.previewMounter = esmRequire.resolve('./preview/mounter.js');
        this.previewDocsTemplate = esmRequire.resolve('./preview/docs.js');
        this.previewViteConfig = esmRequire.resolve('./config/vite.config.mjs');
        this.previewConfig = {
            splitComponentBundle: true,
            strategyName: 'component',
            supportsUseSource: true
        };
        /**
         * determines whether the env in strict mode.
         */
        this.strict = true;
        /**
         * Default npm ignore paths.
         * Will ignore the "artifacts" directory by default.
         */
        this.npmIgnore = [`${CAPSULE_ARTIFACTS_DIR}/`, '.vitest'];
        /**
         * Default package.json modifications.
         */
        this.packageJson = {
            type: 'module',
            main: 'dist/{main}.js',
            types: '{main}.ts',
        };
    }
    /**
     * create an instance of a Bit Component Compiler.
     * Learn more: https://bit.dev/reference/compiling/set-up-compiler
     */
    compiler() {
        return TypescriptCompiler.from({
            esm: true,
            tsconfig: this.tsconfigPath,
            types: this.types,
            typescript,
        });
    }
    /**
     * create an instance of the Bit Tester plugin.
     * learn more: https://bit.dev/reference/testing/set-up-tester
     */
    tester() {
        return VitestTester.from({
            config: this.vitestConfigPath,
        });
        // return JestTester.from({
        //   // jest: esmRequire.resolve('jest'),
        //   config: this.jestConfigPath,
        // });
    }
    /**
     * create an instance for Bit Preview.
     */
    preview() {
        return ReactPreview.from({
            previewConfig: this.previewConfig,
            /**
             * override the default docs template for components.
             */
            docsTemplate: this.previewDocsTemplate,
            /**
             * mounters are used to mount components to DOM
             * during preview. use them for wrapping your components
             * with routing, theming, data fetching or other types
             * of providers.
             */
            mounter: this.previewMounter,
        });
    }
    /**
     * returns an instance of the default TypeScript extractor.
     * used by default for type inference for both JS and TS.
     */
    schemaExtractor() {
        return TypeScriptExtractor.from({
            tsconfig: this.tsconfigPath,
            // @ts-ignore
            includeFiles: ['**/*.composition{s,}.* '],
        });
    }
    /**
     * add a Bit Linter plugin.
     * learn more: https://bit.dev/reference/testing/set-up-tester
     */
    linter() {
        return OxlintLinter.from({
            binDir: this.envCurrentPath,
            oxlintNodeOptions: {
                configPath: this.oxlintConfigPath,
                tsconfigPath: this.tsconfigPath,
                typeAware: true,
            }
        });
    }
    /**
     * create a formatter instance.
     * learn more: https://bit.dev/reference/formatting/set-up-formatter
     */
    formatter() {
        return PrettierFormatter.from({
            configPath: this.prettierConfigPath,
        });
    }
    /**
     * Add your build pipeline.
     * learn more: https://bit.dev/docs/react-env/build-pipelines
     */
    build() {
        return Pipeline.from(compact([
            TypescriptTask.from({
                tsconfig: this.tsconfigPath,
                types: this.types,
                typescript,
            }),
            this.strict ? OxlintTask.from({
                binDir: this.envCurrentPath,
                oxlintNodeOptions: {
                    configPath: this.oxlintConfigPath,
                    tsconfigPath: this.tsconfigPath,
                    typeAware: true,
                }
            }) : undefined,
            VitestTask.from({
                config: this.vitestConfigPath,
            }),
            CssModulesTransformTask.from(),
            // JestTask.from({
            //   config: this.jestConfigPath,
            // }),
        ]));
    }
    /**
     * add build tasks to execute upon [snap](https://bit.dev/docs/snaps).
     * use the snap pipeline for staging and test deployments
     */
    snap() {
        return Pipeline.from([]);
    }
    /**
     * add build tasks to execute upon [tag](https://bit.dev/docs/tags).
     * use the tag pipeline for deployments, or other tasks required for
     * publishing a semantic version for a component.
     */
    tag() {
        return Pipeline.from([]);
    }
    /**
     * add configuration writers to the workspace, to sync IDE features with tools operated by Bit
     */
    workspaceConfig() {
        return ConfigWriterList.from([
            TypescriptConfigWriter.from({
                tsconfig: this.tsconfigPath,
                types: this.types,
            }),
            OxlintConfigWriter.from({
                configPath: this.oxlintConfigPath,
            }),
            PrettierConfigWriter.from({
                configPath: this.prettierConfigPath,
            }),
        ]);
    }
    /**
     * a list of starters for new projects. this helps create a quick and
     * standardized
     */
    starters() {
        return StarterList.from([ReactWorkspaceStarter.from()]);
    }
    /**
     * set a list of component templates to use across your
     * workspaces. new workspaces would be set to include
     * your envs by default.
     */
    generators() {
        return TemplateList.from([
            ReactComponentTemplate.from(),
            ReactViteTemplate.from(),
            ReactHookTemplate.from(),
            ReactWebpackTemplate.from(),
            ReactJSComponentTemplate.from(),
            ReactEnvTemplate.from(),
            ThemeTemplate.from(),
        ]);
    }
    // apps(): EnvHandler<AppTypeList> {
    //   return AppTypeList.from([
    //     ReactSsrType.from(),
    //     ViteReactType.from(),
    //     ReactAppType.from(),
    //   ]);
    // }
    /**
     * configure and control the packaging process of components.
     */
    package() {
        return PackageGenerator.from({
            packageJson: this.packageJson,
            npmIgnore: this.npmIgnore,
        });
    }
    get types() {
        const packagePath = esmRequire.resolve('@teambit/typescript.typescript-compiler');
        const packageRoot = findRoot(packagePath);
        return [
            ...resolveTypes(this.envCurrentPath, [this.tsTypesPath]),
            ...resolveTypes(packageRoot, [GLOBAL_TYPES_DIR]),
        ];
    }
}
export default new ReactEnv();
