// @bitdev/react.app-types.vite-react@3.0.11/vite-react.ts
import {
  AppBuildContext,
  AppBuildResult,
  AppContext,
  Application,
  ApplicationInstance,
  DeployFn,
} from '@teambit/application';
import { CAPSULE_ARTIFACTS_DIR } from '@teambit/builder';
import { createRequire } from "node:module";
import { join, resolve } from 'node:path';
import { existsSync } from 'node:fs';
import { compact } from 'lodash-es';
import { Port } from '@teambit/toolbox.network.get-port';
import { Workspace, WorkspaceAspect } from '@teambit/workspace';
import type { AnymatchPattern, ConfigEnv, InlineConfig } from 'vite';
import { createSsrServer } from './server.js';
import { loadVite } from './load-vite.js';
import type { ViteReactAppOptions } from './vite-react-app-type.js';

const require = createRequire(import.meta.url);

// add `source` into the list particularly for bit components
const DEFAULT_MAIN_FIELDS = ['module', 'source', 'main', 'jsnext:main', 'jsnext'];
const DEFAULT_PEERS = ['react', 'react-dom', 'graphql', 'react-router-dom', '@apollo/client'];

const DEFAULT_CONFIG_FILENAME = 'vite.config';
const DEFAULT_SERVER_CONFIG_FILENAME = 'vite-server.config';

export class ViteReact implements Application {
  /**
   * WARNING:
   * PLEASE DO NOT CALL THIS CONSTRUCTOR DIRECTLY.
   * USE `ViteReact.from()` INSTEAD.
   *
   * This constructor may have breaking changes in the future.
   * For better long-term maintenance, it's recommended to call
   * `ViteReact.from(options)` instead.
   */
  constructor(
    /**
     * name of the application.
     */
    readonly name: string = 'react-vite',

    /**
     * default port for the dev server and preview.
     */
    readonly portRange: [number, number] = [3000, 3200],

    /**
     * peer dependencies to alias from the app component dependencies.
     */
    readonly peers: string[] = DEFAULT_PEERS,

    /**
     * vit config. overrides default react config.
     */
    readonly viteConfigPath: string = '',

    /**
     * use true to ignite vit in SSR mode.
     */
    readonly ssr: boolean = false,

    /**
     * entries to include.
     */
    readonly serverRoot: string = 'server.app-root',

    /**
     * filename of the server.
     */
    readonly serverFilename: string = 'server',

    /**
     * build config for the vite build of the server side code.
     * used for both the express server, and the ssr.
     * defaults to cjs.
     */
    readonly viteServerBuildConfigPath: string = '',

    /**
     * name of the artifact.
     */
    readonly artifactName: string = 'app-bundle',

    /**
     * deploy function.
     */
    readonly deploy?: DeployFn,

    /**
     * instance of vite. 
     * uses the any type annotation to avoid type issues between versions. can be either a promise or a value.
     * @type typeof Vite
     */
    readonly vite: Promise<typeof import('vite')> = loadVite(),
    
    /**
     * determine whether to expose the network.
     */
    readonly exposeNetwork: boolean = true,
  ) {}

  readonly defaultPeers = DEFAULT_PEERS;

  /**
   * run the dev server for a given context.
   * can run in both pwa or ssr modes.
   */
  async run(context: AppContext): Promise<ApplicationInstance> {
    const vite = await this.vite;
    const workspace = context.getAspect<Workspace>(WorkspaceAspect.id);
    const [fromPort, toPort] = this.portRange;
    const port = Number.isNaN(context.port)
      ? await Port.getPort(fromPort, toPort)
      : context.port;

    const rootComponentDir = context.hostRootDir!;
    const appId = context.appComponent.id;
    const isInWorkspace = workspace ? workspace.hasId(appId) : false;
    const componentDir = isInWorkspace 
      ? workspace?.componentDir(context.appComponent.id) as string
      : rootComponentDir;
    const publicDir = componentDir;
    const mode = 'development';
    const userConfig = await this.loadViteConfig({
      mode,
      command: 'serve'
    }, componentDir, this.viteConfigPath);
    const env = await this.getEnvFile(mode, componentDir, context.envVariables);
    const viteConfig = await this.computeConfig(componentDir, userConfig, workspace, env);
    const devServer = await vite.createServer(viteConfig);

    if (this.ssr) {
      const server = await createSsrServer({
        publicDir,
        indexHtmlPath: join(componentDir, 'index.html'),
        serverPath: join(componentDir, this.serverRoot),
        serverEntryFile: join(componentDir, this.serverRoot),
        dev: true,
        viteServer: devServer,
      });

      server.listen(port);
      // eslint-disable-next-line no-console
      console.log(`🚀 ${this.name} app is listening on port ${port}`)

      return {
        appName: this.name,
        port
      };
    }

    await devServer.listen(port);
    devServer.printUrls();

    return {
      appName: this.name,
      port
    };
  }

  async getEnvFile(mode: string, rootDir: string, overrides?: Record<string, string>) {
    const vite = await this.vite;
    const dotenv = vite.loadEnv(mode, rootDir);
    return {
      ...overrides,
      ...dotenv 
    }
  }

  private findViteConfig(componentDir: string, configPath?: string, defaultConfigPath?: string): string | undefined {
    if (configPath) {
      return resolve(componentDir, configPath);
    }
    const defaultPath = defaultConfigPath || DEFAULT_CONFIG_FILENAME;
    const defaultPaths = [
      resolve(componentDir, `${defaultPath}.ts`),
      resolve(componentDir, `${defaultPath}.mjs`),
      resolve(componentDir, `${defaultPath}.js`),
    ];
    return defaultPaths.find((candidatePath) => {
      return existsSync(candidatePath);
    });
  }

  async loadViteConfig(
    config: ConfigEnv,
    componentDir: string,
    path: string,
    defaultConfigFilename: string = DEFAULT_CONFIG_FILENAME
  ): Promise<InlineConfig> {
    const vite = await this.vite;
    const configPath = this.findViteConfig(componentDir, path, defaultConfigFilename);
    if (!configPath) return {};
    try {
      const viteConfig = await vite.loadConfigFromFile(config, configPath);
      return viteConfig?.config;
    } catch (e) {
      return {};
    }
  }

  private async computeConfig(
    root: string,
    config: InlineConfig = {},
    workspace?: Workspace,
    envVars: Record<string, string> = {}
  ): Promise<InlineConfig> {
    const vite = await this.vite;
    const components = workspace ? await workspace.list() : [];
    const packageList = compact(
      components.map((c) => workspace?.componentPackageName(c))
    );
    const ignored: AnymatchPattern[] = packageList.map(
      (p) => `!**/node_modules/${p}/**`
    );
    const userHosts = config.server?.allowedHosts;
    const allHosts = userHosts ? config.server?.allowedHosts : true;
    const extraConfig: InlineConfig = {
      configFile: false,
      root,
      define: {
        'process.env': JSON.stringify({ ...envVars }),
      },
      cacheDir: join(root, 'node_modules', '.vite'),
      optimizeDeps: { exclude: packageList },
      server: {
        allowedHosts: allHosts,
        host: this.exposeNetwork,
        middlewareMode: this.ssr,
        watch: { ignored },
      },
      resolve: {
        mainFields: DEFAULT_MAIN_FIELDS,
        dedupe: this.peers,
      },
      appType: this.ssr ? 'custom' : undefined,
    }

    return vite.mergeConfig(config, extraConfig);
  }

  private async computeBuildClientConfig(root: string, config: InlineConfig = {}): Promise<InlineConfig> {
    const vite = await this.vite;
    const extraConfig: InlineConfig = {
      configFile: false,
      root,
      resolve: {
        mainFields: DEFAULT_MAIN_FIELDS,
        dedupe: this.peers,
      },
      build: {
        emptyOutDir: false,
        outDir: join(root, this.outputDir),
        chunkSizeWarningLimit: 1000
      },
    };

    return vite.mergeConfig(config, extraConfig);
  }

  private async computeBuildConfig(
    root: string,
    input: string,
    outputFilename: string,
    config: InlineConfig = {}
  ): Promise<InlineConfig> {
    const vite = await this.vite;
    const outDir = join(root, this.outputDir);
    const extraConfig: InlineConfig = {
      configFile: false,
      root,
      resolve: {
        mainFields: DEFAULT_MAIN_FIELDS,
        dedupe: this.peers,
      },
      build: {
        emptyOutDir: false,
        commonjsOptions: {
          ignoreDynamicRequires: true,
        },
        rollupOptions: {
          input,
          output: {
            entryFileNames: `${outputFilename}.mjs`,
            chunkFileNames: `${outputFilename}-chunk.mjs`,
            format: 'esm',
          },
        },
        ssr: this.ssr,
        outDir,
      }
    };
    return vite.mergeConfig(config, extraConfig);
  }

  private get outputDir() {
    return join(CAPSULE_ARTIFACTS_DIR, this.artifactName);
  }

  /**
   * app build.
   * building three entries, one for the server
   */
  async build(context: AppBuildContext): Promise<AppBuildResult> {
    const vite = await this.vite;
    const root = context.capsule.path;
    const config = await this.loadViteConfig({
      mode: 'production',
      command: 'build'
    }, root, this.viteConfigPath);

    const serverConfig = await this.loadViteConfig({
      mode: 'production',
      command: 'build'
    }, root, this.viteServerBuildConfigPath, DEFAULT_SERVER_CONFIG_FILENAME);

    // client build
    const viteConfig = await this.computeBuildClientConfig(root, config);
    // (ssr-only) server build
    const viteSsrConfig = this.ssr ? await this.computeBuildConfig(
      root,
      join(root, this.serverRoot),
      'server-ssr',
      config
    ) : undefined;
    // (ssr-only) runner build
    const ssrRunner = this.ssr ? await this.computeBuildConfig(
      root,
      require.resolve('./server-runner.js'),
      this.serverFilename,
      serverConfig
    ) : undefined;

    // avoid to be affected by custom `NODE_ENV` by like Vitest
    const currentNodeEnv = process.env.NODE_ENV;
    process.env.NODE_ENV = 'production';
    await Promise.all(
      [viteConfig, viteSsrConfig, ssrRunner]
        .filter(Boolean)
        .map(async (vConfig) => vite.build(vConfig))
    );
    process.env.NODE_ENV = currentNodeEnv;

    return {
      artifacts: [
        {
          name: this.artifactName,
          globPatterns: [this.outputDir],
        },
      ],
      metadata: {
        publicDir: this.outputDir
      }
    };
  }

  static from(options: ViteReactAppOptions) {
    return new ViteReact(
      options.name,
      options.defaultPort,
      options.peers,
      options.viteConfigPath,
      options.ssr,
      options.serverRoot,
      options.serverFilename,
      options.viteServerBuildConfigPath,
      options.artifactName,
      options.deploy,
      options.vite,
      options.exposeNetwork
    );
  }
}
