// @bitdev/react.app-types.vite-react@3.0.11/vite-react-app-type.ts
import { ApplicationType, Application, DeployFn } from "@teambit/application";
import { Workspace } from '@teambit/workspace';
import { EnvContext } from '@teambit/envs';
import { ViteReact } from "./vite-react.js";

export type ViteReactAppOptions = {
  /**
   * name of the app.
   * defaults to `react-vite`.
   */
  name?: string;

  /**
   * port range to use.
   * defaults to `[3000, 3200]`.
   */
  defaultPort?: [number, number]

  /**
   * packages to ensure a single instance.
   * defaults to: `['react', 'react-dom', 'graphql', 'react-router', 'react-router-dom', '@apollo/client']`.
   */
  peers?: string[],

  /**
   * config to inject to vite.
   * defaults to: `vite.config.mjs`.
   */
  viteConfigPath?: string,

  /**
   * define whether to use ssr.
   */
  ssr?: boolean;

  /**
   * root for the server.
   * defaults to: `server.app-root`
   */
  serverRoot?: string,

  /**
   * name of the server file to output.
   * defaults to: `server.js`
   */
  serverFilename?: string,

  /**
   * build config for the vite build of the server side code.
   * used for both the express server, and the ssr.
   * defaults to: `vite-server.config.js`
   */
  viteServerBuildConfigPath?: string,

  /**
   * name of the artifact.
   * defaults to: `app-bundle`
   */
  artifactName?: string,

  /**
   * deploy function.
   */
  deploy?: DeployFn;

  /**
   * instance of vite. 
   * uses the any type annotation to avoid type issues between versions. can be either a promise or a value.
   * @type typeof Vite
   */
  vite?: any;

  /**
   * determine whether to expose the network.
   * defaults to true.
   */
  exposeNetwork?: boolean;
};

export class ViteReactType implements ApplicationType<ViteReactAppOptions> {
  name = 'react-vite';

  constructor(
    private workspace?: Workspace
  ) {}
  
  /**
   * create an instance of a vite app.
   */
  createApp(options: ViteReactAppOptions): Application {
    return ViteReact.from(options);
  }

  /**
   * create a new definition of the vite app type.
   * use this for to configure this app type on your env.
   */
  static from() {
    return (context: EnvContext) => {
      // TODO: fix properly in bit
      try {
        const workspace: Workspace = context.getAspect('teambit.workspace/workspace');
        return new ViteReactType(workspace);
      } catch (err) {
        return new ViteReactType();
      }
    };
  }
}
