// @teambit/compositions.ui.composition-live-controls@0.0.7/dist/composition-live-controls.js
function resolveControlMap(controls) {
    if (Array.isArray(controls)) {
        return controls;
    }
    if (typeof controls === 'object') {
        return Object.keys(controls).map((key) => ({
            ...controls[key],
            id: key,
        }));
    }
    return [];
}
function humanizeControlId(id) {
    const normalized = id
        .replace(/[_\-.]+/g, ' ')
        .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
        .replace(/([A-Z]+)([A-Z][a-z0-9])/g, '$1 $2')
        .trim();
    if (!normalized)
        return id;
    return normalized
        .split(/\s+/)
        .filter(Boolean)
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}
function resolveControlLabel(control) {
    const explicitLabel = (control.displayName ?? control.label)?.trim();
    if (explicitLabel)
        return explicitLabel;
    return humanizeControlId(control.id);
}
function resolveControlInput(control) {
    const { type } = control;
    let newInput = control.input;
    if (!newInput) {
        if (typeof type === 'string') {
            switch (type) {
                case 'boolean':
                    newInput = 'boolean';
                    break;
                case 'string':
                    newInput = 'text';
                    break;
                case 'number':
                    newInput = 'number';
                    break;
                case 'date':
                    newInput = 'date';
                    break;
                case 'object':
                    newInput = 'json';
                    break;
                default:
                    newInput = 'text';
            }
        }
        else if (typeof type === 'function') {
            if (type === Boolean) {
                newInput = 'boolean';
            }
            else if (type === String) {
                newInput = 'text';
            }
            else if (type === Number) {
                newInput = 'number';
            }
            else if (type === Date) {
                newInput = 'date';
            }
            else if (type === Object) {
                newInput = 'json';
            }
            else {
                newInput = 'text';
            }
        }
        else {
            newInput = 'text';
        }
    }
    return { ...control, input: newInput, label: resolveControlLabel(control) };
}
/**
 * Controls can be an array or a map.
 * This function is designed to resolve controls into array.
 */
export function resolveControls(controls) {
    return resolveControlMap(controls).map(resolveControlInput);
}
function getTypeFromValue(value) {
    if (value === null) {
        return null;
    }
    if (typeof value === 'string') {
        return String;
    }
    if (typeof value === 'number') {
        return Number;
    }
    if (typeof value === 'boolean') {
        return Boolean;
    }
    if (value instanceof Date) {
        return Date;
    }
    if (Array.isArray(value)) {
        return Array;
    }
    if (typeof value === 'object') {
        return Object;
    }
    return String;
}
/**
 * Resolves controls from values.
 * It will return an array of controls based on the type of values in each key.
 */
export function resolveControlsFromValues(values) {
    return Object.keys(values).map((key) => {
        const control = {
            id: key,
            type: getTypeFromValue(values[key]),
        };
        return resolveControlInput(control);
    });
}
/**
 * Applies default values from controls to props.
 */
export function resolveValues(values, controls) {
    const initValue = {};
    controls.forEach((control) => {
        const { id, defaultValue } = control;
        if (id in values) {
            initValue[id] = values[id];
        }
        else {
            initValue[id] = defaultValue;
        }
    });
    return initValue;
}
/**
 * Resolves all the data from given values and controls.
 */
export function resolveAll(values, controls) {
    const hasControls = Array.isArray(controls) ? controls.length > 0 : Boolean(controls && Object.keys(controls).length > 0);
    const cs = hasControls ? resolveControls(controls) : resolveControlsFromValues(values);
    const vs = resolveValues(values, cs);
    return [vs, cs];
}
export const BROADCAST_READY_KEY = 'composition-live-controls:ready';
export const BROADCAST_UPDATE_KEY = 'composition-live-controls:update';
export const BROADCAST_DESTROY_KEY = 'composition-live-controls:destroy';
export function broadcastReady(target, timestamp, controls, values, channel = 'default') {
    target.postMessage({
        type: BROADCAST_READY_KEY,
        payload: JSON.parse(JSON.stringify({
            timestamp,
            controls,
            values,
            channel,
        })),
    }, '*');
}
export function broadcastUpdate(target, id, { key, value }, channel = 'default') {
    target.postMessage({
        type: BROADCAST_UPDATE_KEY,
        payload: JSON.parse(JSON.stringify({
            timestamp: id,
            key,
            value,
            channel,
        })),
    }, '*');
}
export function broadcastDestroy(target, id, channel = 'default') {
    target.postMessage({
        type: BROADCAST_DESTROY_KEY,
        payload: JSON.parse(JSON.stringify({
            timestamp: id,
            channel,
        })),
    }, '*');
}
export function getReadyListener(event, callback) {
    if (!event.data || event.data.type !== BROADCAST_READY_KEY)
        return;
    callback(JSON.parse(JSON.stringify(event.data.payload)));
}
export function getUpdateListener(event, callback) {
    if (!event.data || event.data.type !== BROADCAST_UPDATE_KEY)
        return;
    callback(JSON.parse(JSON.stringify(event.data.payload)));
}
export function getDestroyListener(event, callback) {
    if (!event.data || event.data.type !== BROADCAST_DESTROY_KEY)
        return;
    callback(JSON.parse(JSON.stringify(event.data.payload)));
}
// e.g. `http://localhost:3000/preview/bitdev.react/react-env@xxx/#teambit.vite/examples/foo?preview=compositions&env=bitdev.react/react-env@xxx&name=BasicFoo&fullscreen=true&livecontrols=true`
// e.g. `http://xxx.bit.cloud/api/bitdev.react/react-env@xxx/#teambit.vite/examples/foo?preview=compositions&env=bitdev.react/react-env@xxx&name=BasicFoo&fullscreen=true&livecontrols=true`
export function needLiveControls({ hash }) {
    const params = new URLSearchParams(hash.slice(hash.indexOf('?') + 1));
    return params.has('livecontrols') && params.get('livecontrols') === 'true';
}
export function getLiveControlsChannel(location) {
    const params = new URLSearchParams(location.hash.slice(location.hash.indexOf('?') + 1));
    return params.get('lcchannel') || 'default';
}
