// @teambit/compositions.ui.composition-live-controls@0.0.7/dist/live-controls-registry.js
/**
 * Low-level registry for live controls channel state.
 * Manages channel subscriptions, state storage, and broadcasts.
 * Does not contain any diff-specific logic - that belongs in DiffControlsModel.
 */
export class LiveControlsRegistry {
    constructor() {
        this.channels = new Map();
        this.state = new Map();
        this.windowToChannel = new Map();
        this.DEFAULT_CHANNEL = 'default';
    }
    static getInstance() {
        if (!this._instance) {
            this._instance = new LiveControlsRegistry();
        }
        return this._instance;
    }
    normalizeChannel(channel) {
        return channel || this.DEFAULT_CHANNEL;
    }
    getChannelNames() {
        return Array.from(new Set([...this.channels.keys(), ...this.state.keys()]));
    }
    ensureState(key) {
        if (!this.state.has(key)) {
            this.state.set(key, { defs: [], values: {}, ready: false, timestamps: new Map() });
        }
        return this.state.get(key);
    }
    ensureSubscribers(key) {
        if (!this.channels.has(key)) {
            this.channels.set(key, []);
        }
        return this.channels.get(key);
    }
    setWindowChannel(iframeWindow, channel) {
        const normalized = this.normalizeChannel(channel);
        const existing = this.windowToChannel.get(iframeWindow);
        if (normalized !== this.DEFAULT_CHANNEL || !existing) {
            this.windowToChannel.set(iframeWindow, normalized);
        }
    }
    resolveChannel(channel, iframeWindow) {
        const normalized = this.normalizeChannel(channel);
        if (normalized !== this.DEFAULT_CHANNEL)
            return normalized;
        if (!iframeWindow)
            return normalized;
        return this.windowToChannel.get(iframeWindow) || normalized;
    }
    register(channel, iframeWindow, timestamp) {
        const key = this.normalizeChannel(channel);
        this.setWindowChannel(iframeWindow, key);
        const list = this.ensureSubscribers(key);
        if (!list.find((s) => s.iframeWindow === iframeWindow)) {
            list.push({ iframeWindow, timestamp });
        }
        const st = this.ensureState(key);
        st.timestamps.set(iframeWindow, timestamp);
    }
    unregister(channel, iframeWindow) {
        const key = this.normalizeChannel(channel);
        const list = this.channels.get(key);
        if (list) {
            this.channels.set(key, list.filter((s) => s.iframeWindow !== iframeWindow));
        }
        const st = this.state.get(key);
        if (st) {
            st.timestamps.delete(iframeWindow);
            if (st.timestamps.size === 0)
                st.ready = false;
        }
        const mappedChannel = this.windowToChannel.get(iframeWindow);
        if (!mappedChannel || mappedChannel === key || key === this.DEFAULT_CHANNEL) {
            this.windowToChannel.delete(iframeWindow);
        }
    }
    getSubscribers(channel) {
        return this.channels.get(this.normalizeChannel(channel)) || [];
    }
    get hasAnySubscribers() {
        return [...this.channels.values()].some((list) => list.length > 0);
    }
    registerReadyState(channel, defs, values, timestamp, win) {
        const key = this.normalizeChannel(channel);
        const st = this.ensureState(key);
        st.timestamps.set(win, timestamp);
        st.defs = defs.length ? defs : st.defs;
        st.values = { ...st.values, ...values };
        st.ready = st.timestamps.size > 0;
        this.state.set(key, st);
    }
    updateValue(channel, key, value) {
        const st = this.state.get(this.normalizeChannel(channel));
        if (st)
            st.values[key] = value;
    }
    getState(channel) {
        return this.state.get(this.normalizeChannel(channel));
    }
    broadcastUpdate(channel, key, value) {
        const normalizedChannel = this.normalizeChannel(channel);
        const subs = this.getSubscribers(normalizedChannel);
        subs.forEach((sub) => {
            try {
                sub.iframeWindow.postMessage({
                    type: 'composition-live-controls:update',
                    payload: { channel: normalizedChannel, key, value },
                }, '*');
            }
            catch (err) {
                console.error('broadcast failed', err);
            }
        });
    }
    resetChannel(channel) {
        const key = this.normalizeChannel(channel);
        const st = this.state.get(key);
        if (!st)
            return;
        st.timestamps.clear();
        st.ready = false;
        st.defs = [];
        st.values = {};
    }
    // ---- Multi-channel utilities ----
    /**
     * Merge state from multiple channels into a single result.
     * Used by useLiveControls hook to aggregate multiple channels.
     */
    getMergedState(channels) {
        const defsMap = {};
        const values = {};
        let ready = false;
        channels
            .map((ch) => this.normalizeChannel(ch))
            .forEach((ch) => {
            const s = this.state.get(ch);
            if (!s)
                return;
            s.defs.forEach((d) => {
                defsMap[d.id] = d;
            });
            Object.assign(values, s.values);
            if (s.ready)
                ready = true;
        });
        return { defs: Object.values(defsMap), values, ready };
    }
    /**
     * Broadcast an update to multiple channels.
     */
    broadcastUpdateToChannels(channels, key, value) {
        const uniqueChannels = [...new Set(channels.map((ch) => this.normalizeChannel(ch)))];
        uniqueChannels.forEach((ch) => this.broadcastUpdate(ch, key, value));
    }
    /**
     * Reset multiple channels at once.
     */
    resetTimestamps(channels) {
        const uniqueChannels = [...new Set(channels.map((ch) => this.normalizeChannel(ch)))];
        uniqueChannels.forEach((ch) => this.resetChannel(ch));
    }
    cleanupAll() {
        this.channels.clear();
        this.state.clear();
        this.windowToChannel.clear();
    }
}
