// @teambit/rspack.modules.rspack-config-mutator@1.0.22/dist/rspack-config-mutator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RspackConfigMutator = void 0;
exports.runTransformers = runTransformers;
exports.runAppTransformers = runAppTransformers;
const html_modules_inject_html_element_1 = require("@teambit/html.modules.inject-html-element");
const merge_1 = require("./merge");
class RspackConfigMutator {
    constructor(config) {
        this.config = config;
        // this is an alias for consistency with webpack-config-mutator.
        this.raw = config;
    }
    /**
     * Merge the given configuration with the current configuration.
     * @param config Configuration to merge.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.merge({ mode: "development" });
     * ```
     */
    merge(config) {
        this.config = (0, merge_1.mergeRspackConfig)(this.config, config);
        return this;
    }
    /**
     * Clone the mutator with the current configuration.
     * @returns New mutator instance.
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.clone();
     * ```
     */
    clone() {
        return new RspackConfigMutator((0, merge_1.mergeRspackConfig)({}, this.config));
    }
    /**
     * Set the mode of the configuration.
     * @param mode Mode to set.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.setMode("development");
     * ```
     */
    setMode(mode) {
        this.config.mode = mode;
        return this;
    }
    /**
     * Set the target of the configuration.
     * @param target Target to set.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.setTarget("web");
     * ```
     */
    setTarget(target) {
        this.config.target = target;
        return this;
    }
    /**
     * Add a plugin to the configuration.
     * @param plugin Plugin to add.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.addPlugin(new HtmlWebpackPlugin());
     * ```
     */
    addPlugin(plugin, position) {
        this.config.plugins = this.config.plugins || [];
        if (position === 'prepend') {
            this.config.plugins.unshift(plugin);
        }
        else {
            this.config.plugins.push(plugin);
        }
        return this;
    }
    /**
     * Remove a plugin from the configuration.
     * @param plugin Plugin name to remove.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.removePlugin("HtmlWebpackPlugin");
     * ```
     */
    removePlugin(plugin) {
        this.config.plugins = this.config.plugins || [];
        this.config.plugins = this.config.plugins.filter((p) => {
            if (p) {
                return p.constructor.name !== plugin;
            }
            return false;
        });
        return this;
    }
    /**
     * Add multiple plugins to the configuration.
     * @param plugins Plugins to add.
     * @returns RspackConfigMutator instance
     * @example
     * ```ts
     * const mutator = new RspackConfigMutator({});
     * mutator.addPlugins([new HtmlWebpackPlugin(), new MiniCssExtractPlugin()]);
     * ```
     */
    addPlugins(plugins, position) {
        this.config.plugins = this.config.plugins || [];
        if (position === 'prepend') {
            this.config.plugins.unshift(...plugins);
        }
        else {
            this.config.plugins.push(...plugins);
        }
        return this;
    }
    /**
     * To be used to ignore replace packages with global variable
     * Useful when trying to offload libs to CDN
     * @param externalDeps
     * @returns RspackConfigMutator instance
     */
    addExternals(externalDeps) {
        if (!externalDeps)
            return this;
        let { externals } = this.config;
        if (!externals) {
            externals = externalDeps;
        }
        else if (Array.isArray(externalDeps)) {
            externals = externalDeps.concat(externals);
        }
        else if (Array.isArray(externals) ||
            externalDeps.constructor === Function ||
            externalDeps.constructor === RegExp) {
            externals = [externalDeps].concat(externals);
        }
        else if (externalDeps instanceof Object && externals instanceof Object) {
            // @ts-ignore
            externals = {
                ...externals,
                ...externalDeps,
            };
        }
        this.config.externals = externals;
        return this;
    }
    /**
     * Add aliases
     * @param aliases
     * @returns RspackConfigMutator instance
     */
    addAliases(aliases) {
        if (!this.config.resolve) {
            this.config.resolve = {};
        }
        if (!this.config.resolve.alias) {
            this.config.resolve.alias = {};
        }
        Object.assign(this.config.resolve.alias, aliases);
        return this;
    }
    /**
     * Add a custom element to the html template
     * @param element
     * @returns
     * @example
     * addElementToHtmlTemplate({ parent: 'head', position: 'append', tag: 'script', attributes: { src: 'https://cdn.com/script.js', async: true } });
     * addElementToHtmlTemplate({ parent: 'body', position: 'prepend', tag: 'script', content: 'console.log("hello world")' });
     */
    addElementToHtmlTemplate(element) {
        if (!this.config.plugins || this.config.plugins.length === 0) {
            return this;
        }
        const HtmlWebpackPluginConstructorName = 'HtmlWebpackPlugin';
        const HtmlRspackPluginConstructorName = 'HtmlRspackPlugin';
        const supportedHtmlPlugins = [
            HtmlWebpackPluginConstructorName,
            HtmlRspackPluginConstructorName,
        ];
        const htmlPlugins = this.config.plugins.filter((plugin) => supportedHtmlPlugins.includes(plugin.constructor.name));
        if (!htmlPlugins) {
            return this;
        }
        // iterate over all html plugins and add the scripts to the html
        htmlPlugins.forEach((htmlPlugin) => {
            const pluginConstructorName = htmlPlugin.constructor.name;
            if (pluginConstructorName === HtmlWebpackPluginConstructorName) {
                const htmlWebpackPlugin = htmlPlugin;
                const htmlContent = typeof htmlWebpackPlugin.userOptions?.templateContent === 'function'
                    ? htmlWebpackPlugin?.userOptions?.templateContent({})
                    : htmlWebpackPlugin?.userOptions?.templateContent;
                const newHtmlContent = (0, html_modules_inject_html_element_1.inject)(htmlContent, element);
                htmlWebpackPlugin.userOptions.templateContent = newHtmlContent;
            }
            else if (pluginConstructorName === HtmlRspackPluginConstructorName) {
                const htmlRspackPlugin = htmlPlugin;
                const htmlContent = htmlRspackPlugin?._args?.[0].templateContent ?? '';
                const newHtmlContent = (0, html_modules_inject_html_element_1.inject)(htmlContent, element);
                htmlRspackPlugin._args[0].templateContent = newHtmlContent;
            }
        });
        return this;
    }
    findRule(ruleTest, rules) {
        return (rules || this.config.module?.rules)?.find(((rule) => {
            if (!rule)
                return false;
            if (rule?.oneOf?.length > 0) {
                return !!this.findRule(ruleTest, rule.oneOf);
            }
            if (typeof rule.test === 'string') {
                return rule.test.includes(ruleTest);
            }
            if (typeof rule.test === 'object' && rule.test instanceof RegExp) {
                return rule.test.test(ruleTest);
            }
            return false;
        }));
    }
    addTypescript({ react: enableReact } = { react: false }) {
        const tsRule = '.ts';
        const tsReactRule = '.tsx';
        const tsRuleRegex = enableReact ? /\.tsx?$/ : /\.ts$/;
        const isRuleProvided = !!this.findRule(tsRule) || !!this.findRule(tsReactRule);
        if (isRuleProvided) {
            throw new Error('A custom typescript rule was detected, please remove it in order to use this utility.');
        }
        const isDevelopment = this.config.mode !== 'production';
        this.merge({
            module: {
                rules: [{
                        test: tsRuleRegex,
                        loader: "builtin:swc-loader",
                        options: {
                            sourceMap: true,
                            jsc: {
                                parser: {
                                    syntax: "typescript",
                                    tsx: enableReact,
                                },
                                transform: {
                                    react: {
                                        runtime: "automatic",
                                        development: isDevelopment,
                                    },
                                },
                            },
                            env: {
                                targets: 'Chrome >= 48',
                            },
                        },
                    }],
            }
        });
        return this;
    }
    addCssModules() {
        const isDevelopment = this.config.mode !== 'production';
        this.merge({
            experiments: {
                css: true,
            },
            module: {
                parser: {
                    'css/auto': {
                        namedExports: false,
                    },
                    css: {
                        namedExports: false,
                    },
                    'css/module': {
                        namedExports: false,
                    },
                },
                generator: {
                    'css/auto': {
                        esModule: false,
                        exportsConvention: 'camel-case-only',
                        localIdentName: isDevelopment ? '[uniqueName]-[id]-[local]' : '[hash]',
                    },
                    css: {
                        esModule: false,
                    },
                    'css/module': {
                        esModule: false,
                        exportsConvention: 'camel-case-only',
                        localIdentName: isDevelopment ? '[uniqueName]-[id]-[local]' : '[hash]',
                    },
                },
            }
        });
        return this;
    }
}
exports.RspackConfigMutator = RspackConfigMutator;
function runTransformers(baseConfig, transformers, 
// @FIX: fix context type, maybe with guard rails
// context: RspackConfigAppTransformContext | RspackConfigTransformContext
context) {
    if (!Array.isArray(transformers))
        return baseConfig;
    const mutator = new RspackConfigMutator(baseConfig);
    return transformers.reduce((c, transformer) => {
        return transformer(mutator, context).config;
    }, baseConfig);
}
function runAppTransformers(baseConfig, transformers, context) {
    if (!Array.isArray(transformers))
        return baseConfig;
    const mutator = new RspackConfigMutator(baseConfig);
    return transformers.reduce((c, transformer) => {
        return transformer(mutator, context).config;
    }, baseConfig);
}
