// @teambit/mdx.modules.mdx-v3-options@0.0.3/dist/mdx-options.js
import remarkDirective from 'remark-directive';
import remarkGfm from 'remark-gfm';
import { default as detectFrontmatter } from 'remark-frontmatter';
import recmaPluginInjectisMDXComponent from 'recma-mdx-is-mdx-component';
import rehypeMdxCodeProps from 'rehype-mdx-code-props';
import yaml from 'yaml';
import { visit } from 'unist-util-visit';
import { remove } from 'unist-util-remove';
import detectiveEs6 from '@teambit/node.deps-detectors.detective-es6';
import icons from './icons.js';
/// Admonitions
const AdmonitionsNameList = [
    `important`,
    `tip`,
    `note`,
    `warning`,
    `danger`,
    `info`,
    `success`,
    `secondary`,
    `caution`,
];
const AdmonitionsMap = {
    info: `important`,
    success: `tip`,
    secondary: `note`,
    danger: `warning`,
};
const AdmonitionsIcons = {
    important: icons[0],
    tip: icons[1],
    note: icons[2],
    warning: icons[3],
    caution: icons[4],
};
function adaptAdmonitions() {
    // https://github.com/elviswolcott/remark-admonitions/issues/49
    // div.admonition.admonition-{name}
    //   div.admonition-heading (optional)
    //     span.admonition-icon
    //       svg
    //     h5 <title>
    //   div.admonition-content
    //     <content>
    return function (tree) {
        visit(tree, (node) => {
            if (node.type === 'containerDirective') {
                if (!AdmonitionsNameList.includes(node.name))
                    return;
                node.name = AdmonitionsMap[node.name] || node.name;
                node.data = node.data || {};
                node.data.hName = 'div';
                node.data.hProperties = { className: `admonition admonition-${node.name}` };
                let label = node.children[0];
                if (label && label.data && label.data.directiveLabel) {
                    label.type = 'heading';
                    label.depth = 5;
                    label.data.hProperties = { className: 'admonition-heading' };
                    const icon = {
                        data: {
                            hName: 'span',
                            hProperties: { className: 'admonition-icon' },
                        },
                        children: [AdmonitionsIcons[node.name]],
                    };
                    label.children.unshift(icon);
                    node.children.shift();
                }
                else {
                    label = null;
                }
                node.children = [
                    label,
                    {
                        type: 'div',
                        data: {
                            hName: 'div',
                            hProperties: { className: 'admonition-content' },
                        },
                        children: node.children,
                    },
                ].filter(Boolean);
            }
        });
    };
}
/// Frontmatter
function extractMetadata() {
    return function transformer(tree, file) {
        visit(tree, 'yaml', (node) => {
            try {
                // eslint-disable-next-line no-param-reassign
                file.data.frontmatter = yaml.parse(node.value, { prettyErrors: true });
            }
            catch (err) {
                throw new Error(`failed extracting metadata/front-matter using Yaml lib, due to an error (please disregard the line/column): ${err.message}`);
            }
        });
    };
}
function extractImports() {
    return function transformer(tree, file) {
        visit(tree, 'mdxjsEsm', (node) => {
            var _a;
            const es6Import = detectiveEs6.default(node.value);
            const imports = Object.keys(es6Import).flatMap((dep) => {
                if (!es6Import[dep].importSpecifiers) {
                    return {
                        fromModule: dep,
                    };
                }
                return es6Import[dep].importSpecifiers.map((importSpecifier) => ({
                    fromModule: dep,
                    identifier: importSpecifier.name,
                    isDefault: importSpecifier.isDefault,
                }));
            });
            // eslint-disable-next-line no-param-reassign
            ((_a = file.data).imports || (_a.imports = [])).push(...imports);
        });
        remove(tree, 'yaml');
    };
}
/// Scope Context
function wrapWithScopeContext() {
    return (tree, file) => {
        const imports = file.data?.imports || [];
        const ids = imports.reduce((identifiers, importSpecifier) => {
            const newIds = [];
            if (importSpecifier.identifier)
                newIds.push(importSpecifier.identifier);
            return identifiers.concat(newIds);
        }, []);
        const importNode = {
            "type": "mdxjsEsm",
            "value": "import { MDXScopeProvider } from '@teambit/mdx.ui.mdx-scope-context';",
            "data": {
                "estree": {
                    "type": "Program",
                    "body": [
                        {
                            "type": "ImportDeclaration",
                            "specifiers": [
                                {
                                    "type": "ImportSpecifier",
                                    "imported": {
                                        "type": "Identifier",
                                        "name": "MDXScopeProvider"
                                    },
                                    "local": {
                                        "type": "Identifier",
                                        "name": "MDXScopeProvider"
                                    },
                                }
                            ],
                            "source": {
                                "type": "Literal",
                                "value": "@teambit/mdx.ui.mdx-scope-context",
                                "raw": "'@teambit/mdx.ui.mdx-scope-context'"
                            },
                        }
                    ],
                    "sourceType": "module",
                    "comments": []
                }
            }
        };
        const wrapNode = {
            type: 'mdxJsxFlowElement',
            name: 'MDXScopeProvider',
            attributes: [
                {
                    type: 'mdxJsxAttribute',
                    name: 'components',
                    // value: `{${ids.join(', ')}}`,
                    value: {
                        type: 'mdxJsxAttributeValueExpression',
                        value: `{${ids.join(', ')}}`,
                        data: {
                            estree: {
                                type: 'Program',
                                body: [
                                    {
                                        type: 'ExpressionStatement',
                                        expression: {
                                            type: 'ObjectExpression',
                                            properties: ids.map((id) => ({
                                                type: 'Property',
                                                "method": false,
                                                "shorthand": true,
                                                "computed": false,
                                                key: { type: 'Identifier', name: id },
                                                value: { type: 'Identifier', name: id },
                                                kind: 'init',
                                            })),
                                        },
                                    },
                                ],
                            },
                        }
                    },
                },
            ],
            children: [],
        };
        wrapNode.children.push(...tree.children);
        tree.children = [importNode, wrapNode];
    };
}
export const mdxOptions = {
    remarkPlugins: [
        remarkGfm,
        remarkDirective,
        [detectFrontmatter, ['yaml']],
        extractMetadata,
        extractImports,
        adaptAdmonitions,
    ],
    rehypePlugins: [
        wrapWithScopeContext,
        rehypeMdxCodeProps,
    ],
    recmaPlugins: [
        recmaPluginInjectisMDXComponent,
    ],
    jsxImportSource: 'react',
    providerImportSource: '@mdx-js/react',
};
export const basicMdxOptions = {
    remarkPlugins: [
        remarkGfm,
        remarkDirective,
        extractImports,
        adaptAdmonitions,
    ],
    rehypePlugins: [
        rehypeMdxCodeProps,
    ],
    recmaPlugins: [
        recmaPluginInjectisMDXComponent,
    ],
    jsxImportSource: 'react',
    providerImportSource: '@mdx-js/react',
};
