// @bitdev/react.preview.react-docs-template@1.0.10/react-docs-template.tsx
import React, { ComponentType } from 'react';
import { DocsRootProps } from '@teambit/docs';
// import { DocsApp } from '@teambit/react.ui.docs-app';
import '@bitdev/react.preview.react-docs-app/dist/index.css';
// @ts-ignore
import { DocsApp } from '@bitdev/react.preview.react-docs-app';

/**
 * a reference to a provider function.
 */
export type DocsProvider = ComponentType<{ children: any }>;

type Root = { render: (element: React.ReactElement) => void };

// Both imports are dynamic so the bundler never statically validates named exports
// against whichever React version is installed.
// React 18/19: react-dom/client resolves → use createRoot.
// React 17:    react-dom/client rejects  → fall back to legacy ReactDOM.render.
let createRootImpl: ((container: HTMLElement) => Root) | null = null;
import('react-dom/client')
  .then((m: any) => {
    createRootImpl = (container) => m.createRoot(container);
  })
  .catch(() => {
    import('react-dom').then((m: any) => {
      createRootImpl = (container): Root => ({
        render: (element) => m.render(element, container),
      });
    });
  });

function createCompatRoot(container: HTMLElement): Root {
  if (createRootImpl) return createRootImpl(container);
  throw new Error('[react-compat] createCompatRoot called before module init settled');
}

let root: Root | undefined;

export const createDocsTemplate = (
  Provider: DocsProvider = React.Fragment,
  docsRootElement?: Element
) => {
  const DocsTemplate = ({
    componentId,
    docs,
    compositions,
    context,
  }: DocsRootProps) => {
    const rootElm = (docsRootElement ?? document.getElementById('root')) as HTMLElement;
    if (!root) {
      root = createCompatRoot(rootElm);
    }
    root.render(
      <Provider>
        <DocsApp
          componentId={componentId}
          docs={docs}
          compositions={compositions}
          context={context}
        />
      </Provider> as any
    );
  };
  DocsTemplate.apiObject = true;
  return DocsTemplate;
};
