// @teambit/vite.vitest-tester@0.4.13/config-utils.ts
import { relative, resolve, sep } from 'path';
import fsExtra from 'fs-extra';
import lodash from 'lodash';
import {
  TesterContext,
  ComponentPatternsEntry,
  ComponentPatternsMap,
} from '@teambit/tester';
import { Component } from '@teambit/component';
import { Workspace } from '@teambit/workspace';
import { DependencyResolverMain } from '@teambit/dependency-resolver';
import { type InlineConfig as ViteConfig, mergeConfig, loadConfigFromFile } from 'vite';

import { VitestTesterOptions } from './vitest-options.js';

const { flatten, isEmpty } = lodash;
const { existsSync, writeJSONSync } = fsExtra;

// pattern utils

const resolvePattern = (pattern: string, rootDirs: string[]) => {
  return rootDirs.map((dir) => resolve(dir, pattern));
}

const resolvePatternEntry = (patternEntry: ComponentPatternsEntry) => {
  const { componentDir, paths } = patternEntry;
  return paths.map((p) => {
    if (p.relative.startsWith('!')) {
      return `!${componentDir}/${p.relative.slice(1)}`;
    }
    return resolve(componentDir, p.path)
  });
}

export const resolveComponentPattern = (
  component: Component,
  patternEntry: ComponentPatternsEntry,
  opts: VitestTesterOptions,
  context: TesterContext
): string[] => {
  if (opts.resolveSpecPaths) {
    return opts.resolveSpecPaths(component, context);
  }
  const customPatterns = opts.patterns;
  // If pattern were provided to the specific instance of the tester, use them
  if (customPatterns && !isEmpty(customPatterns)) {
    customPatterns.map((customPattern) => {
      const rootDirs = opts.roots || [patternEntry.componentDir];
      return resolvePattern(customPattern, rootDirs);
    });
  }

  const resolvedPatterns = resolvePatternEntry(patternEntry);
  return resolvedPatterns;
}

export const getPatternsMap = (
  opts: VitestTesterOptions,
  context: TesterContext
): ComponentPatternsMap => {
  const target = opts.useDistsInCapsules
    ? context.patterns
    : (context.sourcePatterns || context.patterns);
  return target;
}

export const patternsToArray = (
  opts: VitestTesterOptions,
  context: TesterContext
): string[] => {
  const patternsMap = getPatternsMap(opts, context);
  const patternArray = flatten(
    patternsMap
      .toArray()
      .map(([component, patternEntry]) =>
        resolveComponentPattern(component, patternEntry, opts, context)
      )
  );
  if (sep === '\\') {
    return patternArray.map((pattern) => pattern.replace(/\\/g, '/'));
  }
  return patternArray;
}

const ensurePackageJson = (dir: string) => {
  const packageJsonPath = resolve(dir, 'package.json');
  if (!existsSync(packageJsonPath)) {
    writeJSONSync(packageJsonPath, {});
  }
}

const getCoverageInclude = (context: TesterContext, workspace: Workspace | undefined) => {
  if (!workspace || context.release) {
    return context.components.map(c => {
      return `${c.id.toObject().scope}_${c.id.toObject().name.replaceAll('/', '_')}@${c.id.toObject().version}`;
    });
  };

  return context.components.map(c => {
    const componentDir = workspace.componentDir(c.id);
    return relative(context.rootPath, componentDir);
  });
};

export const getExclude = (opts: VitestTesterOptions) => {
  // https://vitest.dev/config/#exclude default value without dist
  const excludeWithoutDist = [
    '**/node_modules/**',
    '**/cypress/**',
    '**/.{idea,git,cache,output,temp}/**',
    '**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build,eslint,prettier}.config.*'
  ];
  return opts.useDistsInCapsules ? excludeWithoutDist: undefined;
}

export const getConfig = async (
  configPath: string,
  context: TesterContext,
  workspace: Workspace | undefined,
  depResolver: DependencyResolverMain,
  opts: VitestTesterOptions,
  watch?: boolean,
): Promise<ViteConfig> => {
  const packageNames = context.components.map((component) => depResolver.getPackageName(component));

  // TODO: remove this once we have a better way to handle this
  ensurePackageJson(context.rootPath);

  const config: ViteConfig = {
    root: context.rootPath,
    configFile: false,
    envFile: false,
    plugins: [],
    server: {
      fs: {
        strict: false
      },
      watch: {
        ignored: packageNames.map((m) => `!**/node_modules/${m}/**`),
      }
    },
    cacheDir: resolve(context.rootPath, 'node_modules', '.vitest'),
    // @ts-ignore
    test: {
      globals: true,
      environment: 'jsdom',
      root: context.rootPath,
      testTimeout: 10000,
      css: {
        modules: {
          // set this as default since the relative file paths in workspaces and capsules are different
          classNameStrategy: 'non-scoped'
        }
      },
      server: {
        deps: {
          // TODO:
          inline: true,
          // https://github.com/vitest-dev/vitest/issues/7748
          // https://github.com/vitest-dev/vitest/issues/7950
          // inline: [ /^(?!.*vitest).*$/ ],
        },
      },
    }
  };

  if (context.debug) {
    // Nothing to do with Vitest
  }
  if (context.coverage) {
    // @ts-ignore
    config.test!.coverage = {
      provider: typeof opts.coverage !== 'string' ? 'istanbul' : opts.coverage,
      enabled: opts.coverage !== false,
      include: getCoverageInclude(context, workspace),
      reporter: ['text', 'json-summary'],
      exclude: [
        /**
         * Default Vitest excludes
         */
        'coverage/**',
        'dist/**',
        '**\/[.]**',
        'packages/*\/test?(s)/**',
        '**\/*.d.ts',
        '**\/virtual:*',
        '**\/__x00__*',
        '**\/\x00*',
        'cypress/**',
        'test?(s)/**',
        'test?(-*).?(c|m)[jt]s?(x)',
        '**\/*{.,-}{test,spec}?(-d).?(c|m)[jt]s?(x)',
        '**\/__tests__/**',
        '**\/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*',
        '**\/vitest.{workspace,projects}.[jt]s?(on)',
        '**\/.{eslint,mocha,prettier}rc.{?(c|m)js,yml}',
        '**\/dist/**',
      ]
    };
  }
  // @ts-ignore
  config.test!.watch = context.watch || watch;

  const vitestConfig = await loadConfigFromFile({
    command: 'build',
    mode: 'development'
  }, configPath)
  const realVitestConfig: ViteConfig = vitestConfig?.config || {};
  const result = mergeConfig(config, realVitestConfig);
  return result;
};
