// @teambit/vite.vitest-tester@0.4.13/coverage-utils.ts
import { join } from 'path';
import minimatch from 'minimatch';
import type { BaseCoverageOptions } from  'vitest/node';
import type { TesterContext, Tests } from '@teambit/tester';
import type { BuildContext } from '@teambit/builder';
import type { CoverageData, CoverageResult } from '@teambit/defender.coverage-report';
import { getCoverage, coverageByComponent } from '@teambit/defender.coverage-report';

export type MetricKey = 'lines' | 'functions' | 'branches' | 'statements';
export type ThresholdsKey = MetricKey | 100;
export type GeneralThresholdsKey = ThresholdsKey | 'perFile' | 'autoUpdate';

const METRIC_KEY_SET = new Set(['lines', 'functions', 'branches', 'statements']);
const GENERAL_THRESHOLD_KEYS = new Set([
  'lines',
  'functions',
  'branches',
  'statements',
  '100',
  'perFile',
  'autoUpdate',
]);

type MetricThresholds = Partial<Record<MetricKey, number>>;

function getRootDirMap(context: TesterContext): Record<string, string> {
  const rootDirMap: Record<string, string> = {};
  const capsules = (context as any as BuildContext)?.capsuleNetwork?.seedersCapsules || [];
  capsules.forEach(capsule => {
    rootDirMap[capsule.component.id.toString()] = capsule.path;
  });
  return rootDirMap;
}

function matchPattern(filePath: string, rootDir: string, pattern: string): boolean {
  const absoluteFilePath = join(rootDir, filePath);
  return minimatch(absoluteFilePath, pattern)
}

function collectMetricThresholds(
  source: BaseCoverageOptions['thresholds']
): MetricThresholds {
  const thresholds: MetricThresholds = {};
  if (!source) return thresholds;
  const is100 = source[100];
  for (const metric of METRIC_KEY_SET) {
    const value = source[metric as MetricKey];
    if (typeof value === 'number' && Number.isFinite(value)) {
      thresholds[metric as MetricKey] = value;
    } else if (is100) {
      thresholds[metric as MetricKey] = 100;
    }
  }
  return thresholds;
}

function checkMetricThresholds(
  data: CoverageData,
  thresholds: MetricThresholds,
  createMessage: (metric: MetricKey, actual: number, expected: number) => string
): Error[] {
  const errors: Error[] = [];
  if (!data) return errors;
  for (const metric of METRIC_KEY_SET) {
    const thresholdValue = thresholds[metric as MetricKey];
    if (typeof thresholdValue !== 'number') continue;
    const metricSummary = data?.[metric as MetricKey];
    const actualPct = typeof metricSummary?.pct === 'number' ? metricSummary.pct : undefined;
    if (actualPct === undefined) continue;
    if (actualPct === 0 && metricSummary.total === 0) continue;
    if (actualPct + Number.EPSILON < thresholdValue) {
      errors.push(new Error(createMessage(metric as MetricKey, actualPct, thresholdValue)));
    }
  }
  return errors;
}

export function applyThresholdsToResults(
  thresholds: BaseCoverageOptions['thresholds'],
  rootDirMap: Record<string, string>,
  covByComponent: Record<string, CoverageResult>,
  result: Tests
) {
  const generalThresholds = collectMetricThresholds(thresholds);

  for (const [componentId, componentCoverage] of Object.entries(covByComponent)) {
    if (!componentCoverage) continue;
    const componentRootDir = rootDirMap[componentId];
    const coverageErrors: Error[] = [];

    // Check general thresholds
    const generalCoverage = componentCoverage.total;
    if (generalCoverage && Object.keys(generalThresholds).length) {
      coverageErrors.push(
        ...checkMetricThresholds(
          generalCoverage,
          generalThresholds,
          (metric, actualPct, thresholdValue) =>
            `Coverage threshold not met for ${componentId} (${metric}: ${actualPct}% < ${thresholdValue}%)`
        )
      );
    }

    // Check per-pattern thresholds
    const patternThresholds: [string, MetricThresholds][] = Object.entries(thresholds as any).filter(
      ([key]) => !GENERAL_THRESHOLD_KEYS.has(key)
    ).map(([pattern, value]) => {
      return [pattern, collectMetricThresholds(value as any)];
    });
    if (patternThresholds.length && componentRootDir) {
      for (const [pattern, patternThreshold] of patternThresholds) {
        if (!patternThreshold) continue;
        componentCoverage.files.forEach(file => {
          if (!matchPattern(file.path, componentRootDir, pattern)) return;
          coverageErrors.push(
            ...checkMetricThresholds(
              file.data,
              patternThreshold,
              (metric, actualPct, thresholdValue) =>
                `Coverage threshold not met for ${componentId} (${metric}: ${actualPct}% < ${thresholdValue}% in ${file.path} matching "${pattern}")`
            )
          );
        })
      }
    }

    const componentResult = result.components.find(c => c.componentId.toString() === componentId);
    if (componentResult?.results) {
      Object.assign(componentResult.results, {
        coverage: componentCoverage,
      });
    }
    if (coverageErrors.length) {
      componentResult?.errors?.push(...coverageErrors);
    }
  }

  return result;
}

export async function checkCoverageThresholds(
  coverageOptions: BaseCoverageOptions,
  context: TesterContext,
  result: Tests
) : Promise<Tests> {
  const { thresholds } = coverageOptions;
  if (!thresholds) {
    return result;
  }

  const rootDirMap = getRootDirMap(context);
  const covByComponent = coverageByComponent(context, await getCoverage(context));

  return applyThresholdsToResults(thresholds, rootDirMap, covByComponent, result);
}
