// @teambit/vite.vitest-tester@0.4.13/vitest-tester.ts
/* eslint-disable max-classes-per-file */

import { createRequire } from 'module';
import fsExtra from 'fs-extra';
import { proxy } from 'comlink';
import { Logger } from '@teambit/logger';
import { HarmonyWorker } from '@teambit/worker';
import { Workspace, WorkspaceAspect } from '@teambit/workspace';
import { DependencyResolverMain, DependencyResolverAspect } from '@teambit/dependency-resolver';
import {
  Tester,
  CallbackFn,
  TesterContext,
  Tests,
} from '@teambit/tester';
import { DevFilesAspect, DevFilesMain } from '@teambit/dev-files';
import { EnvContext, EnvHandler } from '@teambit/envs';
import { startVitest, VitestPackageInstaller } from 'vitest/node';

import type { VitestWorker } from './vitest.worker.js';
import type { VitestOptions, VitestTesterOptions } from './vitest-options.js';
import {
  patternsToArray,
  getExclude,
  getConfig,
} from './config-utils.js';
import {
  type VitestFile,
  getResultFromFiles,
} from './report-utils.js';
import {
  checkCoverageThresholds,
} from './coverage-utils.js';

const { readFileSync, readJsonSync } = fsExtra;

const require = createRequire(import.meta.url);

export class VitestTester implements Tester {
  constructor(
    readonly id: string,
    readonly configPath: string,
    private vitestWorker: HarmonyWorker<VitestWorker>,
    private logger: Logger,
    private workspace: Workspace | undefined,
    private depResolver: DependencyResolverMain,
    private devFiles: DevFilesMain,
    private opts: VitestTesterOptions = {}
  ) {}

  static from(options: VitestOptions): EnvHandler<Tester> {
    const id = options.name || 'vitest-tester';
    return (context: EnvContext) => {
      const logger = context.createLogger(id);
      const worker = context.createWorker<VitestWorker>(
        id,
        require.resolve('./vitest.worker')
      );
      let workspace: Workspace | undefined;
      try {
        workspace = context.getAspect<Workspace>(WorkspaceAspect.id);
      } catch (err) {
        logger.info('unable to load workspace aspect');
      }
      const depResolver = context.getAspect<DependencyResolverMain>(DependencyResolverAspect.id);
      const devFiles = context.getAspect<DevFilesMain>(DevFilesAspect.id);
      return new VitestTester(
        id,
        options.config,
        worker,
        logger,
        workspace,
        depResolver,
        devFiles,
        options,
      );
    };
  }

  displayName = 'Vitest';

  version() {
    const filePath = require.resolve('vitest/package.json');
    const vitestPackageInfo = readJsonSync(filePath);
    return vitestPackageInfo.version;
  }

  displayConfig() {
    return readFileSync(this.configPath, 'utf8');
  }

  private onTestRunCompleteCb: CallbackFn | undefined;

  async onTestRunComplete(callback: CallbackFn) {
    this.onTestRunCompleteCb = callback;
  }

  async test(context: TesterContext): Promise<Tests> {
    if (this.opts.coverage !== false && context.release) {
      context.coverage = true;
    }

    const viteConfig = await getConfig(
      this.configPath,
      context,
      this.workspace,
      this.depResolver,
      this.opts,
      false
    );

    if (context.coverage) {
      for (const component of context.components) {
        const patterns = this.devFiles.getDevPatterns(component);
        // @ts-ignore
        viteConfig.test!.coverage!.exclude!.push(...patterns.map((pattern) => `${pattern}`));
      }

      // @ts-ignore
      viteConfig.test!.coverage!.exclude!.push('**/*.bit-app.*');
    }

    const include = patternsToArray(this.opts, context);
    const exclude = getExclude(this.opts);

    // @ts-ignore
    viteConfig.test!.include = include;

    if (exclude) {
      // @ts-ignore
      viteConfig.test!.exclude!.push(...exclude);
    }

    if ((context as any).updateSnapshot) {
      // @ts-ignore
      viteConfig.test!.update = true;
    }

    if ((context as any).watch) {
      // @ts-ignore
      viteConfig.test!.watch = true;
    }

    // skip installation checks
    const installErrors: Error[] = [];
    class CustomPackageInstaller extends VitestPackageInstaller {
      async ensureInstalled(dependency: string): Promise<boolean> {
        try {
          require.resolve(dependency, { paths: [viteConfig.root!] });
        }
        catch (err) {
          installErrors.push(err as Error);
          return false;
        }
        return true;
      }
    }

    const vitest = await startVitest('test', undefined, undefined, viteConfig as any, {
      packageInstaller: new CustomPackageInstaller()
    });

    if (context.watch) {
      // hold the process
      await new Promise(() => {});
    }

    await vitest!.close();

    const files = vitest!.state.getFiles();
    const errors = [
      ...installErrors,
      ...Array.from(vitest!.state.errorsSet as Set<Error>),
    ];
    const result = getResultFromFiles(files, errors, this.opts, context);

    if ((viteConfig as any).test!.coverage) {
      await checkCoverageThresholds((viteConfig as any).test!.coverage, context, result);
    }

    return result;
  }

  async watch(context: TesterContext): Promise<Tests> {
    const viteConfig = await getConfig(this.configPath, context, this.workspace, this.depResolver, this.opts, true);
    // @ts-ignore
    viteConfig.test!.include = patternsToArray(this.opts, context);
    // @ts-ignore
    viteConfig.test!.exclude = getExclude(this.opts);

    const workerApi = this.vitestWorker.initiate(
      context.ui
        ? { stdout: true, stderr: true, stdin: true }
        : { stdout: false, stderr: false, stdin: false }
    );

    const onTestCompleteCb = proxy((files?: VitestFile[], errors?: unknown[]) => {
      if (!this.onTestRunCompleteCb || !files) return;
      const tests = getResultFromFiles(files || [], errors || [], this.opts, context);
      this.onTestRunCompleteCb(tests);
    });

    await workerApi.onTestComplete(onTestCompleteCb);
    await workerApi.watch(viteConfig);

    return {} as Tests;
  }
}
