// @bitdev/react.webpack.refresh-overlay@1.0.0/dist/formatWebpackMessages.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = formatWebpackMessages;
/**
 * Production‑grade formatter for Webpack `stats.toJson()` output.
 * - Normalises/cleans messages (ANSI‑free, duplicate‑free)
 * - Enriches with *all* relevant metadata (file, line:col, module, issuer chain, stack)
 * - Preserves CRA behaviour of surfacing syntax errors first
 *
 * @see https://webpack.js.org/api/stats/#errors-and-warnings
 * @param {{errors:any[];warnings:any[]}} json
 * @returns {{errors:string[];warnings:string[]}}
 */
const SYNTAX_LABEL = 'Syntax error:';
// eslint-disable-next-line no-control-regex
const ANSI_RGX = /\u001b\[[0-9;]*m/g; // strip colour
const STACK_LINE_RGX = /^\s*at\s.*?:\d+:\d+(?:\)?$|$)/m; // Node/V8 stack
/* -------------------------------------------------------------------------- */
/* Helpers                                                                     */
/* -------------------------------------------------------------------------- */
/** Strip ANSI escape sequences. */
const stripAnsi = (s) => s.replace(ANSI_RGX, '');
/** Guarantee a primitive string for any "message‑like" input. */
function ensureString(msg) {
    if (typeof msg === 'string')
        return msg;
    if (msg && typeof msg.message === 'string')
        return msg.message;
    if (Array.isArray(msg))
        return msg.map(ensureString).join('\n');
    return String(msg);
}
/** Return `{head, body}` where head is first line, body is the rest. */
function splitHeadBody(txt) {
    const idx = txt.indexOf('\n');
    return idx === -1
        ? { head: txt.trim(), body: '' }
        : { head: txt.slice(0, idx).trim(), body: txt.slice(idx + 1).trim() };
}
/** Provide quick, context‑aware hints. */
function hintFor(head, body) {
    if (/Cannot find module|Module not found/i.test(head))
        return '💡  Check the import path or ensure the file/package exists.';
    if (/Attempted import error/i.test(head))
        return '💡  Verify named exports or add a default export.';
    if (/sass/i.test(body) && /Cannot find module/i.test(body))
        return '💡  Run `bit install sass` (or your package manager) to add Sass.';
    return '';
}
/** Extract first N stack lines for readability. */
function extractStack(body, max = 8) {
    return body
        .split(/\r?\n/)
        .filter((l) => STACK_LINE_RGX.test(l))
        .slice(0, max)
        .map((l) => '  ' + l.trim());
}
/* -------------------------------------------------------------------------- */
/* Core transformer                                                            */
/* -------------------------------------------------------------------------- */
function transform(raw) {
    /** @type {any} */
    const obj = raw && typeof raw === 'object' ? raw : null;
    const msgClean = stripAnsi(ensureString(raw)).trimEnd();
    const { head, body } = splitHeadBody(msgClean);
    const out = [head];
    if (body)
        out.push('', body);
    // ── Enrich with Webpack metadata ─────────────────────────────────────────
    if (obj) {
        const { moduleIdentifier, moduleName, loc, file, details } = obj;
        if (moduleIdentifier || moduleName) {
            out.push('', `❯ Module: ${moduleName || moduleIdentifier}`);
            if (loc)
                out.push(`   at ${loc}`);
        }
        if (file)
            out.push('', `File: ${file}`);
        if (Array.isArray(obj.moduleTrace) && obj.moduleTrace.length) {
            out.push('', 'Issuer chain:');
            obj.moduleTrace.forEach((t) => out.push(`  • ${t.originName} → ${t.moduleName}`));
        }
        if (details)
            out.push('', details.trim());
    }
    // ── Stack trace (first few lines) ───────────────────────────────────────
    const stackLines = extractStack(body);
    if (stackLines.length)
        out.push('', 'Stack trace:', ...stackLines);
    // ── Helpful hint ────────────────────────────────────────────────────────
    const hint = hintFor(head, body);
    if (hint)
        out.push('', hint);
    return out.filter(Boolean).join('\n');
}
/* -------------------------------------------------------------------------- */
/* Public API                                                                  */
/* -------------------------------------------------------------------------- */
function formatWebpackMessages(json) {
    let errors = json.errors.map(transform);
    let warnings = json.warnings.map(transform);
    // Only syntax errors if present (CRA convention)
    if (errors.some((e) => e.includes(SYNTAX_LABEL))) {
        errors = errors.filter((e) => e.includes(SYNTAX_LABEL));
    }
    // Deduplicate by headline (first line)
    const dedup = (arr) => {
        const seen = new Set();
        return arr.filter((msg) => {
            const headline = msg.split('\n')[0];
            if (seen.has(headline))
                return false;
            seen.add(headline);
            return true;
        });
    };
    return {
        errors: dedup(errors),
        warnings: dedup(warnings),
    };
}
