// @bitdev/react.webpack.refresh-overlay@1.0.0/dist/overlay-cta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initOverlayCta = initOverlayCta;
function initOverlayCta({ doc = window.document } = {}) {
    // ensure this runs only once
    if (doc.defaultView.__BIT_OVERLAY_CTA_AUTO_RAN__)
        return;
    // eslint-disable-next-line no-param-reassign
    doc.defaultView.__BIT_OVERLAY_CTA_AUTO_RAN__ = true;
    const isSameOrigin = (f) => {
        try {
            return !!f.contentDocument;
        }
        catch (_a) {
            return false;
        }
    };
    const looksLikeOverlay = (d) => {
        var _a, _b, _c, _d;
        return !!(d.querySelector('[data-ansi-line]') ||
            ((_b = (_a = d.body) === null || _a === void 0 ? void 0 : _a.textContent) === null || _b === void 0 ? void 0 : _b.includes('Failed to compile')) ||
            ((_d = (_c = d.body) === null || _c === void 0 ? void 0 : _c.textContent) === null || _d === void 0 ? void 0 : _d.includes('Unhandled runtime')));
    };
    function getMergedQueryParams(urlish = window.location.href, base = document.baseURI) {
        var _a;
        const merged = new URLSearchParams();
        const url = new URL(urlish, base);
        url.searchParams.forEach((v, k) => merged.append(k, v));
        let fragment = url.hash.startsWith('#') ? url.hash.slice(1) : url.hash;
        while (fragment) {
            const qPos = fragment.indexOf('?');
            if (qPos !== -1) {
                new URLSearchParams(fragment.slice(qPos + 1)).forEach((v, k) => merged.append(k, v));
                fragment = (_a = fragment.slice(qPos + 1).split('#')[1]) !== null && _a !== void 0 ? _a : '';
            }
            else {
                break;
            }
        }
        return merged;
    }
    function ctaDisabled(view) {
        var _a;
        const merged = new URLSearchParams();
        const urls = [];
        let currentView = view;
        const seen = new WeakSet();
        while (currentView && !seen.has(currentView)) {
            seen.add(currentView);
            try {
                const href = (_a = currentView.location) === null || _a === void 0 ? void 0 : _a.href;
                if (href)
                    urls.push(href);
                const { frameElement } = currentView;
                if (frameElement instanceof HTMLIFrameElement && frameElement.src)
                    urls.push(frameElement.src);
            }
            catch (_b) {
                // eslint-disable-next-line no-console
                console.warn('[overlay-cta] cross-origin barrier, skipping');
            }
            try {
                currentView =
                    currentView.parent !== currentView ? currentView.parent : null;
            }
            catch (_c) {
                break;
            }
        }
        urls.forEach((u) => {
            try {
                const p = getMergedQueryParams(u);
                p.forEach((v, k) => merged.set(k, v));
            }
            catch (_a) {
                // eslint-disable-next-line no-console
                console.warn('[overlay-cta] failed to parse URL', u);
            }
        });
        return merged.get('disableCta') === 'true';
    }
    function buildBar(d) {
        const bar = d.createElement('div');
        bar.dataset.overlayCta = '1';
        bar.style.cssText =
            'margin-top:16px;display:flex;' +
                'gap:12px;flex-wrap:wrap;align-items:center;';
        const btnBase = 'display:inline-flex;align-items:center;gap:8px;' +
            'height:32px;padding:0 16px;font:14px/1.2 sans-serif;font-weight:500;' +
            'border-radius:8px;cursor:pointer;';
        const ai = d.createElement('button');
        ai.type = 'button';
        ai.textContent = 'Fix with AI';
        ai.style.cssText =
            `${btnBase}border:none;color:#fff;` +
                `background:linear-gradient(120deg,#5D48FF 41.01%,#FA72A1 161.2%),#2603b3;`;
        ai.onclick = () => {
            var _a, _b;
            const trace = ((_a = d.querySelector('pre')) === null || _a === void 0 ? void 0 : _a.innerText) || 'N/A';
            const last = window.__LAST_BIT_ERROR__ || {};
            const payload = Object.assign(Object.assign({}, last), { error: Object.assign({ stack: trace }, (last.error || {})), trace, type: 'ai-fix-request' });
            (_b = window.parent) === null || _b === void 0 ? void 0 : _b.postMessage({ event: 'AI_FIX_REQUEST', payload }, '*');
        };
        ai.insertAdjacentHTML('afterbegin', `<svg width="18" height="18" viewBox="0 0 21 21" fill="none" style="flex:none">
         <defs><linearGradient id="g" x1="8" y1="-6" x2="28" y2="6"
                gradientUnits="userSpaceOnUse"><stop stop-color="#5D48FF"/>
                <stop offset="1" stop-color="#FA72A1"/></linearGradient></defs>
         <path d="M10.4 1.1c.1-.3.5-.3.6 0l1.6 5.1a6 6 0 003.8 3.8l5.1 1.6c.3.1.3.5 0
                  .6l-5.1 1.6a6 6 0 00-3.8 3.8l-1.6 5.1a.33.33 0 01-.6 0l-1.6-5.1a6 6
                  0 00-3.8-3.8L.9 10.8a.33.33 0 010-.6l5.1-1.6a6 6 0 003.8-3.8L10.4
                  1.1z" fill="url(#g)"/></svg>`);
        const edit = d.createElement('button');
        edit.type = 'button';
        edit.textContent = 'Edit code';
        edit.style.cssText = `${btnBase}background:#0f1b35;border:none;color:#fff;transition:background .2s;`;
        // eslint-disable-next-line no-return-assign
        edit.onmouseenter = () => (edit.style.background = '#18294a');
        // eslint-disable-next-line no-return-assign
        edit.onmouseleave = () => (edit.style.background = '#0f1b35');
        edit.onclick = () => {
            const raw = (typeof process !== 'undefined' &&
                typeof process.env !== 'undefined' &&
                process.env.VSCODE_PROXY_URI) ||
                '';
            const url = raw.replace(/p\d+/, '').replace(/p\{\{port\}\}/, '');
            if (url)
                window.open(url, '_blank');
        };
        edit.insertAdjacentHTML('afterbegin', `<svg width="16" height="16" viewBox="0 0 32 32" fill="currentColor" style="flex:none">
         <path d="M12.667 17.067l-.8 3.6c-.267.933.667 1.867 1.6 1.6l3.6-.8a1.067
                  1.067 0 00.667-.4l10.667-10.667a1.333 1.333 0 000-1.867l-2.8-2.8a1.333
                  1.333 0 00-1.867 0L13.333 15.6a1.067 1.067 0 00-.4.667zM15.2
                  18l9.467-9.467.933.933L16.133 18.933l-1.2.267.267-1.2z"/>
         <path d="M26.4 15.733l-1.867 1.867c1.333.8 2.133 2 2.133
                  3.333 0 2-1.867 4-4.667 4H7.467a2.133 2.133 0 01-2.133-2.133
                  2.267 2.267 0 011.867-2.133l2.933-.533-.933-2.933A7.467 7.467 0
                  019.2 12c0-3.2 2.667-5.867 5.867-5.867h.933l2.133-2.133a13.3
                  13.3 0 00-1.2-.4 9.86 9.86 0 00-1.867-.133C10.4 3.467 6.533 7.2
                  6.533 12c0 .667.133 1.467.4 2.133-.4 0-.8.133-1.2.4-1.733.8-2.8
                  2.4-2.8 4.4 0 2.667 2.133 4.8 4.8 4.8h14.533c4 0 7.333-2.933
                  7.333-6.667 0-2.133-1.067-4.133-2.933-5.333z"/></svg>`);
        const learn = d.createElement('a');
        learn.textContent = 'Learn more ↗';
        learn.href = 'https://bit.cloud/docs';
        learn.target = '_blank';
        learn.rel = 'noopener';
        learn.style.cssText =
            'font:14px/32px sans-serif;font-weight:500;color:#3b82f6;text-decoration:none;';
        bar.append(ai, edit, learn);
        return bar;
    }
    const scanned = new WeakSet();
    function injectIfNeeded(d) {
        var _a;
        const win = d.defaultView;
        if (ctaDisabled(win))
            return;
        const host = ((_a = d.querySelector('pre')) === null || _a === void 0 ? void 0 : _a.parentElement) || d.body;
        if (host && !host.querySelector('[data-overlay-cta]')) {
            host.appendChild(buildBar(d));
        }
    }
    const tryInjectTopLevel = () => {
        if (looksLikeOverlay(doc))
            injectIfNeeded(doc);
    };
    function scanFrames() {
        doc.querySelectorAll('iframe').forEach((f) => {
            var _a;
            if (scanned.has(f) || !isSameOrigin(f))
                return;
            scanned.add(f);
            const tryInject = () => {
                const d = f.contentDocument;
                if (d && looksLikeOverlay(d))
                    injectIfNeeded(d);
            };
            f.addEventListener('load', tryInject, { once: true });
            if (['interactive', 'complete'].includes((_a = f.contentDocument) === null || _a === void 0 ? void 0 : _a.readyState)) {
                tryInject();
            }
        });
    }
    tryInjectTopLevel();
    scanFrames();
    const mo = new MutationObserver(() => {
        scanFrames();
        tryInjectTopLevel();
    });
    if (doc.body)
        mo.observe(doc.body, { childList: true, subtree: true });
    // eslint-disable-next-line consistent-return
    return () => mo.disconnect();
}
