// @teambit/react.ui.component-highlighter@0.2.6/dist/element-highlighter/element-highlighter.compositions.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, createRef, useEffect } from 'react';
import { ElementHighlighter } from './element-highlighter';
import { MockTarget } from '../mock-component';
export const HighlightedElement = ({ style, targetStyle, watchMotion, className }) => {
    const targetRef = createRef();
    return (_jsxs("div", { className: className, style: { padding: '16px 16px 40px 16px', width: 300, fontFamily: 'sans-serif' }, children: [_jsx("div", { ref: targetRef, style: { width: 100, ...targetStyle }, children: "highlight target" }), _jsx(ElementHighlighter, { targetRef: targetRef, components: [MockTarget], style: style, watchMotion: watchMotion, placement: "bottom" })] }));
};
export const Customized = () => {
    return (_jsx(HighlightedElement, { style: {
            '--bit-highlighter-color': '#94deb4',
            '--bit-highlighter-color-hover': '#d0f1de',
            '--bit-highlighter-color-active': '#37b26c',
        } }));
};
export const Sizes = () => {
    return (_jsxs("div", { children: [_jsx(HighlightedElement, { style: { fontSize: 10 } }), _jsx(HighlightedElement, { style: { fontSize: 14 } }), _jsx(HighlightedElement, { style: { fontSize: 18 } })] }));
};
const fps = 30;
const frameInterval = 1000 / fps;
export const MovingElement = () => {
    const [margin, setMargin] = useState(0);
    useEffect(() => {
        const intervalId = setInterval(() => setMargin((x) => (x + 1) % 100), frameInterval);
        return () => clearInterval(intervalId);
    }, []);
    return _jsx(HighlightedElement, { targetStyle: { marginLeft: margin }, watchMotion: true });
};
export const FullscreenElement = () => {
    const targetRef = createRef();
    return (_jsxs("div", { style: { fontFamily: 'sans-serif' }, children: [_jsxs("div", { ref: targetRef, style: {
                    height: '100vh',
                    width: '100%',
                    background: '#bceed4',
                }, children: ["This element will cover the entire document,", _jsx("br", {}), "pushing the highlighter to the edge of the window.", _jsx("br", {}), "The highlighter should remain inside and expand no further than the document."] }), _jsx(ElementHighlighter, { targetRef: targetRef, components: [MockTarget], watchMotion: true })] }));
};
const edgeStyles = { position: 'absolute', background: 'cyan', padding: 30 };
const centerStyles = {
    top: { top: -30, left: '50%', transform: 'translate(-50%,0)' },
    right: { right: -30, top: '50%', transform: 'translate(0, -50%)' },
    bottom: { bottom: -30, left: '50%', transform: 'translate(-50%,0)' },
    left: { left: -30, top: '50%', transform: 'translate(0, -50%)' },
};
export function OffscreenElements() {
    const target01 = createRef();
    const target02 = createRef();
    const target03 = createRef();
    const target04 = createRef();
    return (_jsx("div", { style: { fontFamily: 'sans-serif', height: '100%' }, children: _jsxs("div", { style: { position: 'relative', width: '100%', height: '100%', overflow: 'hidden' }, children: [_jsx("div", { ref: target01, style: { ...edgeStyles, ...centerStyles.top }, children: "top" }), _jsx("div", { ref: target02, style: { ...edgeStyles, ...centerStyles.right }, children: "right" }), _jsx("div", { ref: target03, style: { ...edgeStyles, ...centerStyles.bottom }, children: "bottom" }), _jsx("div", { ref: target04, style: { ...edgeStyles, ...centerStyles.left }, children: "left" }), _jsx(ElementHighlighter, { targetRef: target01, components: [MockTarget], watchMotion: true }), _jsx(ElementHighlighter, { targetRef: target02, components: [MockTarget], watchMotion: true }), _jsx(ElementHighlighter, { targetRef: target03, components: [MockTarget], watchMotion: true }), _jsx(ElementHighlighter, { targetRef: target04, components: [MockTarget], watchMotion: true })] }) }));
}
