// @teambit/oxc.linter.oxlint-node@2.0.2/dist/oxlint-node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OxlintNode = void 0;
const execa_1 = __importDefault(require("execa"));
const child_process_1 = require("child_process");
const resolve_bin_1 = require("./resolve-bin");
const plugins_1 = require("./plugins");
const rules_1 = require("./rules");
const fixes_1 = require("./fixes");
class OxlintNode {
    constructor(binPath, plugins, rules, fixes, formats = ['default'], configPath, tsconfigPath, options = {}, tsgolintBinPath) {
        this.binPath = binPath;
        this.plugins = plugins;
        this.rules = rules;
        this.fixes = fixes;
        this.formats = formats;
        this.configPath = configPath;
        this.tsconfigPath = tsconfigPath;
        this.options = options;
        this.tsgolintBinPath = tsgolintBinPath;
    }
    version() {
        return (0, child_process_1.execFileSync)(this.binPath, ['--version']).toString().trim();
    }
    getConfigCliArg() {
        return this.configPath ? `--config=${this.configPath}` : '';
    }
    getTsConfigCliArg() {
        return this.tsconfigPath ? `--tsconfig=${this.tsconfigPath}` : '';
    }
    getFormatCliArg(format) {
        return `--format=${format}`;
    }
    getExtraCliArgs() {
        const args = [];
        const o = this.options;
        if (o.ignorePath)
            args.push(`--ignore-path=${o.ignorePath}`);
        if (o.ignorePatterns)
            o.ignorePatterns.forEach((p) => args.push(`--ignore-pattern=${p}`));
        if (o.noIgnore)
            args.push('--no-ignore');
        if (o.quiet)
            args.push('--quiet');
        if (o.denyWarnings)
            args.push('--deny-warnings');
        if (o.maxWarnings !== undefined)
            args.push(`--max-warnings=${o.maxWarnings}`);
        if (o.silent)
            args.push('--silent');
        if (o.threads !== undefined)
            args.push(`--threads=${o.threads}`);
        if (o.printConfig)
            args.push('--print-config');
        if (o.reportUnusedDisableDirectivesSeverity) {
            args.push(`--report-unused-disable-directives-severity=${o.reportUnusedDisableDirectivesSeverity}`);
        }
        else if (o.reportUnusedDisableDirectives) {
            args.push('--report-unused-disable-directives');
        }
        if (o.disableNestedConfig)
            args.push('--disable-nested-config');
        if (o.typeAware)
            args.push('--type-aware');
        if (o.typeCheck)
            args.push('--type-check');
        return args;
    }
    async run(paths = []) {
        const result = {};
        await Promise.all(this.formats.map(async (format) => {
            const args = this.toCliArgs(format, paths);
            let stdout;
            const env = {};
            if (format !== 'json')
                env.FORCE_COLOR = '1';
            if (this.tsgolintBinPath)
                env.OXLINT_TSGOLINT_PATH = this.tsgolintBinPath;
            const execaOpts = Object.keys(env).length > 0 ? { env } : {};
            try {
                ({ stdout } = await (0, execa_1.default)(this.binPath, args, execaOpts));
            }
            catch (err) {
                // oxlint exits with code 1 when it finds lint issues — that is expected.
                // Only re-throw if there is no stdout, which indicates a real failure
                // (e.g. binary not found, bad arguments).
                if (!err.stdout)
                    throw err;
                stdout = err.stdout;
            }
            result[format] = stdout;
            if (format === 'json') {
                try {
                    result[format] = JSON.parse(stdout);
                }
                catch {
                    throw new Error(`oxlint returned non-JSON output:\n${stdout}`);
                }
            }
        }));
        return result;
    }
    toCliArgs(format, paths = []) {
        const args = [
            ...paths,
            ...this.plugins.toCliArgs(),
            ...this.rules.toCliArgs(),
            ...this.fixes.toCliArgs(),
            ...this.getExtraCliArgs(),
            this.getConfigCliArg(),
            this.getTsConfigCliArg(),
            this.getFormatCliArg(format),
        ];
        return args.filter(Boolean);
    }
    static create(options) {
        const plugins = plugins_1.Plugins.from(options.pluginsFlags || {});
        const rules = rules_1.Rules.from(options.rulesFlags || []);
        const fixes = fixes_1.Fixes.from(options.fixesFlags || {});
        const { pluginsFlags, rulesFlags, fixesFlags, formats, configPath, tsconfigPath, binPath: optBinPath, tsgolintBinPath: optTsgolintBinPath, ...rest } = options;
        const binPath = optBinPath || (0, resolve_bin_1.resolveBin)('oxlint', { executable: 'oxlint', paths: [__dirname] });
        let tsgolintBinPath = optTsgolintBinPath;
        if (!tsgolintBinPath && (rest.typeAware || rest.typeCheck)) {
            try {
                tsgolintBinPath = (0, resolve_bin_1.resolveBin)('oxlint-tsgolint', { executable: 'tsgolint', paths: [__dirname] });
            }
            catch {
                // tsgolint not found — oxlint will report its own error if needed
            }
        }
        return new OxlintNode(binPath, plugins, rules, fixes, formats, configPath, tsconfigPath, rest, tsgolintBinPath);
    }
}
exports.OxlintNode = OxlintNode;
