// @teambit/preview.react-preview@1.1.8/dist/webpack-error-reporting.transformer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createErrorReportingTransformer = createErrorReportingTransformer;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const html_webpack_inject_plugin_1 = __importDefault(require("html-webpack-inject-plugin"));
// define it here instead of importing it from @teambit/ui-foundation.ui.rendering.html to avoid extra dep
const ERROR_EVENT = "_ERROR_OCCURRED_";
/**
 * Creates a webpack transformer that uses html-webpack-inject-plugin to inject
 * the error reporting script into HTML files
 */
function createErrorReportingTransformer() {
    return (configMutator) => {
        const config = configMutator.raw;
        let errorReportingScript;
        try {
            const scriptPath = path_1.default.resolve(__dirname, './bit-error-reporting.js');
            errorReportingScript = fs_1.default.readFileSync(scriptPath, 'utf8');
            errorReportingScript = errorReportingScript.replace(/\$\{ERROR_EVENT\}/g, ERROR_EVENT);
        }
        catch (error) {
            // eslint-disable-next-line no-console
            console.warn('Failed to load error reporting script:', error);
            errorReportingScript = `
        (function() {
          if (window.parent && window !== window.parent) {
            window.addEventListener('error', function(event) {
              window.parent.postMessage({
                event: '${ERROR_EVENT}',
                payload: {
                  message: event.message || 'Unknown error',
                  filename: event.filename,
                  type: 'runtime-error',
                  context: { url: window.location.href }
                }
              }, '*');
            }, true);
          }
        })();
      `;
        }
        config.plugins = config.plugins || [];
        config.plugins.push(new html_webpack_inject_plugin_1.default({
            externals: [
                {
                    tagName: 'script',
                    attributes: {},
                    innerHTML: errorReportingScript
                }
            ],
            prepend: true
        }));
        return configMutator;
    };
}
