// @teambit/preview.react-preview@1.1.8/dist/webpack-caching.transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCacheControlTransformer = createCacheControlTransformer;
function createCacheControlTransformer() {
    return (configMutator) => {
        const config = configMutator.raw;
        config.devtool = 'eval-cheap-module-source-map';
        if (!config.optimization) {
            config.optimization = {};
        }
        if (!config.devServer) {
            config.devServer = {};
        }
        if (!config.devServer.setupMiddlewares) {
            config.devServer.setupMiddlewares = [];
        }
        config.optimization = {
            ...(config.optimization || {}),
            minimize: false,
            concatenateModules: true,
            moduleIds: 'named',
            runtimeChunk: 'single',
            splitChunks: {
                chunks: 'all',
                maxInitialRequests: 2,
                minSize: 2000000,
                maxSize: 5000000,
                cacheGroups: {
                    vendors: {
                        test: /[\\/]node_modules[\\/]/,
                        name: 'vendors',
                        priority: 30,
                        enforce: true,
                        reuseExistingChunk: true,
                    },
                    main: {
                        test: /\.(js|jsx|ts|tsx)$/,
                        name: 'main',
                        priority: 10,
                        enforce: true,
                        chunks: 'all',
                    }
                },
            },
        };
        const originalSetupMiddlewares = config.devServer.setupMiddlewares;
        config.devServer.setupMiddlewares = (middlewares, devServer) => {
            middlewares.unshift({
                name: 'performance-middleware',
                middleware: (req, res, next) => {
                    if (req.path.endsWith('.js')) {
                        res.setHeader('Cache-Control', 'no-cache');
                    }
                    if (req.path === '/' || req.path.endsWith('.html')) {
                        res.setHeader('Cache-Control', 'public, max-age=1');
                    }
                    next();
                }
            });
            return typeof originalSetupMiddlewares === 'function'
                ? originalSetupMiddlewares(middlewares, devServer)
                : middlewares;
        };
        return configMutator;
    };
}
