// @teambit/rspack.modules.rspack-config-mutator@1.0.22/config-mutator.spec.ts
import type { Configuration, Target, RspackPluginInstance } from '@rspack/core';
import { RspackConfigMutator } from './rspack-config-mutator';

describe('RspackConfigMutator', () => {
  let mutator: RspackConfigMutator;
  let initialConfig: Configuration;

  beforeEach(() => {
    initialConfig = {};
    mutator = new RspackConfigMutator(initialConfig);
  });

  it('should merge configuration correctly', () => {
    const configToMerge: Configuration = { mode: 'development' };
    mutator.merge(configToMerge);
    expect(mutator.config).toEqual({ mode: 'development' });
  });

  it('should set the mode correctly', () => {
    mutator.setMode('production');
    expect(mutator.config.mode).toBe('production');
  });

  it('should set the target correctly', () => {
    const target: Target = 'web';
    mutator.setTarget(target);
    expect(mutator.config.target).toBe(target);
  });

  it('should add a plugin correctly', () => {
    const plugin: RspackPluginInstance = { apply: () => {} };
    mutator.addPlugin(plugin);
    expect(mutator.config.plugins).toContain(plugin);
  });

  it('should remove a plugin correctly', () => {
    const plugin = new (class HtmlWebpackPlugin {
      apply() {}
    })();
    mutator.addPlugin(plugin);
    mutator.removePlugin('HtmlWebpackPlugin');
    expect(mutator.config.plugins).not.toContain(plugin);
  });

  it('should add multiple plugins correctly', () => {
    const plugin1: RspackPluginInstance = { apply: () => {} };
    const plugin2: RspackPluginInstance = { apply: () => {} };
    mutator.addPlugins([plugin1, plugin2]);
    expect(mutator.config.plugins).toContain(plugin1);
    expect(mutator.config.plugins).toContain(plugin2);
  });

  it('should handle null, undefined, and falsy plugins gracefully', () => {
    mutator.addPlugins([null, undefined, false, 0, '']);
    expect(mutator.config.plugins).toEqual([null, undefined, false, 0, '']);
  });

  it('should merge nested configurations correctly', () => {
    const baseConfig: Configuration = {
      mode: 'development',
      devServer: {
        port: 3000,
      },
    };

    const configToMerge: Configuration = {
      devServer: {
        hot: true,
      },
    };

    mutator.merge(baseConfig);
    mutator.merge(configToMerge);

    expect(mutator.config).toEqual({
      mode: 'development',
      devServer: {
        port: 3000,
        hot: true,
      },
    });
  });

  it('should clone the config mutator with the config', () => {
    const baseConfig: Configuration = {
      mode: 'development',
      devServer: {
        port: 3000,
      },
    };

    mutator = new RspackConfigMutator(baseConfig);
    const clonedMutator = mutator.clone();

    expect(mutator).toBe(mutator);
    expect(clonedMutator).toBe(clonedMutator);
    expect(clonedMutator).not.toBe(mutator);
    expect(clonedMutator.config).toEqual({
      mode: 'development',
      devServer: {
        port: 3000,
      },
    });
  });

  it('should set the mode correctly with a nested configuration', () => {
    const baseConfig: Configuration = {
      mode: 'development',
      devServer: {
        port: 3000,
      },
    };

    mutator.merge(baseConfig);
    mutator.setMode('production');

    expect(mutator.config.mode).toBe('production');
  });

  it('should merge correctly with a nested devserver configuration', () => {
    const baseConfig: Configuration = {
      devServer: {
        hot: true,
        port: 5000,
      },
    };

    const configToMerge: Configuration = {
      devServer: {
        headers: {
          'Access-Control-Allow-Origin': '*',
        },
      },
    };

    mutator.merge(baseConfig);

    mutator.merge(configToMerge);

    expect(mutator.config.devServer).toEqual({
      hot: true,
      port: 5000,
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    });
  });

  describe('addTypescript()', () => {
    it('should add typescript rule', () => {
      mutator.addTypescript();
      expect(mutator.config.module!.rules).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            test: /\.ts$/
          })
        ])
      );
    });
  
    it('should add typescript & react rule', () => {
      mutator.addTypescript({ react: true });
      expect(mutator.config.module!.rules).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            test: /\.tsx?$/
          })
        ])
      );
    });

    it('should throw error if typescript rule already exists', () => {
      mutator.addTypescript()
      expect(mutator.addTypescript.bind(mutator)).toThrow('A custom typescript rule was detected, please remove it in order to use this utility.')
    });

    it('should throw error if typescript & react rule already exists', () => {
      mutator.addTypescript({ react: true })
      expect(mutator.addTypescript.bind(mutator, { react: true })).toThrow('A custom typescript rule was detected, please remove it in order to use this utility.')
    });

    it('should only allow single rule for typescript and react and throw error if one already exists', () => {
      mutator.addTypescript()
      expect(mutator.addTypescript.bind(mutator, { react: true })).toThrow('A custom typescript rule was detected, please remove it in order to use this utility.')

      mutator.config.module!.rules = [];

      mutator.addTypescript({ react: true })
      expect(mutator.addTypescript.bind(mutator)).toThrow('A custom typescript rule was detected, please remove it in order to use this utility.')
    });
  });

  it('should add configuration for css modules', () => {
    mutator.addCssModules();

    expect(mutator.config).toEqual(expect.objectContaining({
      experiments: expect.objectContaining({
        css: true,
      })
    }));

    expect(mutator.config.module).toEqual(
      expect.objectContaining({
        parser: expect.objectContaining({
          'css/auto': {
            namedExports: false,
          },
        }),
        generator: expect.objectContaining({
          'css/auto': {
            esModule: false,
            exportsConvention: 'camel-case-only',
            localIdentName: '[uniqueName]-[id]-[local]',
          },
        }),
      })
    );
  });
});
