// @teambit/rspack.modules.rspack-config-mutator@1.0.22/rspack-config-mutator.ts
// TODO(env): replace teambit.node/node@2.0.1 with a workspace-owned ESM
// module env (e.g. teambit.rspack/envs/rspack-module-env) whose PackageGenerator
// emits `"type": "module"` on the component package.json. That eliminates the
// suppression below — @rspack/core@2 is ESM-only and NodeNext rejects type
// imports from ESM packages inside CJS-typed packages, even for `import type`.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore TS1479: type-only import from ESM-only @rspack/core under NodeNext (CJS package)
import type {
  Configuration,
  RspackPluginFunction,
  RspackPluginInstance,
  Target,
  Plugins,
  RuleSetRules
} from '@rspack/core';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import {
  DevServerContext,
  BundlerMode,
  Target as BundlerTarget,
} from '@teambit/bundler';
import {
  inject,
  type InjectedHtmlElement as CustomHtmlElement,
} from '@teambit/html.modules.inject-html-element';
import type { Component } from '@teambit/component';
import { mergeRspackConfig } from './merge';

export type RspackConfigTransformContext =
  GlobalRspackConfigTransformContext & {
    target: BundlerTarget;
  };
export type RspackConfigDevServerTransformContext =
  GlobalRspackConfigTransformContext & DevServerContext;

export type GlobalRspackConfigTransformContext = {
  mode: BundlerMode;
  /**
   * Whether the config is for an env template bundling
   */
  isEnvTemplate?: boolean;
  /**
   * A path for the host root dir
   * Host root dir is usually the env root dir
   * This can be used in different bundle options which run require.resolve
   * for example when configuring webpack aliases or webpack expose loader on the peers deps
   */
  hostRootDir?: string;
};

export type Transformers = RspackConfigTransformer[];

export type RspackConfigTransformer = (
  config: RspackConfigMutator,
  context: RspackConfigTransformContext
) => RspackConfigMutator;

export type RspackConfigDevServerTransformer = (
  config: RspackConfigMutator,
  context: RspackConfigDevServerTransformContext
) => RspackConfigMutator;

export type RspackConfigAppTransformer = (
  config: RspackConfigMutator,
  context: RspackConfigAppTransformContext
) => RspackConfigMutator;

export type RspackConfigAppTransformContext = {
  /**
   * app name
   */
  appName: string;

  /**
   * is dev mode or not.
   */
  dev?: boolean;

  /**
   * is the rspack configuration is intended for server side rendering.
   */
  ssr?: boolean;

  /**
   * services list provided by the platform, if any.
   */
  services?: {
    appName: string;
    url?: string;
    port?: number;
  }[];

  /**
   * component directory in the workspace or the component capsule path in other cases.
   */
  componentDir: string;

  /**
   * App component instance.
   */
  appComponent: Component;
};

export class RspackConfigMutator {
  raw: Configuration;

  constructor(public config: Configuration) {
    // this is an alias for consistency with webpack-config-mutator.
    this.raw = config;
  }

  /**
   * Merge the given configuration with the current configuration.
   * @param config Configuration to merge.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.merge({ mode: "development" });
   * ```
   */
  merge(config: Configuration): RspackConfigMutator {
    this.config = mergeRspackConfig(this.config, config);
    return this;
  }

  /**
   * Clone the mutator with the current configuration.
   * @returns New mutator instance.
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.clone();
   * ```
   */
  clone(): RspackConfigMutator {
    return new RspackConfigMutator(mergeRspackConfig({}, this.config));
  }

  /**
   * Set the mode of the configuration.
   * @param mode Mode to set.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.setMode("development");
   * ```
   */
  setMode(
    mode: 'development' | 'production' | 'none' | undefined
  ): RspackConfigMutator {
    this.config.mode = mode;
    return this;
  }

  /**
   * Set the target of the configuration.
   * @param target Target to set.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.setTarget("web");
   * ```
   */
  setTarget(target: Target): RspackConfigMutator {
    this.config.target = target;
    return this;
  }

  /**
   * Add a plugin to the configuration.
   * @param plugin Plugin to add.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.addPlugin(new HtmlWebpackPlugin());
   * ```
   */
  addPlugin(
    plugin:
      | false
      | ''
      | 0
      | RspackPluginInstance
      | RspackPluginFunction
      | null
      | undefined,
    position?: 'append' | 'prepend'
  ): RspackConfigMutator {
    this.config.plugins = this.config.plugins || [];
    if (position === 'prepend') {
      this.config.plugins.unshift(plugin);
    } else {
      this.config.plugins.push(plugin);
    }

    return this;
  }

  /**
   * Remove a plugin from the configuration.
   * @param plugin Plugin name to remove.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.removePlugin("HtmlWebpackPlugin");
   * ```
   */
  removePlugin(plugin: string): RspackConfigMutator {
    this.config.plugins = this.config.plugins || [];
    this.config.plugins = this.config.plugins.filter((p) => {
      if (p) {
        return p.constructor.name !== plugin;
      }
      return false;
    });
    return this;
  }

  /**
   * Add multiple plugins to the configuration.
   * @param plugins Plugins to add.
   * @returns RspackConfigMutator instance
   * @example
   * ```ts
   * const mutator = new RspackConfigMutator({});
   * mutator.addPlugins([new HtmlWebpackPlugin(), new MiniCssExtractPlugin()]);
   * ```
   */
  addPlugins(
    plugins: (
      | false
      | ''
      | 0
      | RspackPluginInstance
      | RspackPluginFunction
      | null
      | undefined
    )[],
    position?: 'append' | 'prepend'
  ): RspackConfigMutator {
    this.config.plugins = this.config.plugins || [];
    if (position === 'prepend') {
      this.config.plugins.unshift(...plugins);
    } else {
      this.config.plugins.push(...plugins);
    }

    return this;
  }

  /**
   * To be used to ignore replace packages with global variable
   * Useful when trying to offload libs to CDN
   * @param externalDeps
   * @returns RspackConfigMutator instance
   */
  addExternals(
    externalDeps: Configuration['externals'] | any
  ): RspackConfigMutator {
    if (!externalDeps) return this;
    let { externals } = this.config;
    if (!externals) {
      externals = externalDeps;
    } else if (Array.isArray(externalDeps)) {
      externals = externalDeps.concat(externals);
    } else if (
      Array.isArray(externals) ||
      externalDeps.constructor === Function ||
      externalDeps.constructor === RegExp
    ) {
      externals = [externalDeps].concat(externals);
    } else if (externalDeps instanceof Object && externals instanceof Object) {
      // @ts-ignore
      externals = {
        ...externals,
        ...externalDeps,
      };
    }

    this.config.externals = externals;
    return this;
  }

  /**
   * Add aliases
   * @param aliases
   * @returns RspackConfigMutator instance
   */
  addAliases(aliases: {
    [index: string]: string | false | string[];
  }): RspackConfigMutator {
    if (!this.config.resolve) {
      this.config.resolve = {};
    }
    if (!this.config.resolve.alias) {
      this.config.resolve.alias = {};
    }
    Object.assign(this.config.resolve.alias, aliases);
    return this;
  }

  /**
   * Add a custom element to the html template
   * @param element
   * @returns
   * @example
   * addElementToHtmlTemplate({ parent: 'head', position: 'append', tag: 'script', attributes: { src: 'https://cdn.com/script.js', async: true } });
   * addElementToHtmlTemplate({ parent: 'body', position: 'prepend', tag: 'script', content: 'console.log("hello world")' });
   */
  addElementToHtmlTemplate(element: CustomHtmlElement): RspackConfigMutator {
    if (!this.config.plugins || this.config.plugins.length === 0) {
      return this;
    }

    const HtmlWebpackPluginConstructorName = 'HtmlWebpackPlugin';
    const HtmlRspackPluginConstructorName = 'HtmlRspackPlugin';
    const supportedHtmlPlugins = [
      HtmlWebpackPluginConstructorName,
      HtmlRspackPluginConstructorName,
    ];

    const htmlPlugins = this.config.plugins.filter((plugin) =>
      supportedHtmlPlugins.includes(plugin!.constructor.name)
    ) as Plugins;

    if (!htmlPlugins) {
      return this;
    }

    // iterate over all html plugins and add the scripts to the html
    htmlPlugins.forEach((htmlPlugin) => {
      const pluginConstructorName = htmlPlugin!.constructor.name;
      if (pluginConstructorName === HtmlWebpackPluginConstructorName) {
        const htmlWebpackPlugin = htmlPlugin as HtmlWebpackPlugin;
        const htmlContent =
          typeof htmlWebpackPlugin.userOptions?.templateContent === 'function'
            ? (htmlWebpackPlugin?.userOptions?.templateContent({}) as string)
            : (htmlWebpackPlugin?.userOptions?.templateContent as string);

        const newHtmlContent = inject(htmlContent, element);

        htmlWebpackPlugin.userOptions.templateContent = newHtmlContent;
      } else if (pluginConstructorName === HtmlRspackPluginConstructorName) {
        const htmlRspackPlugin = htmlPlugin as any;
        const htmlContent = htmlRspackPlugin?._args?.[0].templateContent ?? '';
        const newHtmlContent = inject(htmlContent, element);
        htmlRspackPlugin._args[0].templateContent = newHtmlContent;
      }
    });

    return this;
  }

  private findRule(ruleTest: string, rules?: any): any {
    return (rules || this.config.module?.rules as RuleSetRules)?.find(((rule: any) => {
      if(!rule) return false;

      if(rule?.oneOf?.length > 0) {
        return !!this.findRule(ruleTest, rule.oneOf);
      }

      if(typeof rule.test === 'string') {
        return rule.test.includes(ruleTest);
      }
      
      if(typeof rule.test === 'object' && rule.test instanceof RegExp) {
        return rule.test.test(ruleTest);
      }

      return false;
    }));
  }

  addTypescript({ react: enableReact }: { react?: boolean } = { react: false }): RspackConfigMutator {
    const tsRule = '.ts';
    const tsReactRule = '.tsx';
    const tsRuleRegex = enableReact ? /\.tsx?$/ : /\.ts$/;
    const isRuleProvided = !!this.findRule(tsRule) || !!this.findRule(tsReactRule);

    if(isRuleProvided) {
      throw new Error('A custom typescript rule was detected, please remove it in order to use this utility.')
    }

    const isDevelopment = this.config.mode !== 'production';

    this.merge({
      module: {
        rules: [{
          test: tsRuleRegex,
          loader: "builtin:swc-loader",
          options: {
            sourceMap: true,
            jsc: {
              parser: {
                syntax: "typescript",
                tsx: enableReact,
              },
              transform: {
                react: {
                  runtime: "automatic",
                  development: isDevelopment,
                },
              },
            },
            env: {
              targets: 'Chrome >= 48',
            },
          },
        }],
      }
    });

    return this;
  }

  addCssModules(): RspackConfigMutator {
    const isDevelopment = this.config.mode !== 'production';

    this.merge({
      experiments: {
        css: true,
      },
      module: {
        parser: {
          'css/auto': {
            namedExports: false,
          },
          css: {
            namedExports: false,
          },
          'css/module': {
            namedExports: false,
          },
        },
        generator: {
          'css/auto': {
            esModule: false,
            exportsConvention: 'camel-case-only',
            localIdentName: isDevelopment ? '[uniqueName]-[id]-[local]' : '[hash]',
          },
          css: {
            esModule: false,
          },
          'css/module': {
            esModule: false,
            exportsConvention: 'camel-case-only',
            localIdentName: isDevelopment ? '[uniqueName]-[id]-[local]' : '[hash]',
          },
        },
      }
    });

    return this;
  }
}

export function runTransformers(
  baseConfig: Configuration,
  transformers: Array<RspackConfigTransformer>,
  // @FIX: fix context type, maybe with guard rails
  // context: RspackConfigAppTransformContext | RspackConfigTransformContext
  context: any
): Configuration {
  if (!Array.isArray(transformers)) return baseConfig;
  const mutator = new RspackConfigMutator(baseConfig);
  return transformers.reduce((c, transformer) => {
    return transformer(mutator, context).config;
  }, baseConfig);
}

export function runAppTransformers(
  baseConfig: Configuration,
  transformers: RspackConfigAppTransformer[],
  context: RspackConfigAppTransformContext
): Configuration {
  if (!Array.isArray(transformers)) return baseConfig;
  const mutator = new RspackConfigMutator(baseConfig);
  return transformers.reduce((c, transformer) => {
    return transformer(mutator, context).config;
  }, baseConfig);
}
