// @bitdev/react.generators.react-templates@6.0.5/theme-template/files/tokens-file.ts
import { ComponentContext } from "@teambit/generator";


export function themeTokenFile(context: ComponentContext) {
  return {
    relativePath: `${context.name}-tokens.ts`,
    content: `

/**
 * ${context.namePascalCase} tokens theme.
 * Include all tokens in this object.
 */
export function ${context.nameCamelCase}Tokens() {
  const tokens = {
    /**
     * Color Palette
     */
    colors: {
      primary: {
        default: '#6750A4', // Main brand color (core identity)
        hover: '#7A63B7', // Brand color for hover states.
        active: '#554090', // Brand color for hover states.
      },
      secondary: {
        default: '#e8def8', // Brand accent color (subtle contrast)
        hover: '#e0d6f0', // Brand accent hover color
        active: '#d8cef0', // Brand accent active color
      },
      surface: {
        background: '#f8f9fa', // Default background color (light, airy)
        primary: '#ffffff', // Primary content surface (clean, crisp)
        secondary: '#f0f0f5', // Secondary surface (subtle differentiation)
      },
      text: {
        primary: '#212121', // Main text color (high contrast)
        default: '#212121', // Main text color (high contrast)
        secondary: '#666666', // Secondary text (subtle emphasis)
        inverse: '#ffffff', // Text on dark backgrounds (clear readability)
      },
      status: {
        positive: { default: '#4CAF50', subtle: '#c8e6c9' },
        negative: { default: '#F44336', subtle: '#ffcdd2' },
        warning: { default: '#FFC107', subtle: '#ffecb3' },
        info: { default: '#2196F3', subtle: '#bbdefb' },
      },
      overlay: 'rgba(0, 0, 0, 0.5)', // Semi-transparent overlay (modals, etc.)
    },

    borders: {
      default: {
        color: '#e0e0e0', // Subtle border color (clean separation)
        width: '1px', // Default border width
        style: 'solid', // Default border style
      },
      focus: {
        color: '#6750A4',
        width: '2px',
        style: 'solid',
        offset: '2px',
      },
      radius: {
        small: '4px', // For small elements (e.g., input fields)
        medium: '8px', // For standard elements (e.g., buttons, cards)
        large: '16px', // For larger elements (e.g., modals)
      },
    },


    /**
     * Typography System
     */
    typography: {
      fontFamily: "'Poppins', sans-serif, Arial", // Modern, clean typeface
      sizes: {
        display: { large: '60px', medium: '48px', small: '36px' },
        heading: {
          h1: '32px',
          h2: '28px',
          h3: '24px',
          h4: '20px',
          h5: '18px',
          h6: '16px',
        },
        body: { large: '18px', medium: '16px', default: '16px', small: '14px' },
        caption: { default: '12px', medium: '14px' },
      },
      lineHeight: {
        base: '1.5', // Comfortable reading experience
        heading: '1.2', // Tighter leading for headings
      },
      fontWeight: {
        regular: '400',
        medium: '500',
        semiBold: '600',
        bold: '700',
      },
      letterSpacing: {
        tight: '-0.02em', // For headings
        normal: '0',
        wide: '0.03em', // For specific elements
      },
    },

    /**
     * Spacing & Layout
     */
    spacing: {
      default: '8px',
      medium: '8px',
      small: '4px',
      large: '16px',
      xl: '24px',
      x4: '32px',
    },

    layout: {
      /**
       * Maximum width size for pages
       */
      maxPageWidth: '1280px',

      /**
       * Spacing between columns or elements
       */
      gutter: '24px',
    },

    /**
     * Visual Effects
     */
    effects: {
      shadows: {
        xs: '0px 1px 2px rgba(0, 0, 0, 0.08)', // Extra small shadow
        small: '0px 2px 4px rgba(0, 0, 0, 0.1)',
        medium: '0px 4px 8px rgba(0, 0, 0, 0.15)',
        large: '0px 8px 16px rgba(0, 0, 0, 0.2)',
        xLarge: '0px 12px 24px rgba(0, 0, 0, 0.25)', // Extra large shadow
        inset: 'inset 0px 1px 2px rgba(0, 0, 0, 0.1)', // Inset shadow for depth
        raised: '0px 4px 12px rgba(0, 0, 0, 0.15), 0px 2px 4px rgba(0, 0, 0, 0.1)', // Raised effect
      },
      opacity: { disabled: '0.5', hover: '0.8', faint: '0.2', semiOpaque: '0.7' },
      gradients: {
        primary: 'linear-gradient(to right, #6750A4, #9370DB)',
        secondary: 'linear-gradient(to bottom, #e8def8, #f0e6f8)',
        radial: 'radial-gradient(circle, #6750A4, #4A3780)', // Radial gradient example
      },
      blur: {
        small: 'blur(4px)',
        medium: 'blur(8px)',
        large: 'blur(16px)',
      },
    },

    /**
     * Interaction & Motion
     */
    interactions: {
      cursor: { pointer: 'pointer', disabled: 'not-allowed', text: 'text', grab: 'grab', grabbing: 'grabbing' },
      zIndex: { base: '1', modal: '100', tooltip: '200', overlay: '300', sticky: '50' },
      gradients: {
        primary: 'linear-gradient(135deg, #f9ad01, #FF4A49)',
        secondary: 'linear-gradient(135deg, #2C7489, #3A8EA6)',
        subtle: 'linear-gradient(to bottom, rgba(249, 249, 249, 0.8), rgba(240, 240, 240, 0.6))',
        codeBlock: 'linear-gradient(to right, rgba(245, 247, 250, 1), rgba(235, 240, 245, 1))',
      },
      transitions: {
        duration: { fast: '0.15s', medium: '0.3s', slow: '0.5s', verySlow: '1s' },
        easing: {
          easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
          easeOut: 'ease-out',
          easeIn: 'ease-in',
          spring: 'cubic-bezier(0.175, 0.885, 0.32, 1.275)' // Spring-like easing
        },
        property: {
          all: 'all',
          transform: 'transform',
          opacity: 'opacity',
          color: 'color',
          shadow: 'box-shadow',
        },
      },
      hoverEffect: {
        scale: 'scale(1.05)', // Slight scale on hover
        translateY: 'translateY(-2px)', // Slight vertical lift on hover
        shadow: '0px 6px 12px rgba(0, 0, 0, 0.18)', // Enhanced shadow on hover
      },
    },
  };

  return tokens;
}

// create a theme type schema to allow new theme to override
// or implement a different theme variation like dark theme.
// in case tokens are dynamically loaded from a json file, please declare this variable an an interface.
/**
 * Use tokens from this schema as css variables in your components.
 * For example, use \`surfaceColor\` as css variable \`--surface-color\`
 */
export type ${context.namePascalCase}Schema = ReturnType<typeof ${context.nameCamelCase}Tokens>;
`
  }
}
